// @osdk/client/build/browser/observable/internal/list/InterfaceListQuery.js
/*
 * Copyright 2025 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { additionalContext } from "../../../Client.js";
import { ObjectDefRef } from "../../../object/convertWireToOsdkObjects/InternalSymbols.js";
import { reloadDataAsFullObjects } from "../utils/reloadDataAsFullObjects.js";
import { ListQuery, PIVOT_IDX, RDP_IDX, RIDS_IDX } from "./ListQuery.js";
export class InterfaceListQuery extends ListQuery {
  createObjectSet(store) {
    const rdpConfig = this.cacheKey.otherKeys[RDP_IDX];
    const pivotInfo = this.cacheKey.otherKeys[PIVOT_IDX];
    const rids = this.cacheKey.otherKeys[RIDS_IDX];
    if (pivotInfo != null) {
      const sourceSet = createSourceSetForPivot(store, pivotInfo, rids);
      let objectSet = sourceSet.where(this.canonicalWhere).pivotTo(pivotInfo.linkName);
      if (rdpConfig != null) {
        objectSet = objectSet.withProperties(rdpConfig);
      }

      // intersectWith for pivot queries is deferred to fetchPageData
      // where the target type can be resolved asynchronously
      return objectSet;
    }
    const objectTypeDef = {
      type: "interface",
      apiName: this.apiName
    };
    const clientCtx = store.client[additionalContext];
    let objectSet;
    if (rids != null) {
      objectSet = clientCtx.objectSetFactory(objectTypeDef, clientCtx, {
        type: "static",
        objects: [...rids]
      });
    } else {
      objectSet = store.client(objectTypeDef);
    }
    if (rdpConfig != null) {
      objectSet = objectSet.withProperties(rdpConfig);
    }
    return objectSet.where(this.canonicalWhere);
  }
  async revalidateObjectType(objectType) {
    if (await super.revalidateObjectType(objectType)) return true;

    // For interface queries: also check if the invalidated concrete type
    // implements this query's interface. e.g. invalidating "Employee"
    // should revalidate a query for "Assignable" if Employee implements it.
    try {
      const objectMetadata = await this.store.client.fetchMetadata({
        type: "object",
        apiName: objectType
      });
      return this.apiName in objectMetadata.interfaceMap;
    } catch {
      return true;
    }
  }
  async postProcessFetchedData(data) {
    return reloadDataAsFullObjects(this.store.client, data);
  }
  wrapObject(object) {
    return this.options.resolveToObjectType ? object : object.$as(this.apiName);
  }
  createPayload(params) {
    const resolvedList = params.resolvedData?.map(obj => this.wrapObject(obj));
    return {
      ...super.createPayload(params),
      resolvedList
    };
  }
  extractRelevantObjects(changes) {
    const matchesApiName = ([, object]) => {
      return this.apiName in object[ObjectDefRef].interfaceMap;
    };
    const added = Array.from(changes.addedObjects).filter(matchesApiName).map(([, object]) => this.wrapObject(object));
    const modified = Array.from(changes.modifiedObjects).filter(matchesApiName).map(([, object]) => this.wrapObject(object));
    return {
      added: {
        all: added,
        strictMatches: new Set(),
        sortaMatches: new Set()
      },
      modified: {
        all: modified,
        strictMatches: new Set(),
        sortaMatches: new Set()
      }
    };
  }
}
function createSourceSetForPivot(store, pivotInfo, rids) {
  const clientCtx = store.client[additionalContext];
  if (rids != null) {
    return clientCtx.objectSetFactory({
      type: "object",
      apiName: pivotInfo.sourceType
    }, clientCtx, {
      type: "static",
      objects: [...rids]
    });
  }
  if (pivotInfo.sourceTypeKind === "interface") {
    return store.client({
      type: "interface",
      apiName: pivotInfo.sourceType
    });
  }
  return store.client({
    type: "object",
    apiName: pivotInfo.sourceType
  });
}
