// @osdk/api/build/esm/shapes/ShapeDefinition.js
/*
 * Copyright 2025 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const SourcePrimaryKeySymbol = Symbol.for("osdk.query.$sourcePk");
const LegacyShapesSymbol = Symbol.for("osdk.shapes.$primaryKey");
export function isSourcePkSymbol(value) {
  return typeof value === "symbol" && (value === SourcePrimaryKeySymbol || value === LegacyShapesSymbol);
}

/** Configuration for how a derived link is loaded at runtime. */

/** The result of calling `ShapeLinkBuilder.as()`, binding a link traversal to a target shape. */

/** A single link traversal step in a derived link's object set definition. */

/** Sort clause for a derived link's result set. */

/** A set operation (union/intersect/subtract) applied to a derived link's object set. */

/** Serializable definition of a derived link's object set: link traversals, filters, ordering, and set operations. */

/** How a shape handles null values for a selected property. */

/** Per-property configuration stored in a shape's `__props` map. */

/** A derived link definition stored in a shape's `__derivedLinks` array. */

/**
 * Type projection from an OSDK type. `__props` is the single source of truth for property config.
 *
 * @typeParam BASE - The base ObjectTypeDefinition or InterfaceDefinition
 * @typeParam SELECTED_PROPS - Record of selected properties and their transformed types
 * @typeParam DERIVED_LINKS - Record of derived link names to their result types
 */

export class ShapeNullabilityError extends Error {
  constructor(shape, violations) {
    const props = violations.map(v => v.property).join(", ");
    const shapeName = shape.__debugName ?? shape.__shapeId;
    super(`Shape "${shapeName}" requires these properties to be non-null: ${props}`);
    this.shape = shape;
    this.violations = violations;
    this.name = "ShapeNullabilityError";
  }
}
