// @osdk/api/build/esm/shapes/ShapeBuilder.js
/*
 * Copyright 2025 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { computeShapeId } from "./computeShapeId.js";
import { SourcePrimaryKeySymbol } from "./ShapeDefinition.js";
class ShapeBuilderImpl {
  #state;
  constructor(state) {
    this.#state = state;
  }
  #addProps(props, type) {
    const newProps = {
      ...this.#state.props
    };
    for (const prop of props) {
      newProps[prop] = {
        nullabilityOp: {
          type
        }
      };
    }
    return newProps;
  }
  select(...props) {
    return new ShapeBuilderImpl({
      ...this.#state,
      props: this.#addProps(props, "select")
    });
  }
  require(...props) {
    return new ShapeBuilderImpl({
      ...this.#state,
      props: this.#addProps(props, "require")
    });
  }
  dropIfNull(...props) {
    return new ShapeBuilderImpl({
      ...this.#state,
      props: this.#addProps(props, "dropIfNull")
    });
  }
  withDefault(prop, defaultValue) {
    return new ShapeBuilderImpl({
      ...this.#state,
      props: {
        ...this.#state.props,
        [prop]: {
          nullabilityOp: {
            type: "withDefault",
            defaultValue
          }
        }
      }
    });
  }
  withTransform(prop, transform) {
    return new ShapeBuilderImpl({
      ...this.#state,
      props: {
        ...this.#state.props,
        [prop]: {
          nullabilityOp: {
            type: "withTransform",
            transform
          }
        }
      }
    });
  }
  deriveLink(name, builder) {
    const linkBuilder = createShapeLinkBuilder(this.#state.baseTypeApiName);
    const linkResult = builder(linkBuilder);
    return new ShapeBuilderImpl({
      ...this.#state,
      props: {
        ...this.#state.props
      },
      derivedLinks: [...this.#state.derivedLinks, {
        name,
        objectSetDef: linkResult.objectSetDef,
        targetShape: linkResult.targetShape,
        config: linkResult.config
      }]
    });
  }
  build() {
    const shapeId = computeShapeId({
      baseTypeApiName: this.#state.baseTypeApiName,
      props: this.#state.props,
      derivedLinks: this.#state.derivedLinks
    });
    return {
      __shapeId: shapeId,
      __debugName: this.#state.debugName,
      __baseType: this.#state.baseType,
      __baseTypeApiName: this.#state.baseTypeApiName,
      __props: Object.freeze({
        ...this.#state.props
      }),
      __derivedLinks: Object.freeze([...this.#state.derivedLinks]),
      __selectedPropsType: {},
      __derivedLinksType: {}
    };
  }
}
class ShapeLinkBuilderImpl {
  #state;
  $primaryKey = SourcePrimaryKeySymbol;
  constructor(state) {
    this.#state = state;
  }
  pivotTo(link) {
    const newSegments = [...this.#state.segments, {
      type: "pivotTo",
      linkName: link,
      sourceType: this.#state.currentTypeApiName
      // targetType is resolved at link execution time when link metadata is available
    }];
    return new ShapeLinkBuilderImpl({
      ...this.#state,
      segments: newSegments,
      // currentTypeApiName is a hint for debugging; actual type resolution happens at execution
      currentTypeApiName: `${this.#state.currentTypeApiName}.${link}`
    });
  }

  /** Sets the where filter for this link. Calling multiple times replaces the previous clause (last-write-wins). */
  where(clause) {
    return new ShapeLinkBuilderImpl({
      ...this.#state,
      where: clause
    });
  }
  #setOp(type, others) {
    return new ShapeLinkBuilderImpl({
      ...this.#state,
      setOperations: [...this.#state.setOperations, ...others.map(other => ({
        type,
        other: other.toObjectSetDef()
      }))]
    });
  }
  union(...others) {
    return this.#setOp("union", others);
  }
  intersect(...others) {
    return this.#setOp("intersect", others);
  }
  subtract(...others) {
    return this.#setOp("subtract", others);
  }
  orderBy(property, direction = "asc") {
    return new ShapeLinkBuilderImpl({
      ...this.#state,
      orderBy: [...this.#state.orderBy, {
        property: property,
        direction
      }]
    });
  }
  limit(n) {
    if (!Number.isInteger(n) || n <= 0) {
      throw new Error(`ShapeLinkBuilder.limit() requires a positive integer, got: ${n}`);
    }
    return new ShapeLinkBuilderImpl({
      ...this.#state,
      limit: n
    });
  }
  distinct() {
    return new ShapeLinkBuilderImpl({
      ...this.#state,
      distinct: true
    });
  }
  as(shape, config = {}) {
    // The brand is a compile-time only marker, at runtime it's just undefined
    return {
      targetShape: shape,
      objectSetDef: this.toObjectSetDef(),
      config
    };
  }
  toObjectSetDef() {
    return {
      segments: Object.freeze([...this.#state.segments]),
      where: this.#state.where,
      orderBy: this.#state.orderBy.length > 0 ? Object.freeze([...this.#state.orderBy]) : undefined,
      limit: this.#state.limit,
      distinct: this.#state.distinct || undefined,
      setOperations: this.#state.setOperations.length > 0 ? Object.freeze([...this.#state.setOperations]) : undefined
    };
  }
}
export function createShapeLinkBuilder(sourceTypeApiName) {
  return new ShapeLinkBuilderImpl({
    sourceTypeApiName,
    currentTypeApiName: sourceTypeApiName,
    segments: [],
    orderBy: [],
    distinct: false,
    setOperations: []
  });
}

/**
 * Creates a new ShapeBuilder for the given base type.
 *
 * @param baseType - The base ObjectTypeDefinition or InterfaceDefinition
 * @param debugName - Optional name for debugging in DevTools
 *
 * @example
 * ```typescript
 * const SlimPlayer = createShapeBuilder(Player, "SlimPlayer")
 *   .require("name", "age")
 *   .withDefault("position", "Unknown")
 *   .build();
 * ```
 */
export function createShapeBuilder(baseType, debugName) {
  const state = {
    baseType,
    baseTypeApiName: baseType.apiName,
    debugName,
    props: {},
    derivedLinks: []
  };
  return new ShapeBuilderImpl(state);
}
