// @osdk/client/build/esm/observable/internal/aggregation/AggregationsHelper.js
/*
 * Copyright 2025 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { getWireObjectSet } from "../../../objectSet/createObjectSet.js";
import { AbstractHelper } from "../AbstractHelper.js";
import { ObjectAggregationQuery } from "./ObjectAggregationQuery.js";
export class AggregationsHelper extends AbstractHelper {
  constructor(store, cacheKeys, whereCanonicalizer, rdpCanonicalizer, intersectCanonicalizer) {
    super(store, cacheKeys);
    this.whereCanonicalizer = whereCanonicalizer;
    this.rdpCanonicalizer = rdpCanonicalizer;
    this.intersectCanonicalizer = intersectCanonicalizer;
  }
  observe(options, subFn) {
    return super.observe(options, subFn);
  }
  async observeAsync(options, subFn) {
    const query = this.getQueryWithObjectSet(options);
    await query.ensureInvalidationTypesReady();
    return this._subscribe(query, options, subFn);
  }
  getQuery(options) {
    return this.getOrCreateQuery(options, undefined);
  }
  getQueryWithObjectSet(options) {
    const serializedObjectSet = JSON.stringify(getWireObjectSet(options.objectSet));
    return this.getOrCreateQuery(options, serializedObjectSet);
  }
  getOrCreateQuery(options, serializedObjectSet) {
    const {
      type,
      where,
      withProperties,
      intersectWith,
      aggregate
    } = options;
    const {
      apiName
    } = type;
    const typeKind = "type" in type ? type.type : "interface";
    const canonWhere = this.whereCanonicalizer.canonicalize(where ?? {});
    const canonRdp = withProperties ? this.rdpCanonicalizer.canonicalize(withProperties) : undefined;
    const canonIntersect = intersectWith && intersectWith.length > 0 ? this.intersectCanonicalizer.canonicalize(intersectWith) : undefined;
    const canonAggregate = this.canonicalizeAggregate(aggregate);
    const aggregationCacheKey = this.cacheKeys.get("aggregation", typeKind, apiName, serializedObjectSet, canonWhere, canonRdp, canonIntersect, canonAggregate);
    return this.store.queries.get(aggregationCacheKey, () => {
      if (typeKind !== "object") {
        throw new Error("Only ObjectTypeDefinition is currently supported for aggregations");
      }
      return new ObjectAggregationQuery(this.store, this.store.subjects.get(aggregationCacheKey), aggregationCacheKey, options);
    });
  }
  canonicalizeAggregate(aggregate) {
    return JSON.parse(JSON.stringify(aggregate));
  }
}
