// @osdk/client/build/browser/observable/internal/links/LinksHelper.js
/*
 * Copyright 2025 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { AbstractHelper } from "../AbstractHelper.js";
import { SpecificLinkQuery } from "./SpecificLinkQuery.js";
export class LinksHelper extends AbstractHelper {
  constructor(store, cacheKeys, whereCanonicalizer, orderByCanonicalizer, selectCanonicalizer) {
    super(store, cacheKeys);
    this.whereCanonicalizer = whereCanonicalizer;
    this.orderByCanonicalizer = orderByCanonicalizer;
    this.selectCanonicalizer = selectCanonicalizer;
  }
  getQuery(options) {
    const {
      apiName,
      type: sourceTypeKind
    } = options.srcType;
    const canonWhere = this.whereCanonicalizer.canonicalize(options.where ?? {});
    const canonOrderBy = this.orderByCanonicalizer.canonicalize(options.orderBy ?? {});
    const canonSelect = options.select && options.select.length > 0 ? this.selectCanonicalizer.canonicalize(options.select) : undefined;
    const linkCacheKey = this.cacheKeys.get("specificLink", apiName, sourceTypeKind, options.sourceUnderlyingObjectType, options.pk, options.linkName, canonWhere, canonOrderBy, canonSelect, options.$includeAllBaseObjectProperties ? true : undefined, options.resolveToObjectType ? true : undefined);
    return this.store.queries.get(linkCacheKey, () => {
      return new SpecificLinkQuery(this.store, this.store.subjects.get(linkCacheKey), linkCacheKey, options);
    });
  }
}
