// @osdk/client/build/cjs/applyStreamingQuery-6ZZRUAB2.cjs
'use strict';

var chunkJDRCSS7F_cjs = require('./chunk-JDRCSS7F.cjs');
var chunkF4L4RFUT_cjs = require('./chunk-F4L4RFUT.cjs');
require('./chunk-ROJIXXGF.cjs');
var chunkH6PXPU6F_cjs = require('./chunk-H6PXPU6F.cjs');
require('./chunk-Q7SFCCGT.cjs');

// ../../node_modules/.pnpm/@osdk+foundry.functions@2.63.0/node_modules/@osdk/foundry.functions/build/esm/public/Query.js
var _streamingExecute = [1, "/v2/functions/queries/{0}/streamingExecute", 7, , "application/octet-stream"];
function streamingExecute($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _streamingExecute, ...args);
}

// src/queries/applyStreamingQuery.ts
async function* applyStreamingQuery(client, query, params) {
  const qd = await client.ontologyProvider.getQueryDefinition(query.apiName, query.isFixedVersion ? query.version : void 0);
  if (client.flushEdits != null) {
    await client.flushEdits();
  }
  const response = await streamingExecute(chunkF4L4RFUT_cjs.addUserAgentAndRequestContextHeaders(chunkF4L4RFUT_cjs.augmentRequestContext(client, (_) => ({
    finalMethodCall: "applyStreamingQuery"
  })), query), query.apiName, {
    ontology: await client.ontologyRid,
    parameters: params ? await chunkJDRCSS7F_cjs.remapQueryParams(params, client, qd.parameters) : {},
    version: query.isFixedVersion ? query.version : void 0,
    branch: client.branch
  }, {
    transactionId: client.transactionId,
    preview: true
  });
  if (response.body == null) {
    throw new Error("streamingExecute returned no response body");
  }
  const definitions = await chunkJDRCSS7F_cjs.getRequiredDefinitions(qd.output, client);
  const reader = response.body.getReader();
  for await (const line of chunkJDRCSS7F_cjs.parseNdjsonStream(chunkJDRCSS7F_cjs.iterateReadableStream(reader))) {
    if (line.type === "error") {
      const err = new Error(`${line.errorName} (${line.errorCode}) [${line.errorInstanceId}]: ${line.errorDescription ?? ""}`);
      Object.assign(err, line);
      throw err;
    }
    const remapped = await chunkJDRCSS7F_cjs.remapQueryResponse(client, qd.output, line.value, definitions);
    if (qd.output.type === "array" && Array.isArray(remapped)) {
      for (const item of remapped) {
        yield item;
      }
    } else {
      yield remapped;
    }
  }
}

exports.applyStreamingQuery = applyStreamingQuery;
