// @osdk/client/build/cjs/chunk-XLX6T4OB.cjs
'use strict';

var chunkYJG67XL4_cjs = require('./chunk-YJG67XL4.cjs');
var chunkTQUNF2DZ_cjs = require('./chunk-TQUNF2DZ.cjs');
var chunkPIWGMK3H_cjs = require('./chunk-PIWGMK3H.cjs');
var chunkCEBSLQQC_cjs = require('./chunk-CEBSLQQC.cjs');
var chunkJDRCSS7F_cjs = require('./chunk-JDRCSS7F.cjs');
var chunkF4L4RFUT_cjs = require('./chunk-F4L4RFUT.cjs');
var chunkROJIXXGF_cjs = require('./chunk-ROJIXXGF.cjs');
var chunkH6PXPU6F_cjs = require('./chunk-H6PXPU6F.cjs');
var chunkQ7SFCCGT_cjs = require('./chunk-Q7SFCCGT.cjs');
var unstable = require('@osdk/api/unstable');
var invariant2 = require('tiny-invariant');
var shared_client_impl = require('@osdk/shared.client.impl');
var api = require('@osdk/api');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var invariant2__default = /*#__PURE__*/_interopDefault(invariant2);

// src/actions/ActionValidationError.ts
var ActionValidationError = class extends Error {
  constructor(validation) {
    super("Validation Error: " + JSON.stringify(validation, null, 2));
    this.validation = validation;
  }
};

// ../../node_modules/.pnpm/@osdk+foundry.ontologies@2.63.0/node_modules/@osdk/foundry.ontologies/build/esm/public/Action.js
var _apply = [1, "/v2/ontologies/{0}/actions/{1}/apply", 7];
function apply($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _apply, ...args);
}
var _applyBatch = [1, "/v2/ontologies/{0}/actions/{1}/applyBatch", 3];
function applyBatch($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _applyBatch, ...args);
}

// ../../node_modules/.pnpm/@osdk+foundry.ontologies@2.63.0/node_modules/@osdk/foundry.ontologies/build/esm/public/ActionTypeV2.js
var _get = [0, "/v2/ontologies/{0}/actionTypes/{1}", 2];
function get($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _get, ...args);
}

// ../../node_modules/.pnpm/@osdk+foundry.ontologies@2.63.0/node_modules/@osdk/foundry.ontologies/build/esm/public/ObjectTypeV2.js
var _getFullMetadata = [0, "/v2/ontologies/{0}/objectTypes/{1}/fullMetadata", 2];
function getFullMetadata($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _getFullMetadata, ...args);
}

// ../../node_modules/.pnpm/@osdk+foundry.ontologies@2.63.0/node_modules/@osdk/foundry.ontologies/build/esm/public/OntologyInterface.js
var _get2 = [0, "/v2/ontologies/{0}/interfaceTypes/{1}", 2];
function get2($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _get2, ...args);
}

// ../../node_modules/.pnpm/@osdk+foundry.ontologies@2.63.0/node_modules/@osdk/foundry.ontologies/build/esm/public/OntologyScenario.js
var OntologyScenario_exports = {};
chunkQ7SFCCGT_cjs.__export(OntologyScenario_exports, {
  createScenario: () => createScenario,
  listScenarioEditedEntityTypes: () => listScenarioEditedEntityTypes,
  listScenarioEditedLinkTypes: () => listScenarioEditedLinkTypes,
  listScenarioEditedLinks: () => listScenarioEditedLinks,
  listScenarioEditedObjectTypes: () => listScenarioEditedObjectTypes,
  listScenarioEditedObjects: () => listScenarioEditedObjects
});
var _createScenario = [1, "/v2/ontologies/{0}/scenarios/create", 3];
function createScenario($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _createScenario, ...args);
}
var _listScenarioEditedObjectTypes = [0, "/v2/ontologies/{0}/scenarios/{1}/objectTypes/edited", 2];
function listScenarioEditedObjectTypes($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _listScenarioEditedObjectTypes, ...args);
}
var _listScenarioEditedLinkTypes = [0, "/v2/ontologies/{0}/scenarios/{1}/objectTypes/{2}/outgoingLinkTypes/edited", 2];
function listScenarioEditedLinkTypes($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _listScenarioEditedLinkTypes, ...args);
}
var _listScenarioEditedEntityTypes = [0, "/v2/ontologies/{0}/scenarios/{1}/editedEntityTypes", 2];
function listScenarioEditedEntityTypes($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _listScenarioEditedEntityTypes, ...args);
}
var _listScenarioEditedObjects = [0, "/v2/ontologies/{0}/scenarios/{1}/objects/{2}/edited", 2];
function listScenarioEditedObjects($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _listScenarioEditedObjects, ...args);
}
var _listScenarioEditedLinks = [0, "/v2/ontologies/{0}/scenarios/{1}/objects/{2}/links/{3}/edited", 2];
function listScenarioEditedLinks($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _listScenarioEditedLinks, ...args);
}

// ../../node_modules/.pnpm/@osdk+foundry.ontologies@2.63.0/node_modules/@osdk/foundry.ontologies/build/esm/public/QueryType.js
var _get3 = [0, "/v2/ontologies/{0}/queryTypes/{1}", 2];
function get3($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _get3, ...args);
}

// ../../node_modules/.pnpm/@osdk+foundry.ontologies@2.63.0/node_modules/@osdk/foundry.ontologies/build/esm/public/TimeSeriesPropertyV2.js
var _getFirstPoint = [0, "/v2/ontologies/{0}/objects/{1}/{2}/timeseries/{3}/firstPoint", 2];
function getFirstPoint($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _getFirstPoint, ...args);
}
var _getLastPoint = [0, "/v2/ontologies/{0}/objects/{1}/{2}/timeseries/{3}/lastPoint", 2];
function getLastPoint($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _getLastPoint, ...args);
}
var _streamPoints = [1, "/v2/ontologies/{0}/objects/{1}/{2}/timeseries/{3}/streamPoints", 3, , "*/*"];
function streamPoints($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _streamPoints, ...args);
}

// ../../node_modules/.pnpm/@osdk+foundry.ontologies@2.63.0/node_modules/@osdk/foundry.ontologies/build/esm/public/TimeSeriesValueBankProperty.js
var _getLatestValue = [0, "/v2/ontologies/{0}/objects/{1}/{2}/timeseries/{3}/latestValue", 2];
function getLatestValue($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _getLatestValue, ...args);
}
var _streamValues = [1, "/v2/ontologies/{0}/objects/{1}/{2}/timeseries/{3}/streamValues", 3, , "*/*"];
function streamValues($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _streamValues, ...args);
}

// src/scenarios/ScenarioClient.ts
function isScenarioClient(value) {
  return value != null && (typeof value === "object" || typeof value === "function") && typeof value.getScenarioReference === "function";
}
function buildScenarioClient(parent, scenarioRid) {
  const ctx = parent[chunkF4L4RFUT_cjs.additionalContext];
  if (ctx.transactionId != null) {
    console.warn("withScenario / createScenario: the supplied client has an active transaction. Ignoring the transaction and scoping to the scenario instead.");
  }
  if (ctx.scenarioRid != null) {
    throw new Error("withScenario / createScenario: the supplied client already has an active scenario. Scenarios cannot be nested.");
  }
  const inner = createClientWithScenario(scenarioRid, ctx.baseUrl, ctx.ontologyRid, ctx.tokenProvider, {
    logger: ctx.logger,
    UNSTABLE_DO_NOT_USE_BRANCH: ctx.branch
  }, ctx.fetch);
  const innerCtx = inner[chunkF4L4RFUT_cjs.additionalContext];
  async function getEditedEntityTypes() {
    const ontologyRid = await innerCtx.ontologyRid;
    const response = await OntologyScenario_exports.listScenarioEditedEntityTypes(innerCtx, ontologyRid, scenarioRid, {
      preview: true
    });
    return {
      objectTypes: response.objectTypes,
      linkTypes: response.linkTypes
    };
  }
  async function getEditedEntities(objectType, options) {
    const ontologyRid = await innerCtx.ontologyRid;
    const response = await OntologyScenario_exports.listScenarioEditedObjects(innerCtx, ontologyRid, scenarioRid, objectType.apiName, {
      pageSize: options?.pageSize,
      pageToken: options?.pageToken,
      preview: true
    });
    const data = response.data.map((entry) => {
      const wire = entry;
      return {
        $apiName: wire.__apiName,
        $primaryKey: wire.__primaryKey
      };
    });
    return {
      data,
      nextPageToken: response.nextPageToken
    };
  }
  async function* editedEntitiesAsyncIter(objectType, options) {
    const seen = /* @__PURE__ */ new Set();
    let pageToken;
    do {
      const page = await getEditedEntities(objectType, {
        pageSize: options?.pageSize,
        pageToken
      });
      for (const obj of page.data) {
        const key = obj.$primaryKey;
        if (seen.has(key)) continue;
        seen.add(key);
        yield obj;
      }
      pageToken = page.nextPageToken;
    } while (pageToken != null);
  }
  async function getEditedLinkTypes(sourceObjectType) {
    const ontologyRid = await innerCtx.ontologyRid;
    const response = await OntologyScenario_exports.listScenarioEditedLinkTypes(innerCtx, ontologyRid, scenarioRid, sourceObjectType.apiName, {
      preview: true
    });
    return response.data;
  }
  async function getEditedLinks(sourceObjectType, linkType, options) {
    const ontologyRid = await innerCtx.ontologyRid;
    const response = await OntologyScenario_exports.listScenarioEditedLinks(innerCtx, ontologyRid, scenarioRid, sourceObjectType.apiName, linkType, {
      pageSize: options?.pageSize,
      pageToken: options?.pageToken,
      preview: true
    });
    const data = [];
    for (const entry of response.data) {
      const sourceWire = entry.sourceObject;
      if (sourceWire?.__apiName == null || sourceWire.__primaryKey == null) {
        continue;
      }
      const source = {
        $apiName: sourceWire.__apiName,
        $primaryKey: sourceWire.__primaryKey
      };
      for (const linked of entry.linkedObjects) {
        const targetWire = linked.targetObject;
        if (targetWire?.__apiName == null || targetWire.__primaryKey == null) {
          continue;
        }
        const target = {
          $apiName: targetWire.__apiName,
          $primaryKey: targetWire.__primaryKey
        };
        data.push({
          source,
          target,
          linkType: linked.linkType ?? linkType
        });
      }
    }
    return {
      data,
      nextPageToken: response.nextPageToken
    };
  }
  async function* editedLinksAsyncIter(sourceObjectType, linkType, options) {
    let pageToken;
    do {
      const page = await getEditedLinks(sourceObjectType, linkType, {
        pageSize: options?.pageSize,
        pageToken
      });
      for (const link of page.data) {
        yield link;
      }
      pageToken = page.nextPageToken;
    } while (pageToken != null);
  }
  return Object.defineProperties(inner, {
    getScenarioReference: {
      value: () => scenarioRid
    },
    getEditedEntityTypes: {
      value: getEditedEntityTypes
    },
    getEditedEntities: {
      value: getEditedEntities
    },
    editedEntitiesAsyncIter: {
      value: editedEntitiesAsyncIter
    },
    getEditedLinkTypes: {
      value: getEditedLinkTypes
    },
    getEditedLinks: {
      value: getEditedLinks
    },
    editedLinksAsyncIter: {
      value: editedLinksAsyncIter
    }
  });
}

// src/util/isOntologyObjectV2.ts
function isOntologyObjectV2(o) {
  return o && typeof o === "object" && typeof o.__apiName === "string" && o.__primaryKey != null;
}

// src/util/toDataValue.ts
async function toDataValue(value, client, actionMetadata) {
  if (value == null) {
    return value;
  }
  if (Array.isArray(value) || value instanceof Set) {
    const values = Array.from(value);
    if (values.some((dataValue) => chunkJDRCSS7F_cjs.isAttachmentUpload(dataValue) || chunkJDRCSS7F_cjs.isAttachmentFile(dataValue))) {
      const converted = [];
      for (const value2 of values) {
        converted.push(await toDataValue(value2, client));
      }
      return converted;
    }
    const promiseArray = Array.from(value, async (innerValue) => await toDataValue(innerValue, client));
    return Promise.all(promiseArray);
  }
  if (chunkJDRCSS7F_cjs.isAttachmentUpload(value)) {
    const attachment = await chunkF4L4RFUT_cjs.upload(client, value.data, {
      filename: value.name
    });
    return await toDataValue(attachment.rid, client);
  }
  if (chunkJDRCSS7F_cjs.isAttachmentFile(value)) {
    const attachment = await chunkF4L4RFUT_cjs.upload(client, value, {
      filename: value.name
    });
    return await toDataValue(attachment.rid, client);
  }
  if (chunkJDRCSS7F_cjs.isMediaUpload(value)) {
    const mediaRef = await chunkROJIXXGF_cjs.MediaSet_exports.uploadMedia(client, value.data, {
      filename: value.fileName,
      preview: true
    });
    return await toDataValue(mediaRef, client);
  }
  if (chunkJDRCSS7F_cjs.isMedia(value)) {
    return value.getMediaReference();
  }
  if (chunkJDRCSS7F_cjs.isMediaReference(value)) {
    return value;
  }
  if (isOntologyObjectV2(value)) {
    return await toDataValue(value.__primaryKey, client);
  }
  if (chunkJDRCSS7F_cjs.isObjectSpecifiersObject(value)) {
    return await toDataValue(value.$primaryKey, client);
  }
  if (chunkF4L4RFUT_cjs.isWireObjectSet(value)) {
    return value;
  }
  if (chunkF4L4RFUT_cjs.isObjectSet(value)) {
    return chunkF4L4RFUT_cjs.getWireObjectSet(value);
  }
  if (chunkJDRCSS7F_cjs.isInterfaceActionParam(value)) {
    return {
      objectTypeApiName: value.$objectType,
      primaryKeyValue: value.$primaryKey
    };
  }
  if (isScenarioClient(value)) {
    return value.getScenarioReference();
  }
  if (typeof value === "object") {
    return Object.entries(value).reduce(async (promisedAcc, [key, structValue]) => {
      const acc = await promisedAcc;
      acc[key] = await toDataValue(structValue, client);
      return acc;
    }, Promise.resolve({}));
  }
  return value;
}

// src/actions/applyAction.ts
async function applyAction(client, action, parameters, options = {}) {
  const clientWithHeaders = chunkF4L4RFUT_cjs.addUserAgentAndRequestContextHeaders(chunkF4L4RFUT_cjs.augmentRequestContext(client, (_) => ({
    finalMethodCall: "applyAction"
  })), action);
  if (Array.isArray(parameters)) {
    !(client.transactionId == null) ? process.env.NODE_ENV !== "production" ? invariant2__default.default(false, "Batch actions are not supported for staged edit functions or when supplying a transaction ID") : invariant2__default.default(false) : void 0;
    !(client.scenarioRid == null) ? process.env.NODE_ENV !== "production" ? invariant2__default.default(false, "Batch actions are not supported when scoped to a scenario") : invariant2__default.default(false) : void 0;
    const response = await applyBatch(clientWithHeaders, await client.ontologyRid, action.unsanitizedApiName ?? action.apiName, {
      requests: parameters ? await remapBatchActionParams(parameters, client, await client.ontologyProvider.getActionDefinition(action.unsanitizedApiName ?? action.apiName)) : [],
      options: {
        returnEdits: options?.$returnEdits ? "ALL" : "NONE"
      }
    }, {
      branch: client.branch
    });
    const edits = response.edits;
    return options?.$returnEdits ? edits?.type === "edits" ? remapActionResponse(response) : edits : void 0;
  } else {
    const response = await apply(clientWithHeaders, await client.ontologyRid, action.unsanitizedApiName ?? action.apiName, {
      parameters: await remapActionParams(parameters, client, await client.ontologyProvider.getActionDefinition(action.unsanitizedApiName ?? action.apiName)),
      options: {
        mode: options?.$validateOnly ? "VALIDATE_ONLY" : "VALIDATE_AND_EXECUTE",
        returnEdits: options?.$returnEdits ? "ALL_V2_WITH_DELETIONS" : "NONE"
      }
    }, {
      branch: client.branch,
      transactionId: client.transactionId,
      scenarioRid: client.scenarioRid
    });
    if (options?.$validateOnly) {
      return response.validation;
    }
    if (response.validation && response.validation?.result === "INVALID") {
      const validation = response.validation;
      throw new ActionValidationError(validation);
    }
    const edits = response.edits;
    return options?.$returnEdits ? edits?.type === "edits" ? remapActionResponse(response) : edits : void 0;
  }
}
async function remapActionParams(params, client, actionMetadata) {
  if (params == null) {
    return {};
  }
  const parameterMap = {};
  for (const [key, value] of Object.entries(params)) {
    parameterMap[key] = await toDataValue(value, client);
  }
  return parameterMap;
}
async function remapBatchActionParams(params, client, actionMetadata) {
  const remappedParams = await Promise.all(params.map(async (param) => {
    return {
      parameters: await remapActionParams(param, client)
    };
  }));
  return remappedParams;
}
function remapActionResponse(response) {
  const editResponses = response?.edits;
  if (editResponses?.type === "edits") {
    const remappedActionResponse = {
      type: editResponses.type,
      deletedLinksCount: editResponses.deletedLinksCount,
      deletedObjectsCount: editResponses.deletedObjectsCount,
      addedLinks: [],
      deletedLinks: [],
      addedObjects: [],
      deletedObjects: [],
      modifiedObjects: [],
      editedObjectTypes: []
    };
    const editedObjectTypesSet = /* @__PURE__ */ new Set();
    for (const edit of editResponses.edits) {
      if (edit.type === "addLink" || edit.type === "deleteLink") {
        const osdkEdit = {
          linkTypeApiNameAtoB: edit.linkTypeApiNameAtoB,
          linkTypeApiNameBtoA: edit.linkTypeApiNameBtoA,
          aSideObject: edit.aSideObject,
          bSideObject: edit.bSideObject
        };
        edit.type === "addLink" ? remappedActionResponse.addedLinks.push(osdkEdit) : remappedActionResponse.deletedLinks?.push(osdkEdit);
        editedObjectTypesSet.add(edit.aSideObject.objectType);
        editedObjectTypesSet.add(edit.bSideObject.objectType);
      } else if (edit.type === "addObject" || edit.type === "deleteObject" || edit.type === "modifyObject") {
        const osdkEdit = {
          objectType: edit.objectType,
          primaryKey: edit.primaryKey
        };
        if (edit.type === "addObject") {
          remappedActionResponse.addedObjects.push(osdkEdit);
        } else if (edit.type === "deleteObject") {
          remappedActionResponse.deletedObjects?.push(osdkEdit);
        } else if (edit.type === "modifyObject") {
          remappedActionResponse.modifiedObjects.push(osdkEdit);
        }
        editedObjectTypesSet.add(edit.objectType);
      } else {
        if (process.env.NODE_ENV !== "production") {
          console.warn(`Unexpected edit type: ${JSON.stringify(edit)}`);
        }
      }
    }
    remappedActionResponse.editedObjectTypes = [...editedObjectTypesSet];
    return remappedActionResponse;
  }
}
function getTimeRange(body) {
  if ("$startTime" in body || "$endTime" in body) {
    return {
      type: "absolute",
      startTime: body.$startTime,
      endTime: body.$endTime
    };
  }
  return body.$before ? {
    type: "relative",
    startTime: {
      when: "BEFORE",
      value: body.$before,
      unit: api.TimeseriesDurationMapping[body.$unit]
    }
  } : {
    type: "relative",
    endTime: {
      when: "AFTER",
      value: body.$after,
      unit: api.TimeseriesDurationMapping[body.$unit]
    }
  };
}
async function* asyncIterPointsHelper(iterator) {
  const reader = iterator.body?.getReader();
  for await (const point of chunkJDRCSS7F_cjs.parseStreamedResponse(chunkJDRCSS7F_cjs.iterateReadableStream(reader))) {
    yield {
      time: point.time,
      value: point.value
    };
  }
}

// src/createGeotimeSeriesProperty.ts
var GeotimeSeriesPropertyImpl = class {
  #triplet;
  #client;
  constructor(client, objectApiName, primaryKey, propertyName, initialValue) {
    this.#client = client;
    this.#triplet = [objectApiName, primaryKey, propertyName];
    if (initialValue != null) {
      this.lastFetchedValue = initialValue;
    }
  }
  async getLatestValue() {
    const latestPointPromise = getLatestValue(this.#client, await this.#client.ontologyRid, ...this.#triplet);
    latestPointPromise.then(
      (latestPoint) => this.lastFetchedValue = latestPoint,
      // eslint-disable-next-line no-console
      (err) => void console.error(err)
    );
    return latestPointPromise;
  }
  async getAllValues(query) {
    const allPoints = [];
    for await (const point of this.asyncIterValues(query)) {
      allPoints.push(point);
    }
    return allPoints;
  }
  async *asyncIterValues(query) {
    const streamPointsIterator = await streamValues(this.#client, await this.#client.ontologyRid, ...this.#triplet, query ? {
      range: getTimeRange(query)
    } : {});
    for await (const timeseriesPoint of asyncIterPointsHelper(streamPointsIterator)) {
      yield timeseriesPoint;
    }
  }
};

// src/createMediaReferenceProperty.ts
var MediaReferencePropertyImpl = class {
  #mediaReference;
  #triplet;
  #client;
  constructor(args) {
    const {
      client,
      objectApiName,
      primaryKey,
      propertyName,
      mediaReference
    } = args;
    this.#client = client;
    this.#triplet = [objectApiName, primaryKey, propertyName];
    this.#mediaReference = mediaReference;
  }
  async fetchContents() {
    return chunkPIWGMK3H_cjs.getMediaContent(this.#client, await this.#client.ontologyRid, ...this.#triplet, {
      preview: true
      // TODO: Can turn this back off when backend is no longer in beta.
    });
  }
  async fetchMetadata() {
    const r = await chunkPIWGMK3H_cjs.getMediaMetadata(this.#client, await this.#client.ontologyRid, ...this.#triplet, {
      preview: true
      // TODO: Can turn this back off when backend is no longer in beta.
    });
    return {
      path: r.path,
      sizeBytes: Number(r.sizeBytes),
      mediaType: r.mediaType
    };
  }
  /**
   * Fetches type-specific metadata via `MediaSets.metadata`. There is no Ontologies-backed
   * full-metadata endpoint, so this routes through the media set view using the embedded media
   * reference.
   */
  async fetchFullMetadata() {
    const {
      mediaSetRid,
      mediaItemRid,
      token
    } = this.#mediaReference.reference.mediaSetViewItem;
    const raw = await chunkROJIXXGF_cjs.MediaSet_exports.metadata(this.#client, mediaSetRid, mediaItemRid, {
      preview: true
    }, token ? {
      ReadToken: token
    } : void 0);
    return {
      itemMetadata: chunkF4L4RFUT_cjs.validateMediaItemMetadata(raw)
    };
  }
  getMediaReference() {
    return this.#mediaReference;
  }
};

// src/createTimeseriesProperty.ts
var TimeSeriesPropertyImpl = class {
  #triplet;
  #client;
  constructor(client, objectApiName, primaryKey, propertyName) {
    this.#client = client;
    this.#triplet = [objectApiName, primaryKey, propertyName];
  }
  async getFirstPoint() {
    return getFirstPoint(this.#client, await this.#client.ontologyRid, ...this.#triplet);
  }
  async getLastPoint() {
    return getLastPoint(this.#client, await this.#client.ontologyRid, ...this.#triplet);
  }
  async getAllPoints(query) {
    const allPoints = [];
    for await (const point of this.asyncIterPoints(query)) {
      allPoints.push(point);
    }
    return allPoints;
  }
  async *asyncIterPoints(query) {
    const streamPointsIterator = await streamPoints(this.#client, await this.#client.ontologyRid, ...this.#triplet, query ? {
      range: getTimeRange(query)
    } : {});
    for await (const timeseriesPoint of asyncIterPointsHelper(streamPointsIterator)) {
      yield timeseriesPoint;
    }
  }
};

// src/object/formatting/formatBoolean.ts
function formatBoolean(value, rule) {
  return value ? rule.valueIfTrue : rule.valueIfFalse;
}

// src/object/formatting/propertyFormattingUtils.ts
function resolvePropertyReference(ref, objectData) {
  if (ref.type === "constant") {
    return ref.value;
  } else if (ref.type === "propertyType") {
    const value = objectData[ref.propertyApiName];
    return value != null ? String(value) : void 0;
  }
  return void 0;
}
function getBrowserLocale() {
  if (typeof navigator !== "undefined" && navigator.language != null) {
    return navigator.language;
  }
  return "en-US";
}

// src/object/formatting/formatDateTime.ts
function formatDateTime(value, format, timezone, objectData, locale, userTimezoneOverride) {
  const date = value instanceof Date ? value : new Date(value);
  if (isNaN(date.getTime())) {
    return void 0;
  }
  const resolvedTimezone = resolveTimezone(timezone, objectData, userTimezoneOverride);
  switch (format.type) {
    case "localizedFormat":
      return formatLocalized(date, format, locale, resolvedTimezone);
    case "stringFormat":
      return void 0;
    default:
      return void 0;
  }
}
var INVALID_DATE_STRING = "Invalid date or timezone";
function formatLocalized(date, format, locale, timezone) {
  if (format.format === "DATE_FORMAT_ISO_INSTANT") {
    return date.toISOString();
  }
  if (format.format === "DATE_FORMAT_RELATIVE_TO_NOW") {
    return formatRelativeToNow(date, locale, timezone);
  }
  const options = getLocalizedFormatOptions(format.format);
  try {
    return new Intl.DateTimeFormat(locale, timezone ? {
      ...options,
      timeZone: timezone
    } : options).format(date);
  } catch (_e) {
    return INVALID_DATE_STRING;
  }
}
function getLocalizedFormatOptions(format) {
  switch (format) {
    case "DATE_FORMAT_DATE":
      return FORMAT_DATE;
    case "DATE_FORMAT_YEAR_AND_MONTH":
      return FORMAT_YEAR_AND_MONTH;
    case "DATE_FORMAT_DATE_TIME":
      return FORMAT_DATE_TIME;
    case "DATE_FORMAT_DATE_TIME_SHORT":
      return FORMAT_DATE_TIME_SHORT;
    case "DATE_FORMAT_TIME":
      return FORMAT_TIME;
    default:
      return {};
  }
}
var FORMAT_DATE = {
  day: "numeric",
  month: "short",
  weekday: "short",
  year: "numeric"
};
var FORMAT_YEAR_AND_MONTH = {
  month: "short",
  year: "numeric"
};
var FORMAT_DATE_TIME = {
  day: "numeric",
  month: "short",
  weekday: "short",
  year: "numeric",
  hour: "numeric",
  minute: "numeric",
  second: "numeric"
};
var FORMAT_DATE_TIME_SHORT = {
  day: "numeric",
  month: "short",
  year: "numeric",
  hour: "numeric",
  minute: "numeric"
};
var FORMAT_TIME = {
  hour: "numeric",
  minute: "numeric",
  second: "numeric"
};
var DATE_TIME_SHORT_WITH_WEEKDAY = {
  day: "numeric",
  month: "short",
  weekday: "short",
  year: "numeric",
  hour: "numeric",
  minute: "numeric"
};
function resolveTimezone(timezone, objectData, userTimezoneOverride) {
  if (userTimezoneOverride != null) {
    return userTimezoneOverride;
  }
  if (timezone == null) {
    return void 0;
  }
  switch (timezone.type) {
    case "user":
      return void 0;
    case "static":
      return resolvePropertyReference(timezone.zoneId, objectData);
    default:
      return void 0;
  }
}
var SECOND_MS = 1e3;
var MINUTE_MS = 60 * SECOND_MS;
var HOUR_MS = 60 * MINUTE_MS;
var DAY_MS = 24 * HOUR_MS;
function formatRelativeToNow(date, locale, timezone) {
  const now = Date.now();
  const diff = date.valueOf() - now;
  const absDiff = Math.abs(diff);
  if (absDiff >= DAY_MS) {
    const dtf = new Intl.DateTimeFormat(locale, {
      ...DATE_TIME_SHORT_WITH_WEEKDAY,
      timeZone: timezone
    });
    return dtf.format(date);
  }
  const rtf = new Intl.RelativeTimeFormat(locale, {
    numeric: "auto"
  });
  const units = [["hour", HOUR_MS], ["minute", MINUTE_MS], ["second", SECOND_MS]];
  for (const [unit, ms] of units) {
    if (absDiff >= ms) {
      return rtf.format(Math.floor(diff / ms), unit);
    }
  }
  return rtf.format(0, "second");
}

// src/object/formatting/formatNumber.ts
function formatNumber(value, numberType, objectData, locale) {
  switch (numberType.type) {
    case "standard":
      return formatStandardNumber(value, numberType.baseFormatOptions, locale);
    case "fixedValues":
      return numberType.values[value];
    case "currency":
      return formatCurrency(value, numberType, objectData, locale);
    case "standardUnit":
      return formatStandardUnit(value, numberType, objectData, locale);
    case "customUnit":
      return formatCustomUnit(value, numberType, objectData, locale);
    case "affix":
      return formatAffix(value, numberType, objectData, locale);
    case "duration":
      return void 0;
    case "scale":
      return formatScale(value, numberType, locale);
    case "ratio":
      return formatRatio(value, numberType, locale);
    default:
      return void 0;
  }
}
function formatStandardNumber(value, options, locale) {
  return formatNumberWithAffixes(value, options, locale);
}
function formatCurrency(value, rule, objectData, locale) {
  const currencyCode = resolvePropertyReference(rule.currencyCode, objectData);
  if (currencyCode == null) {
    return formatStandardNumber(value, rule.baseFormatOptions, locale);
  }
  const options = {
    ...convertToIntlOptions(rule.baseFormatOptions),
    style: "currency",
    currency: currencyCode,
    currencyDisplay: rule.style === "COMPACT" ? "narrowSymbol" : "symbol"
  };
  return formatWithIntl(value, options, locale);
}
function formatStandardUnit(value, rule, objectData, locale) {
  const unit = resolvePropertyReference(rule.unit, objectData);
  if (unit == null) {
    return formatStandardNumber(value, rule.baseFormatOptions, locale);
  }
  const options = {
    ...convertToIntlOptions(rule.baseFormatOptions),
    style: "unit",
    unit
  };
  try {
    return formatWithIntl(value, options, locale);
  } catch (error) {
    const suffix = ` ${unit}`;
    return formatNumberWithAffixes(value, rule.baseFormatOptions, locale, "", suffix);
  }
}
function formatCustomUnit(value, rule, objectData, locale) {
  const unit = resolvePropertyReference(rule.unit, objectData);
  const suffix = unit != null ? ` ${unit}` : "";
  return formatNumberWithAffixes(value, rule.baseFormatOptions, locale, "", suffix);
}
function formatAffix(value, rule, objectData, locale) {
  const prefix = rule.affix.prefix != null ? resolvePropertyReference(rule.affix.prefix, objectData) ?? "" : "";
  const suffix = rule.affix.postfix != null ? resolvePropertyReference(rule.affix.postfix, objectData) ?? "" : "";
  return formatNumberWithAffixes(value, rule.baseFormatOptions, locale, prefix, suffix);
}
function getScaleDivisor(scaleType) {
  switch (scaleType) {
    case "THOUSANDS":
      return 1e3;
    case "MILLIONS":
      return 1e6;
    case "BILLIONS":
      return 1e9;
    default:
      throw new Error(`Unknown scale type: ${scaleType}`);
  }
}
function getLocalizedCompactSuffix(scaleDivisor, locale) {
  const compactFormatter = new Intl.NumberFormat(locale, {
    notation: "compact",
    compactDisplay: "short"
  });
  const parts = compactFormatter.formatToParts(scaleDivisor);
  const compactPart = parts.find((p) => p.type === "compact");
  return compactPart?.value ?? "";
}
function formatScale(value, rule, locale) {
  const scaleDivisor = getScaleDivisor(rule.scaleType);
  const scaledValue = value / scaleDivisor;
  const suffix = getLocalizedCompactSuffix(scaleDivisor, locale);
  return formatNumberWithAffixes(scaledValue, rule.baseFormatOptions, locale, "", suffix);
}
function getRatioScaledValue(value, ratioType) {
  switch (ratioType) {
    case "PERCENTAGE":
      return value * 100;
    case "PER_MILLE":
      return value * 1e3;
    case "BASIS_POINTS":
      return value * 1e4;
    default:
      throw new Error(`Unknown ratio type: ${ratioType}`);
  }
}
function formatRatio(value, rule, locale) {
  if (rule.ratioType === "PERCENTAGE") {
    const options = {
      ...convertToIntlOptions(rule.baseFormatOptions),
      style: "percent"
    };
    return formatWithIntl(value, options, locale);
  }
  const scaledValue = getRatioScaledValue(value, rule.ratioType);
  let suffix;
  switch (rule.ratioType) {
    case "PER_MILLE":
      suffix = "\u2030";
      break;
    case "BASIS_POINTS":
      suffix = " bps";
      break;
    default:
      rule.ratioType;
      throw new Error(`Unknown ratio type: ${rule.ratioType}`);
  }
  return formatNumberWithAffixes(scaledValue, rule.baseFormatOptions, locale, "", suffix);
}
function mapNotation(notation) {
  switch (notation) {
    case "STANDARD":
      return "standard";
    case "SCIENTIFIC":
      return "scientific";
    case "ENGINEERING":
      return "engineering";
    case "COMPACT":
      return "compact";
  }
}
function mapRoundingMode(mode) {
  switch (mode) {
    case "CEIL":
      return "ceil";
    case "FLOOR":
      return "floor";
    case "ROUND_CLOSEST":
      return "halfExpand";
    default:
      return void 0;
  }
}
function convertToIntlOptions(options) {
  return {
    useGrouping: options.useGrouping,
    minimumIntegerDigits: options.minimumIntegerDigits,
    minimumFractionDigits: options.minimumFractionDigits,
    maximumFractionDigits: options.maximumFractionDigits,
    minimumSignificantDigits: options.minimumSignificantDigits,
    maximumSignificantDigits: options.maximumSignificantDigits,
    notation: options.notation ? mapNotation(options.notation) : void 0,
    roundingMode: options.roundingMode ? mapRoundingMode(options.roundingMode) : void 0,
    convertNegativeToParenthesis: options.convertNegativeToParenthesis
  };
}
function maybeConvertNegativeToParenthesis(formatted, value, shouldConvert) {
  if (shouldConvert && value < 0) {
    return formatted.replace(/^-/, "(") + ")";
  }
  return formatted;
}
function formatWithIntl(value, options, locale) {
  const formatter = new Intl.NumberFormat(locale, options);
  const formatted = formatter.format(value);
  return maybeConvertNegativeToParenthesis(formatted, value, options.convertNegativeToParenthesis ?? false);
}
function formatNumberWithAffixes(value, baseOptions, locale, prefix, suffix) {
  const intlOptions = convertToIntlOptions(baseOptions);
  const formatted = formatWithIntl(value, intlOptions, locale);
  return `${prefix || ""}${formatted}${suffix || ""}`;
}

// src/object/formatting/applyPropertyFormatter.ts
function applyPropertyFormatter(propertyValue, propertyDefinition, objectData, options = {}) {
  if (propertyDefinition?.valueFormatting == null || propertyValue == null) {
    return void 0;
  }
  return formatPropertyValue(propertyValue, propertyDefinition.valueFormatting, objectData, options);
}
function formatPropertyValue(value, rule, objectData, options) {
  switch (rule.type) {
    case "boolean":
      if (typeof value !== "boolean") {
        return void 0;
      }
      return formatBoolean(value, rule);
    case "number":
      if (typeof value !== "number") {
        return void 0;
      }
      return formatNumber(value, rule.numberType, objectData, options.locale ?? getBrowserLocale());
    case "date":
    case "timestamp":
      if (typeof value !== "string") {
        return void 0;
      }
      return formatDateTime(new Date(value), rule.format, rule.type === "timestamp" ? rule.displayTimezone : void 0, objectData, options.locale ?? getBrowserLocale(), options.timezoneId);
    default:
      return void 0;
  }
}

// src/ontology/OntologyProvider.ts
var InterfaceDefinitions = Symbol("InterfaceDefinitions" );

// src/object/convertWireToOsdkObjects/InternalSymbols.ts
var UnderlyingOsdkObject = Symbol("Underlying Object" );
var ObjectDefRef = Symbol("ObjectDefinition" );
var InterfaceDefRef = Symbol("InterfaceDefinition" );
var ClientRef = Symbol("ClientRef" );
var PropertySecuritiesRef = Symbol("Property Securities" );

// src/object/convertWireToOsdkObjects/getDollarLink.ts
function get$link(holder) {
  const client = holder[ClientRef];
  const objDef = holder[ObjectDefRef];
  const rawObj = holder[UnderlyingOsdkObject];
  return Object.freeze(Object.fromEntries(Object.keys(objDef.links).map((linkName) => {
    const linkDef = objDef.links[linkName];
    const objectSet = client.objectSetFactory(objDef, client).where({
      [objDef.primaryKeyApiName]: rawObj.$primaryKey
    }).pivotTo(linkName);
    const value = !linkDef.multiplicity ? {
      fetchOne: (options) => chunkF4L4RFUT_cjs.fetchSingle(client, objDef, options ?? {}, chunkF4L4RFUT_cjs.getWireObjectSet(objectSet)),
      fetchOneWithErrors: (options) => chunkF4L4RFUT_cjs.fetchSingleWithErrors(client, objDef, options ?? {}, chunkF4L4RFUT_cjs.getWireObjectSet(objectSet))
    } : objectSet;
    return [linkName, value];
  })));
}
function get$linkForInterface(holder) {
  const client = holder[UnderlyingOsdkObject][ClientRef];
  const objDef = holder[UnderlyingOsdkObject][ObjectDefRef];
  const interfaceDef = holder[InterfaceDefRef];
  const rawObj = holder[UnderlyingOsdkObject];
  return Object.freeze(Object.fromEntries(Object.keys(interfaceDef.links).map((linkName) => {
    const linkDef = interfaceDef.links[linkName];
    const objectSet = client.objectSetFactory(interfaceDef, client).intersect(client.objectSetFactory(objDef, client).where({
      [objDef.primaryKeyApiName]: rawObj.$primaryKey
    })).pivotTo(linkName);
    const linkTargetDef = linkDef.targetType === "object" ? {
      type: "object",
      apiName: linkDef.targetTypeApiName
    } : {
      type: "interface",
      apiName: linkDef.targetTypeApiName
    };
    const value = !linkDef.multiplicity ? {
      fetchOne: (options) => chunkF4L4RFUT_cjs.fetchSingle(client, linkTargetDef, options ?? {}, chunkF4L4RFUT_cjs.getWireObjectSet(objectSet)),
      fetchOneWithErrors: (options) => chunkF4L4RFUT_cjs.fetchSingleWithErrors(client, linkTargetDef, options ?? {}, chunkF4L4RFUT_cjs.getWireObjectSet(objectSet))
    } : objectSet;
    return [linkName, value];
  })));
}

// src/object/convertWireToOsdkObjects/createOsdkInterface.ts
function extractValueByImplementation(underlying, impl) {
  switch (impl.type) {
    case "localProperty":
      return underlying[impl.propertyApiName];
    case "structField": {
      const struct = underlying[impl.propertyApiName];
      return struct == null ? void 0 : struct[impl.structFieldApiName];
    }
    case "struct":
      return Object.fromEntries(Object.entries(impl.mapping).map(([fieldName, entry]) => {
        if (entry.type === "structFieldOfProperty") {
          const struct = underlying[entry.propertyApiName];
          return [fieldName, struct == null ? void 0 : struct[entry.structFieldApiName]];
        }
        return [fieldName, underlying[entry.propertyApiName]];
      }));
    case "reduced":
      return extractValueByImplementation(underlying, impl.implementation);
  }
}
function createOsdkInterface(underlying, interfaceDef) {
  const [objApiNamespace] = chunkF4L4RFUT_cjs.extractNamespace(interfaceDef.apiName);
  const objDef = underlying[ObjectDefRef];
  return Object.freeze(Object.defineProperties({}, {
    // first to minimize hidden classes
    [UnderlyingOsdkObject]: {
      value: underlying
    },
    "$apiName": {
      value: interfaceDef.apiName,
      enumerable: true
    },
    "$as": {
      value: underlying.$as,
      enumerable: false
    },
    "$objectType": {
      value: underlying.$objectType,
      enumerable: "$objectType" in underlying
    },
    "$primaryKey": {
      value: underlying.$primaryKey,
      enumerable: "$primaryKey" in underlying
    },
    "$objectSpecifier": {
      value: underlying.$objectSpecifier,
      enumerable: "$objectSpecifier" in underlying
    },
    "$title": {
      value: underlying.$title,
      enumerable: "$title" in underlying
    },
    "$rid": {
      value: underlying.$rid,
      enumerable: "$rid" in underlying
    },
    "$clone": {
      value: clone,
      enumerable: false
    },
    "$propertySecurities": {
      value: remapPropertySecuritiesForInterface(underlying.$propertySecurities, underlying[ObjectDefRef], interfaceDef, objApiNamespace),
      enumerable: "$propertySecurities" in underlying
    },
    "$__EXPERIMENTAL__NOT_SUPPORTED_YET__metadata": {
      value: {
        "ObjectMetadata": underlying[ObjectDefRef],
        "InterfaceMetadata": interfaceDef
      },
      enumerable: false
    },
    "$__EXPERIMENTAL__NOT_SUPPORTED_YET__getFormattedValue": {
      value: underlying.$__EXPERIMENTAL__NOT_SUPPORTED_YET__getFormattedValue,
      enumerable: false
    },
    "$link": {
      get() {
        return get$linkForInterface(this);
      }
    },
    [InterfaceDefRef]: {
      value: interfaceDef
    },
    ...Object.fromEntries(Object.keys(interfaceDef.properties).map((p) => {
      const [apiNamespace, apiName] = chunkF4L4RFUT_cjs.extractNamespace(p);
      const exposedName = apiNamespace === objApiNamespace ? apiName : p;
      const impl = objDef.interfaceImplementations?.[interfaceDef.apiName]?.[p];
      if (impl != null) {
        const value = extractValueByImplementation(underlying, impl);
        return [exposedName, {
          enumerable: value !== void 0,
          value
        }];
      }
      const targetPropName = objDef.interfaceMap[interfaceDef.apiName][p];
      return [exposedName, {
        enumerable: targetPropName in underlying,
        value: underlying[targetPropName]
      }];
    }))
  }));
  function clone(update) {
    if (update == null) {
      return underlying.$clone().$as(interfaceDef);
    }
    for (const key of Object.keys(update)) {
      if (!(key in interfaceDef.properties)) {
        throw new Error(`Invalid property ${key} for interface ${interfaceDef.apiName}`);
      }
    }
    const remappedProps = Object.fromEntries(Object.keys(update).map((p) => mapProperty(p, update[p])).filter((x) => x != null));
    return underlying.$clone(remappedProps).$as(interfaceDef);
  }
  function mapProperty(propertyName, value) {
    const objDef2 = underlying[ObjectDefRef];
    const targetPropName = objDef2.interfaceMap[interfaceDef.apiName][propertyName];
    if (targetPropName == null) {
      throw new Error(`Cannot clone interface with ${propertyName} as property is not implemented by the underlying object type ${objDef2.apiName}`);
    }
    return [targetPropName, value];
  }
}
function remapPropertySecuritiesForInterface(underlyingSecurities, objDef, interfaceDef, objApiNamespace) {
  if (underlyingSecurities == null) return void 0;
  const inverseMap = objDef.inverseInterfaceMap?.[interfaceDef.apiName] ?? {};
  const remapped = {};
  for (const objPropName of Object.keys(underlyingSecurities)) {
    if (objPropName === "$primaryKey" || objPropName === "$title") {
      remapped[objPropName] = underlyingSecurities[objPropName];
      continue;
    }
    const interfacePropName = inverseMap[objPropName];
    if (interfacePropName == null) continue;
    const [apiNamespace, apiName] = chunkF4L4RFUT_cjs.extractNamespace(interfacePropName);
    const key = apiNamespace === objApiNamespace ? apiName : interfacePropName;
    remapped[key] = underlyingSecurities[objPropName];
  }
  return remapped;
}

// src/object/convertWireToOsdkObjects/getDollarAs.ts
var get$as = chunkYJG67XL4_cjs.createSimpleCache(/* @__PURE__ */ new WeakMap(), $asFactory).get;
var osdkObjectToInterfaceView = chunkYJG67XL4_cjs.createSimpleCache(/* @__PURE__ */ new WeakMap(), () => /* @__PURE__ */ new Map());
function assertInterfaceToOtCastIsPermitted(holder, objDef) {
  const interfaceDef = holder[InterfaceDefRef];
  if (interfaceDef == null) return;
  const implementations = objDef.interfaceImplementations?.[interfaceDef.apiName];
  if (implementations == null) return;
  for (const [iptApiName, impl] of Object.entries(implementations)) {
    if (impl.type === "localProperty") continue;
    throw new Error(`Cannot cast interface view of '${interfaceDef.apiName}' to '${objDef.apiName}': property '${iptApiName}' has a non-local implementation (${impl.type}), so the underlying object type cannot be faithfully represented. Load the object type directly.`);
  }
}
function $asFactory(objDef) {
  return function $as(targetMinDef) {
    let targetInterfaceApiName;
    if (typeof targetMinDef === "string") {
      if (targetMinDef === objDef.apiName) {
        assertInterfaceToOtCastIsPermitted(this, objDef);
        return this[UnderlyingOsdkObject];
      }
      if (objDef.interfaceMap?.[targetMinDef] == null) {
        throw new Error(`Object does not implement interface '${targetMinDef}'.`);
      }
      targetInterfaceApiName = targetMinDef;
    } else if (targetMinDef.apiName === objDef.apiName) {
      assertInterfaceToOtCastIsPermitted(this, objDef);
      return this[UnderlyingOsdkObject];
    } else {
      if (targetMinDef.type === "object") {
        throw new Error(`'${targetMinDef.apiName}' is not an interface nor is it '${objDef.apiName}', which is the object type.`);
      }
      targetInterfaceApiName = targetMinDef.apiName;
    }
    const def = objDef[InterfaceDefinitions][targetInterfaceApiName];
    if (!def) {
      throw new Error(`Object does not implement interface '${targetInterfaceApiName}'.`);
    }
    const underlying = this[UnderlyingOsdkObject];
    const existing = osdkObjectToInterfaceView.get(underlying).get(targetInterfaceApiName)?.deref();
    if (existing) return existing;
    const osdkInterface = createOsdkInterface(underlying, def.def);
    osdkObjectToInterfaceView.get(underlying).set(targetInterfaceApiName, new WeakRef(osdkInterface));
    return osdkInterface;
  };
}

// src/object/convertWireToOsdkObjects/createOsdkObject.ts
var specialPropertyTypes = /* @__PURE__ */ new Set(["attachment", "geotimeSeriesReference", "mediaReference", "numericTimeseries", "stringTimeseries", "sensorTimeseries"]);
var securableSpecialKeys = /* @__PURE__ */ new Set(["$primaryKey", "$title"]);
var basePropDefs = {
  "$as": {
    get() {
      return get$as(this[ObjectDefRef]);
    }
  },
  "$link": {
    get() {
      return get$link(this);
    }
  },
  "$clone": {
    value(update) {
      const rawObj = this[UnderlyingOsdkObject];
      const def = this[ObjectDefRef];
      if (update == null) {
        return createOsdkObject(this[ClientRef], def, {
          ...rawObj
        });
      }
      if (def.primaryKeyApiName in update && rawObj[def.primaryKeyApiName] !== update[def.primaryKeyApiName]) {
        throw new Error(`Cannot update ${def.apiName} object with differing primary key values `);
      }
      if (def.titleProperty in update && !("$title" in update)) {
        update.$title = update[def.titleProperty];
      }
      const newObject = {
        ...this[UnderlyingOsdkObject],
        ...update
      };
      return createOsdkObject(this[ClientRef], this[ObjectDefRef], newObject);
    }
  },
  "$objectSpecifier": {
    get() {
      const rawObj = this[UnderlyingOsdkObject];
      return chunkJDRCSS7F_cjs.createObjectSpecifierFromPrimaryKey(this[ObjectDefRef], rawObj.$primaryKey);
    },
    enumerable: true
  },
  "$propertySecurities": {
    get() {
      return this[PropertySecuritiesRef];
    },
    enumerable: true
  },
  "$__EXPERIMENTAL__NOT_SUPPORTED_YET__metadata": {
    get() {
      return {
        ObjectMetadata: this[ObjectDefRef]
      };
    },
    enumerable: false
  },
  "$__EXPERIMENTAL__NOT_SUPPORTED_YET__getFormattedValue": {
    value(propertyApiName, options) {
      const rawObj = this[UnderlyingOsdkObject];
      const def = this[ObjectDefRef];
      const propertyValue = rawObj[propertyApiName];
      return applyPropertyFormatter(propertyValue, def.properties[propertyApiName], rawObj, options);
    },
    enumerable: false
  }
};
function createOsdkObject(client, objectDef, simpleOsdkProperties, derivedPropertyTypeByName = {}, wirePropertySecurities = []) {
  const {
    parsedObject,
    clientPropertySecurities
  } = parseWhenSecuritiesLoaded(wirePropertySecurities, simpleOsdkProperties, objectDef, derivedPropertyTypeByName);
  const rawObj = parsedObject;
  Object.defineProperties(rawObj, {
    [UnderlyingOsdkObject]: {
      enumerable: false,
      value: simpleOsdkProperties
    },
    [PropertySecuritiesRef]: {
      enumerable: false,
      value: clientPropertySecurities
    },
    [ObjectDefRef]: {
      value: objectDef,
      enumerable: false
    },
    // TODO: Potentially update when GA metadata field
    [ClientRef]: {
      value: client,
      enumerable: false
    },
    ...basePropDefs
  });
  for (const propKey of Object.keys(rawObj)) {
    if (propKey in objectDef.properties && typeof objectDef.properties[propKey].type === "string" && specialPropertyTypes.has(objectDef.properties[propKey].type)) {
      rawObj[propKey] = createSpecialProperty(client, objectDef, rawObj, propKey);
    } else if (propKey in derivedPropertyTypeByName) {
      rawObj[propKey] = modifyRdpProperties(client, derivedPropertyTypeByName, rawObj[propKey], propKey);
    }
  }
  return Object.freeze(rawObj);
}
function modifyRdpProperties(client, derivedPropertyTypeByName, rawValue, propKey) {
  if (derivedPropertyTypeByName[propKey].definition.type === "selection" && derivedPropertyTypeByName[propKey].definition.operation.type === "count") {
    const num = Number(rawValue);
    !Number.isSafeInteger(num) ? process.env.NODE_ENV !== "production" ? invariant2__default.default(false, "Count aggregation for derived property " + propKey + " returned a value larger than safe integer.") : invariant2__default.default(false) : void 0;
    return num;
  } else if (derivedPropertyTypeByName[propKey].selectedOrCollectedPropertyType != null && typeof derivedPropertyTypeByName[propKey].selectedOrCollectedPropertyType.type === "string" && specialPropertyTypes.has(derivedPropertyTypeByName[propKey].selectedOrCollectedPropertyType.type)) {
    switch (derivedPropertyTypeByName[propKey].selectedOrCollectedPropertyType?.type) {
      case "attachment":
        if (Array.isArray(rawValue)) {
          return rawValue.map((a) => chunkF4L4RFUT_cjs.hydrateAttachmentFromRidInternal(client, a.rid));
        } else {
          return chunkF4L4RFUT_cjs.hydrateAttachmentFromRidInternal(client, rawValue.rid);
        }
      default:
        process.env.NODE_ENV !== "production" ? invariant2__default.default(false, "Derived property aggregations for Timeseries and Media are not supported") : invariant2__default.default(false) ;
    }
  }
  return rawValue;
}
function createSpecialProperty(client, objectDef, rawObject, p) {
  const rawValue = rawObject[p];
  const propDef = objectDef.properties[p];
  if (process.env.NODE_ENV !== "production") {
    !(propDef != null && typeof propDef.type === "string" && specialPropertyTypes.has(propDef.type)) ? process.env.NODE_ENV !== "production" ? invariant2__default.default(false) : invariant2__default.default(false) : void 0;
  }
  if (propDef.type === "attachment") {
    if (Array.isArray(rawValue)) {
      return rawValue.map((a) => chunkF4L4RFUT_cjs.hydrateAttachmentFromRidInternal(client, a.rid));
    }
    return chunkF4L4RFUT_cjs.hydrateAttachmentFromRidInternal(client, rawValue.rid);
  }
  if (propDef.type === "numericTimeseries" || propDef.type === "stringTimeseries" || propDef.type === "sensorTimeseries") {
    return new TimeSeriesPropertyImpl(client, objectDef.apiName, rawObject[objectDef.primaryKeyApiName], p);
  }
  if (propDef.type === "geotimeSeriesReference") {
    return new GeotimeSeriesPropertyImpl(client, objectDef.apiName, rawObject[objectDef.primaryKeyApiName], p, rawValue.type === "geotimeSeriesValue" ? {
      time: rawValue.timestamp,
      value: {
        type: "Point",
        coordinates: rawValue.position
      }
    } : void 0);
  }
  if (propDef.type === "mediaReference") {
    return new MediaReferencePropertyImpl({
      client,
      objectApiName: objectDef.apiName,
      primaryKey: rawObject[objectDef.primaryKeyApiName],
      propertyName: p,
      mediaReference: rawValue
    });
  }
}
function parseWhenSecuritiesLoaded(wirePropertySecurities, rawObject, objectDef, derivedPropertyTypeByName = {}) {
  if (wirePropertySecurities == null || wirePropertySecurities.length === 0) {
    return {
      parsedObject: rawObject,
      clientPropertySecurities: void 0
    };
  }
  const parsedObject = rawObject;
  const clientPropertySecurities = {};
  for (const propKey of Object.keys(rawObject)) {
    if (propKey in objectDef.properties || propKey in derivedPropertyTypeByName || securableSpecialKeys.has(propKey)) {
      const value = rawObject[propKey];
      if (Array.isArray(value)) {
        const newVal = [];
        const newSecurities = [];
        value.forEach((spv) => {
          !(typeof spv === "object" && spv != null && "value" in spv && "propertySecurityIndex" in spv) ? process.env.NODE_ENV !== "production" ? invariant2__default.default(false, "Expected destructured secured property value object in array") : invariant2__default.default(false) : void 0;
          const securedValue = spv;
          newVal.push(securedValue.value);
          const securityIndex = securedValue.propertySecurityIndex;
          !(securityIndex != null) ? process.env.NODE_ENV !== "production" ? invariant2__default.default(false, "Expected property security index to be defined") : invariant2__default.default(false) : void 0;
          !(securityIndex < wirePropertySecurities.length) ? process.env.NODE_ENV !== "production" ? invariant2__default.default(false, "Expected property security index to be within bounds") : invariant2__default.default(false) : void 0;
          newSecurities.push(wirePropertySecurities[securityIndex].disjunction.map(wireToClientPropertySecurities));
        });
        parsedObject[propKey] = newVal;
        clientPropertySecurities[propKey] = newSecurities;
      } else if (typeof value === "object" && value != null && "value" in value && "propertySecurityIndex" in value) {
        const securedValue = value;
        parsedObject[propKey] = securedValue.value;
        const securityIndex = securedValue.propertySecurityIndex;
        !(securityIndex != null) ? process.env.NODE_ENV !== "production" ? invariant2__default.default(false, "Expected property security index to be defined") : invariant2__default.default(false) : void 0;
        !(securityIndex < wirePropertySecurities.length) ? process.env.NODE_ENV !== "production" ? invariant2__default.default(false, "Expected property security index to be within bounds") : invariant2__default.default(false) : void 0;
        clientPropertySecurities[propKey] = wirePropertySecurities[securityIndex].disjunction.map(wireToClientPropertySecurities);
      } else {
        parsedObject[propKey] = value;
      }
    }
  }
  return {
    parsedObject,
    clientPropertySecurities
  };
}
function wireToClientPropertySecurities(propertySecurity) {
  switch (propertySecurity.type) {
    case "propertyMarkingSummary":
      return {
        type: "propertyMarkings",
        conjunctive: propertySecurity.conjunctive,
        containerConjunctive: propertySecurity.containerConjunctive,
        disjunctive: propertySecurity.disjunctive,
        containerDisjunctive: propertySecurity.containerDisjunctive
      };
    case "errorComputingSecurity":
      return {
        type: "errorComputingSecurity"
      };
    case "unsupportedPolicy":
      return {
        type: "unsupportedPolicy"
      };
  }
}

// src/object/convertWireToOsdkObjects.ts
async function convertWireToOsdkObjects(client, objects, interfaceApiName, derivedPropertyTypeByName, propertySecurities, forceRemoveRid = false, selectedProps, strictNonNull = false, interfaceToObjectTypeMappings = {}, interfaceToObjectTypeMappingsV2 = {}) {
  fixObjectPropertiesInPlace(objects, forceRemoveRid);
  const effectiveMappings = Object.keys(interfaceToObjectTypeMappingsV2).length > 0 ? convertInterfaceToObjectTypeMappingsV2ToV1(interfaceToObjectTypeMappingsV2) : interfaceToObjectTypeMappings;
  const isInterfaceScoped = Object.keys(effectiveMappings).length > 0;
  const ret = [];
  for (const rawObj of objects) {
    const objectDef = await client.ontologyProvider.getObjectDefinition(rawObj.$apiName);
    !objectDef ? process.env.NODE_ENV !== "production" ? invariant2__default.default(false, `Missing definition for '${rawObj.$apiName}'`) : invariant2__default.default(false) : void 0;
    const interfaceToObjMapping = interfaceApiName && isInterfaceScoped ? effectiveMappings[interfaceApiName][rawObj.$apiName] : void 0;
    const ifaceSelected = interfaceApiName && interfaceToObjMapping ? selectedProps ? Object.keys(interfaceToObjMapping).filter((val) => {
      selectedProps?.includes(interfaceToObjMapping[val]);
    }) : [...Object.values(interfaceToObjMapping), objectDef.primaryKeyApiName] : void 0;
    let objProps;
    let conforming = true;
    if (interfaceApiName && ifaceSelected) {
      invariantInterfacesAsViews(objectDef, interfaceApiName, client);
      objProps = ifaceSelected;
    } else {
      objProps = selectedProps ?? Object.keys(objectDef.properties);
    }
    conforming &&= isConforming(client, objectDef, rawObj, objProps);
    if (strictNonNull === "throw" && !conforming) {
      throw new Error("Unable to safely convert objects as some non nullable properties are null");
    } else if (strictNonNull === "drop" && !conforming) {
      continue;
    }
    let osdkObject = createOsdkObject(client, objectDef, rawObj, derivedPropertyTypeByName, propertySecurities);
    if (interfaceApiName && isInterfaceScoped) osdkObject = osdkObject.$as(interfaceApiName);
    ret.push(osdkObject);
  }
  return ret;
}
function isConforming(client, def, obj, propsToCheck) {
  for (const propName of propsToCheck) {
    if (propName in def.properties && def.properties[propName].nullable === false && obj[propName] == null) {
      if (process.env.NODE_ENV !== "production") {
        client.logger?.debug({
          obj: {
            $apiName: obj.$apiName,
            $objectType: obj.$objectType,
            $primaryKey: obj.$primaryKey
          }
        }, `Found object that does not conform to its definition. Expected ${def.apiName}'s ${propName} to not be null.`);
      }
      return false;
    }
  }
  return true;
}
function invariantInterfacesAsViews(objectDef, interfaceApiName, client) {
  if (objectDef.interfaceMap?.[interfaceApiName] == null) {
    const warning = "Interfaces are only supported 'as views' but your metadata object is missing the correct information. This suggests your interfaces have not been migrated to the newer version yet and you cannot use this version of the SDK.";
    if (client.logger) {
      client.logger.warn(warning);
    } else {
      console.error(`WARNING! ${warning}`);
    }
    throw new Error(warning);
  }
}
function fixObjectPropertiesInPlace(objs, forceRemoveRid) {
  for (const obj of objs) {
    if (forceRemoveRid) {
      delete obj.__rid;
    }
    if (obj.__rid) {
      obj.$rid = obj.__rid;
      delete obj.__rid;
    }
    if (obj.__score) {
      obj.$score = obj.__score;
      delete obj.__score;
    }
    obj.$apiName ??= obj.__apiName;
    obj.$objectType = obj.$apiName;
    obj.$primaryKey ??= obj.__primaryKey;
    obj.$title ??= obj.__title;
    delete obj.__apiName;
    delete obj.__primaryKey;
    delete obj.__title;
  }
}
function convertInterfaceToObjectTypeMappingsV2ToV1(mappingsV2) {
  return Object.fromEntries(Object.entries(mappingsV2).map(([interfaceApiName, objectTypeMappingsV2]) => [interfaceApiName, Object.fromEntries(Object.entries(objectTypeMappingsV2).map(([objectTypeName, propertyMappings]) => [objectTypeName, Object.fromEntries(Object.entries(propertyMappings).filter(([, impl]) => impl.type === "localPropertyImplementation").map(([interfaceProp, impl]) => [interfaceProp, impl.propertyApiName]))]))]));
}

// src/object/Cache.ts
function createClientCache(fn) {
  const cache = /* @__PURE__ */ new WeakMap();
  function get4(client, key) {
    if (cache.get(client.clientCacheKey) == null) {
      cache.set(client.clientCacheKey, /* @__PURE__ */ new Map());
    }
    const r = cache.get(client.clientCacheKey).get(key);
    if (r === void 0 && fn !== void 0) {
      return set(client, key, fn(client, key));
    } else {
      return r;
    }
  }
  function set(client, key, value) {
    if (cache.get(client.clientCacheKey) == null) {
      cache.set(client.clientCacheKey, /* @__PURE__ */ new Map());
    }
    cache.get(client.clientCacheKey).set(key, value);
    return value;
  }
  function remove(client, key) {
    if (cache.get(client.clientCacheKey) == null) return false;
    return cache.get(client.clientCacheKey).delete(key);
  }
  return {
    get: get4,
    set,
    remove
  };
}
function createAsyncClientCache(fn, createCacheLocal = createClientCache) {
  const cache = createCacheLocal();
  const inProgress = createCacheLocal();
  const ret = {
    getOrUndefined: function getOrUndefined(client, key) {
      return cache.get(client, key);
    },
    get: async function get4(client, key) {
      return cache.get(client, key) ?? inProgress.get(client, key) ?? ret.set(client, key, fn(client, key));
    },
    set: async function set(client, k, v) {
      try {
        const r = await inProgress.set(client, k, v);
        cache.set(client, k, r);
        inProgress.remove(client, k);
        return r;
      } catch (e) {
        inProgress.remove(client, k);
        throw e;
      }
    }
  };
  return ret;
}

// src/util/deepFreeze.ts
function deepFreeze(obj) {
  Object.getOwnPropertyNames(obj).forEach((name) => {
    const prop = obj[name];
    if (typeof prop === "object" && prop != null && !Object.isFrozen(prop)) {
      deepFreeze(prop);
    }
  });
  return Object.freeze(obj);
}

// src/ontology/loadActionMetadata.ts
async function loadActionMetadata(client, actionType) {
  const r = await get(client, await client.ontologyRid, actionType, {
    branch: client.branch
  });
  const {
    wireActionTypeV2ToSdkActionMetadata
  } = await import('@osdk/generator-converters');
  return wireActionTypeV2ToSdkActionMetadata(r);
}

// src/ontology/loadFullObjectMetadata.ts
async function loadFullObjectMetadata(client, objectType) {
  const full = await getFullMetadata(client, await client.ontologyRid, objectType, {
    preview: true,
    branch: client.branch
  });
  const {
    wireObjectTypeFullMetadataToSdkObjectMetadata
  } = await import('@osdk/generator-converters');
  const ret = wireObjectTypeFullMetadataToSdkObjectMetadata(full, true);
  return {
    ...ret
  };
}

// src/ontology/loadInterfaceMetadata.ts
async function loadInterfaceMetadata(client, objectType) {
  const r = await get2(client, await client.ontologyRid, objectType, {
    preview: true,
    branch: client.branch
  });
  const {
    wireInterfaceTypeV2ToSdkObjectDefinition
  } = await import('@osdk/generator-converters');
  return wireInterfaceTypeV2ToSdkObjectDefinition(r, true);
}

// src/ontology/loadQueryMetadata.ts
async function loadQueryMetadata(client, queryTypeApiNameAndVersion) {
  const [apiName, version] = queryTypeApiNameAndVersion.split(":");
  const r = await get3(client, await client.ontologyRid, apiName, {
    version
  });
  const {
    wireQueryTypeV2ToSdkQueryMetadata
  } = await import('@osdk/generator-converters');
  return wireQueryTypeV2ToSdkQueryMetadata(r);
}

// src/ontology/StandardOntologyProvider.ts
var createStandardOntologyProviderFactory = (client) => {
  return (client2) => {
    async function loadObject(client3, key) {
      const objectDef = await loadFullObjectMetadata(client3, key);
      const interfaceDefs = Object.fromEntries((await Promise.all(objectDef.implements?.map((i) => ret.getInterfaceDefinition(i)) ?? [])).map((i) => [i.apiName, {
        def: i,
        handler: void 0
      }]));
      const fullObjectDef = {
        ...objectDef,
        [InterfaceDefinitions]: interfaceDefs
      };
      return deepFreeze(fullObjectDef);
    }
    async function loadInterface(client3, key) {
      return deepFreeze(await loadInterfaceMetadata(client3, key));
    }
    async function loadQuery(client3, key) {
      return loadQueryMetadata(client3, key);
    }
    async function loadAction(client3, key) {
      const r = await loadActionMetadata(client3, key);
      return r;
    }
    function makeGetter(fn) {
      const cache = createAsyncClientCache((client3, key) => fn(client3, key, false));
      return async (apiName) => {
        return await cache.get(client2, apiName);
      };
    }
    function makeQueryGetter(client3, fn) {
      const queryCache = createAsyncClientCache((client4, key) => {
        return fn(client4, key);
      });
      return async (apiName, version) => {
        const key = version ? `${apiName}:${version}` : apiName;
        return await queryCache.get(client3, key);
      };
    }
    const ret = {
      getObjectDefinition: makeGetter(loadObject),
      getInterfaceDefinition: makeGetter(loadInterface),
      getActionDefinition: makeGetter(loadAction),
      getQueryDefinition: makeQueryGetter(client2, loadQuery)
    };
    return ret;
  };
};

// src/util/UserAgent.ts
var USER_AGENT = `osdk-client/${"2.40.0"}`;
var OBSERVABLE_USER_AGENT = `osdk-observable-client/${"2.40.0"}`;

// src/createMinimalClient.ts
function createMinimalClient(metadata, baseUrl, tokenProvider, options = {}, fetchFn = global.fetch, objectSetFactory = chunkF4L4RFUT_cjs.createObjectSet, createOntologyProviderFactory = createStandardOntologyProviderFactory) {
  if (process.env.NODE_ENV !== "production") {
    try {
      new URL(baseUrl);
    } catch (e) {
      const hint = !baseUrl.startsWith("http://") || !baseUrl.startsWith("https://") ? ". Did you forget to add 'http://' or 'https://'?" : "";
      throw new Error(`Invalid stack URL: ${baseUrl}${hint}`);
    }
  }
  const minimalClient = {
    ...shared_client_impl.createSharedClientContext(baseUrl, tokenProvider, USER_AGENT, fetchFn, options.headers),
    objectSetFactory,
    objectFactory: convertWireToOsdkObjects,
    ontologyRid: metadata.ontologyRid,
    logger: options.logger,
    transactionId: options.transactionId,
    flushEdits: options.flushEdits,
    scenarioRid: options.scenarioRid,
    clientCacheKey: {},
    requestContext: {},
    branch: options.branch,
    narrowTypeInterfaceOrObjectMapping: {}
  };
  return Object.freeze(Object.assign(minimalClient, {
    ontologyProvider: createOntologyProviderFactory(options)(minimalClient),
    subscribeFn: options.subscribeFn ?? ((objectType, objectSet, listener, properties, shouldLoadRids) => import('./ObjectSetListenerWebsocket-JCUV33MX.cjs').then(({
      ObjectSetListenerWebsocket: ObjectSetListenerWebsocket2
    }) => ObjectSetListenerWebsocket2.getInstance(minimalClient).subscribe(objectType, objectSet, listener, properties, shouldLoadRids)))
  }));
}

// src/fetchMetadata.ts
var fetchMetadataInternal = async (client, definition) => {
  if (definition.type === "object") {
    const {
      [InterfaceDefinitions]: interfaceDefs,
      ...objectTypeDef
    } = await client.ontologyProvider.getObjectDefinition(definition.apiName);
    return objectTypeDef;
  } else if (definition.type === "interface") {
    return client.ontologyProvider.getInterfaceDefinition(definition.apiName);
  } else if (definition.type === "action") {
    return client.ontologyProvider.getActionDefinition(definition.unsanitizedApiName ?? definition.apiName);
  } else if (definition.type === "query") {
    return client.ontologyProvider.getQueryDefinition(definition.apiName, definition.isFixedVersion ? definition.version : void 0);
  } else {
    throw new Error("Not implemented for given definition");
  }
};

// src/internal/conversions/makeMediaTransformation.ts
function makeMediaTransformation(transformation) {
  if ("$image" in transformation && transformation.$image != null) {
    const t = transformation.$image;
    return {
      type: "image",
      encoding: {
        type: t.$encoding
      },
      operations: t.$operations.map(convertImageOperation)
    };
  } else if ("$video" in transformation && transformation.$video != null) {
    const t = transformation.$video;
    return {
      type: "video",
      encoding: {
        type: t.$encoding
      },
      operation: convertVideoOperation(t.$operation)
    };
  } else if ("$audio" in transformation && transformation.$audio != null) {
    return {
      type: "audio",
      operation: convertAudioOperation(transformation.$audio.$operation)
    };
  } else if ("$emailToText" in transformation && transformation.$emailToText != null) {
    return {
      type: "emailToText",
      operation: convertEmailToTextOperation(transformation.$emailToText.$operation)
    };
  } else if ("$spreadsheetToText" in transformation && transformation.$spreadsheetToText != null) {
    return {
      type: "spreadsheetToText",
      operation: convertSpreadsheetToTextOperation(transformation.$spreadsheetToText.$operation)
    };
  } else if ("$videoToAudio" in transformation && transformation.$videoToAudio != null) {
    const t = transformation.$videoToAudio;
    return {
      type: "videoToAudio",
      encoding: convertAudioEncoding(t.$encoding),
      operation: convertVideoToAudioOperation(t.$operation)
    };
  } else if ("$audioToText" in transformation && transformation.$audioToText != null) {
    return {
      type: "audioToText",
      operation: convertAudioToTextOperation(transformation.$audioToText.$operation)
    };
  } else if ("$emailToAttachment" in transformation && transformation.$emailToAttachment != null) {
    return {
      type: "emailToAttachment",
      operation: convertEmailToAttachmentOperation(transformation.$emailToAttachment.$operation)
    };
  } else if ("$videoToArchive" in transformation && transformation.$videoToArchive != null) {
    const t = transformation.$videoToArchive;
    return {
      type: "videoToArchive",
      encoding: {
        type: t.$encoding
      },
      operation: convertVideoToArchiveOperation(t.$operation)
    };
  } else if ("$videoToText" in transformation && transformation.$videoToText != null) {
    return {
      type: "videoToText",
      operation: convertVideoToTextOperation(transformation.$videoToText.$operation)
    };
  } else if ("$imageToText" in transformation && transformation.$imageToText != null) {
    return {
      type: "imageToText",
      operation: convertImageToTextOperation(transformation.$imageToText.$operation)
    };
  } else if ("$videoToImage" in transformation && transformation.$videoToImage != null) {
    const t = transformation.$videoToImage;
    return {
      type: "videoToImage",
      encoding: {
        type: t.$encoding
      },
      operation: convertVideoToImageOperation(t.$operation)
    };
  } else if ("$imageToDocument" in transformation && transformation.$imageToDocument != null) {
    return {
      type: "imageToDocument",
      operation: convertImageToDocumentOperation(transformation.$imageToDocument.$operation)
    };
  } else if ("$dicomToImage" in transformation && transformation.$dicomToImage != null) {
    const t = transformation.$dicomToImage;
    return {
      type: "dicomToImage",
      encoding: {
        type: t.$encoding
      },
      operation: convertDicomToImageOperation(t.$operation)
    };
  } else if ("$documentToDocument" in transformation && transformation.$documentToDocument != null) {
    const t = transformation.$documentToDocument;
    return {
      type: "documentToDocument",
      encoding: {
        type: t.$encoding
      },
      operation: convertDocumentToDocumentOperation(t.$operation)
    };
  } else if ("$documentToImage" in transformation && transformation.$documentToImage != null) {
    const t = transformation.$documentToImage;
    return {
      type: "documentToImage",
      encoding: {
        type: t.$encoding
      },
      operation: convertDocumentToImageOperation(t.$operation)
    };
  } else if ("$imageToEmbedding" in transformation && transformation.$imageToEmbedding != null) {
    return {
      type: "imageToEmbedding",
      operation: convertImageToEmbeddingOperation(transformation.$imageToEmbedding.$operation)
    };
  } else {
    return {
      type: "documentToText",
      operation: convertDocumentToTextOperation(transformation.$documentToText.$operation)
    };
  }
}
function convertAudioEncoding(encoding) {
  if (typeof encoding === "string") {
    if (encoding === "mp3") {
      return {
        type: "mp3"
      };
    } else {
      return {
        type: "ts"
      };
    }
  }
  return {
    type: "wav",
    sampleRate: encoding.$wav.$sampleRate,
    audioChannelLayout: encoding.$wav.$audioChannelLayout != null ? {
      type: "numberOfChannels",
      numberOfChannels: encoding.$wav.$audioChannelLayout.$numberOfChannels
    } : void 0
  };
}
function convertPageRange(range) {
  return {
    startPageInclusive: range.$startPageInclusive,
    endPageExclusive: range.$endPageExclusive
  };
}
function convertOcrOutputFormat(fmt) {
  if ("$hocr" in fmt) {
    return {
      type: "hocr"
    };
  } else {
    return {
      type: "text"
    };
  }
}
function convertOcrLanguageOrScript(item) {
  if ("$language" in item && item.$language != null) {
    return {
      type: "language",
      language: item.$language
    };
  } else {
    const script = item.$script;
    return {
      type: "script",
      script
    };
  }
}
function convertOcrParameters(params) {
  return {
    outputFormat: convertOcrOutputFormat(params.$outputFormat),
    languages: params.$languages.map(convertOcrLanguageOrScript)
  };
}
function convertLayoutAwareExtractionParameters(params) {
  return {
    languages: params.$languages
  };
}
function convertDocumentTextExtractionConfig(config) {
  return {
    format: config.$format,
    mode: config.$mode,
    languages: config.$languages.map(convertOcrLanguageOrScript)
  };
}
function convertTranscribeOutputFormat(fmt) {
  if ("$plainTextNoSegmentData" in fmt && fmt.$plainTextNoSegmentData != null) {
    return {
      type: "plainTextNoSegmentData",
      addTimestamps: fmt.$plainTextNoSegmentData.$addTimestamps
    };
  } else if ("$json" in fmt) {
    return {
      type: "json"
    };
  } else {
    return {
      type: "pttml"
    };
  }
}
function convertLlmSpec(spec) {
  return {
    type: "chat",
    chat: {
      modelLocator: {
        type: "apiName",
        apiName: spec.$chat.$modelLocator
      },
      systemPrompt: spec.$chat.$systemPrompt,
      userPrompt: spec.$chat.$userPrompt,
      maxTokens: spec.$chat.$maxTokens
    }
  };
}
function convertVlmPreprocessingConfig(config) {
  if ("$layoutAware" in config && config.$layoutAware != null) {
    return {
      type: "layoutAware",
      layoutAware: {
        transformationConfig: convertDocumentTextExtractionConfig(config.$layoutAware.$transformationConfig),
        cropConfig: config.$layoutAware.$cropConfig != null ? {
          tablePrompt: config.$layoutAware.$cropConfig.$tablePrompt
        } : void 0
      }
    };
  } else {
    const ext = config.$extractText;
    return {
      type: "extractText",
      extractText: convertDocumentTextExtractionConfig(ext)
    };
  }
}
function convertImageSpec(spec) {
  return {
    resizingMode: spec.$resizingMode,
    height: spec.$height,
    width: spec.$width,
    mimeType: spec.$mimeType
  };
}
function convertImageOperation(op) {
  if ("$resize" in op && op.$resize != null) {
    return {
      type: "resize",
      width: op.$resize.$width,
      height: op.$resize.$height,
      autoOrient: op.$resize.$autoOrient
    };
  } else if ("$resizeToFitBoundingBox" in op && op.$resizeToFitBoundingBox != null) {
    return {
      type: "resizeToFitBoundingBox",
      width: op.$resizeToFitBoundingBox.$width,
      height: op.$resizeToFitBoundingBox.$height
    };
  } else if ("$rotate" in op && op.$rotate != null) {
    return {
      type: "rotate",
      angle: op.$rotate.$angle
    };
  } else if ("$crop" in op && op.$crop != null) {
    return {
      type: "crop",
      xOffset: op.$crop.$xOffset,
      yOffset: op.$crop.$yOffset,
      width: op.$crop.$width,
      height: op.$crop.$height
    };
  } else if ("$grayscale" in op) {
    return {
      type: "grayscale"
    };
  } else {
    const tile = op.$tile;
    return {
      type: "tile",
      zoom: tile.$zoom,
      x: tile.$x,
      y: tile.$y
    };
  }
}
function convertVideoOperation(op) {
  if ("$transcode" in op) {
    return {
      type: "transcode"
    };
  } else {
    const chunk = op.$chunk;
    return {
      type: "chunk",
      chunkDurationMilliseconds: chunk.$chunkDurationMilliseconds,
      chunkIndex: chunk.$chunkIndex
    };
  }
}
function convertAudioOperation(op) {
  if ("$channel" in op && op.$channel != null) {
    return {
      type: "channel",
      encodeFormat: convertAudioEncoding(op.$channel.$encodeFormat),
      channel: op.$channel.$channel
    };
  } else if ("$chunk" in op && op.$chunk != null) {
    return {
      type: "chunk",
      encodeFormat: convertAudioEncoding(op.$chunk.$encodeFormat),
      chunkDurationMilliseconds: op.$chunk.$chunkDurationMilliseconds,
      chunkIndex: op.$chunk.$chunkIndex
    };
  } else {
    const convert = op.$convert;
    return {
      type: "convert",
      encodeFormat: convertAudioEncoding(convert.$encodeFormat)
    };
  }
}
function convertVideoToAudioOperation(_op) {
  return {
    type: "extractAudio"
  };
}
function convertAudioToTextOperation(op) {
  if ("$transcribe" in op && op.$transcribe != null) {
    return {
      type: "transcribe",
      language: op.$transcribe.$language,
      diarize: op.$transcribe.$diarize,
      outputFormat: op.$transcribe.$outputFormat != null ? convertTranscribeOutputFormat(op.$transcribe.$outputFormat) : void 0,
      performanceMode: op.$transcribe.$performanceMode
    };
  } else {
    const waveform = op.$waveform;
    return {
      type: "waveform",
      peaksPerSecond: waveform.$peaksPerSecond
    };
  }
}
function convertEmailToTextOperation(op) {
  return {
    type: "getEmailBody",
    outputFormat: op.$getEmailBody.$outputFormat
  };
}
function convertEmailToAttachmentOperation(op) {
  return {
    type: "getEmailAttachment",
    mimeType: op.$getEmailAttachment.$mimeType,
    attachmentIndex: op.$getEmailAttachment.$attachmentIndex
  };
}
function convertSpreadsheetToTextOperation(op) {
  return {
    type: "convertSheetToJson",
    sheetName: op.$convertSheetToJson.$sheetName
  };
}
function convertVideoToArchiveOperation(op) {
  return {
    type: "extractSceneFrames",
    encoding: {
      type: op.$extractSceneFrames.$encoding
    },
    sceneScore: op.$extractSceneFrames.$sceneScore
  };
}
function convertVideoToTextOperation(op) {
  return {
    type: "getTimestampsForSceneFrames",
    sceneScore: op.$getTimestampsForSceneFrames.$sceneScore
  };
}
function convertImageToTextOperation(op) {
  if ("$extractLayoutAwareContent" in op && op.$extractLayoutAwareContent != null) {
    return {
      type: "extractLayoutAwareContent",
      parameters: convertLayoutAwareExtractionParameters(op.$extractLayoutAwareContent.$parameters)
    };
  } else {
    const ocr = op.$ocr;
    return {
      type: "ocr",
      parameters: convertOcrParameters(ocr.$parameters)
    };
  }
}
function convertVideoToImageOperation(op) {
  if ("$extractFirstFrame" in op && op.$extractFirstFrame != null) {
    return {
      type: "extractFirstFrame",
      height: op.$extractFirstFrame.$height,
      width: op.$extractFirstFrame.$width
    };
  } else {
    const ext = op.$extractFramesAtTimestamps;
    return {
      type: "extractFramesAtTimestamps",
      height: ext.$height,
      width: ext.$width,
      timestamp: ext.$timestamp
    };
  }
}
function convertImageToDocumentOperation(_op) {
  return {
    type: "createPdf"
  };
}
function convertDicomToImageOperation(op) {
  return {
    type: "renderImageLayer",
    layerNumber: op.$renderImageLayer.$layerNumber,
    height: op.$renderImageLayer.$height,
    width: op.$renderImageLayer.$width
  };
}
function convertDocumentToDocumentOperation(op) {
  if ("$slicePdfRange" in op && op.$slicePdfRange != null) {
    return {
      type: "slicePdfRange",
      startPageInclusive: op.$slicePdfRange.$startPageInclusive,
      endPageExclusive: op.$slicePdfRange.$endPageExclusive,
      strictlyEnforceEndPage: op.$slicePdfRange.$strictlyEnforceEndPage
    };
  } else {
    return {
      type: "convertDocument"
    };
  }
}
function convertDocumentToImageOperation(op) {
  if ("$renderPage" in op && op.$renderPage != null) {
    return {
      type: "renderPage",
      pageNumber: op.$renderPage.$pageNumber,
      height: op.$renderPage.$height,
      width: op.$renderPage.$width
    };
  } else {
    const bb = op.$renderPageToFitBoundingBox;
    return {
      type: "renderPageToFitBoundingBox",
      pageNumber: bb.$pageNumber,
      width: bb.$width,
      height: bb.$height
    };
  }
}
function convertImageToEmbeddingOperation(op) {
  return {
    type: "generateEmbedding",
    modelId: op.$generateEmbedding.$modelId
  };
}
function convertDocumentToTextOperation(op) {
  if ("$extractAllText" in op) {
    return {
      type: "extractAllText"
    };
  } else if ("$extractTableOfContents" in op) {
    return {
      type: "extractTableOfContents"
    };
  } else if ("$getPdfPageDimensions" in op) {
    return {
      type: "getPdfPageDimensions"
    };
  } else if ("$extractFormFields" in op) {
    return {
      type: "extractFormFields"
    };
  } else if ("$extractUnstructuredTextFromPage" in op && op.$extractUnstructuredTextFromPage != null) {
    return {
      type: "extractUnstructuredTextFromPage",
      pageNumber: op.$extractUnstructuredTextFromPage.$pageNumber
    };
  } else if ("$extractTextFromPagesToArray" in op && op.$extractTextFromPagesToArray != null) {
    return {
      type: "extractTextFromPagesToArray",
      startPage: op.$extractTextFromPagesToArray.$startPage,
      endPage: op.$extractTextFromPagesToArray.$endPage
    };
  } else if ("$ocrOnPage" in op && op.$ocrOnPage != null) {
    return {
      type: "ocrOnPage",
      pageNumber: op.$ocrOnPage.$pageNumber,
      parameters: convertOcrParameters(op.$ocrOnPage.$parameters)
    };
  } else if ("$ocrOnPages" in op && op.$ocrOnPages != null) {
    return {
      type: "ocrOnPages",
      pageNumber: op.$ocrOnPages.$pageNumber,
      parameters: convertOcrParameters(op.$ocrOnPages.$parameters)
    };
  } else if ("$extractLayoutAwareContent" in op && op.$extractLayoutAwareContent != null) {
    return {
      type: "extractLayoutAwareContent",
      parameters: convertLayoutAwareExtractionParameters(op.$extractLayoutAwareContent.$parameters)
    };
  } else if ("$extractLayoutAwareTextV2" in op && op.$extractLayoutAwareTextV2 != null) {
    return {
      type: "extractLayoutAwareTextV2",
      pageRange: op.$extractLayoutAwareTextV2.$pageRange != null ? convertPageRange(op.$extractLayoutAwareTextV2.$pageRange) : void 0,
      config: convertDocumentTextExtractionConfig(op.$extractLayoutAwareTextV2.$config)
    };
  } else if ("$extractTextV2" in op && op.$extractTextV2 != null) {
    return {
      type: "extractTextV2",
      pageRange: op.$extractTextV2.$pageRange != null ? convertPageRange(op.$extractTextV2.$pageRange) : void 0,
      config: convertDocumentTextExtractionConfig(op.$extractTextV2.$config)
    };
  } else {
    const vlm = op.$extractVlmText;
    return {
      type: "extractVlmText",
      llmSpec: convertLlmSpec(vlm.$llmSpec),
      preprocessingConfiguration: vlm.$preprocessingConfiguration != null ? convertVlmPreprocessingConfig(vlm.$preprocessingConfiguration) : void 0,
      imageSpec: vlm.$imageSpec != null ? convertImageSpec(vlm.$imageSpec) : void 0,
      outputFormat: vlm.$outputFormat,
      pageRange: vlm.$pageRange != null ? convertPageRange(vlm.$pageRange) : void 0
    };
  }
}

// src/createClient.ts
var ActionInvoker = class {
  constructor(clientCtx, actionDef) {
    this.applyAction = applyAction.bind(void 0, clientCtx, actionDef);
    this.batchApplyAction = applyAction.bind(void 0, clientCtx, actionDef);
  }
};
var QueryInvoker = class {
  constructor(clientCtx, queryDef) {
    this.executeFunction = chunkJDRCSS7F_cjs.applyQuery.bind(void 0, clientCtx, queryDef);
  }
};
function createClientInternal(objectSetFactory, transactionRid, flushEdits, scenarioRid, subscribeFn, baseUrl, ontologyRid, tokenProvider, options = void 0, fetchFn = fetch) {
  if (typeof ontologyRid === "string") {
    if (!ontologyRid.startsWith("ri.")) {
      throw new Error("Invalid ontology RID");
    }
  } else {
    ontologyRid.then((ontologyRid2) => {
      if (!ontologyRid2.startsWith("ri.")) {
        throw new Error("Invalid ontology RID");
      }
    });
  }
  const clientCtx = createMinimalClient({
    ontologyRid
  }, baseUrl, tokenProvider, {
    ...options,
    logger: options?.logger ?? new chunkTQUNF2DZ_cjs.MinimalLogger(),
    transactionId: transactionRid,
    flushEdits,
    scenarioRid,
    branch: options?.UNSTABLE_DO_NOT_USE_BRANCH,
    subscribeFn
  }, fetchFn, objectSetFactory);
  return createClientFromContext(clientCtx);
}
function createClientFromContext(clientCtx) {
  function clientFn(o) {
    if (o.type === "object" || o.type === "interface") {
      return clientCtx.objectSetFactory(o, clientCtx);
    } else if (o.type === "action") {
      return new ActionInvoker(clientCtx, o);
    } else if (o.type === "query") {
      return new QueryInvoker(clientCtx, o);
    } else if (o.type === "experiment") {
      switch (o.name) {
        case unstable.__EXPERIMENTAL__NOT_SUPPORTED_YET__executeStreamingFunction.name:
          return {
            async *executeStreamingFunction(query, params) {
              const {
                applyStreamingQuery
              } = await import('./applyStreamingQuery-6ZZRUAB2.cjs');
              yield* applyStreamingQuery(clientCtx, query, params);
            }
          };
        case unstable.__EXPERIMENTAL__NOT_SUPPORTED_YET__getBulkLinks.name:
          return {
            async *getBulkLinks(objs, linkTypes) {
              const {
                createBulkLinksAsyncIterFactory
              } = await import('./createBulkLinksAsyncIterFactory-F4NZOXU7.cjs');
              yield* createBulkLinksAsyncIterFactory(clientCtx)(objs, linkTypes);
            }
          };
        case unstable.__EXPERIMENTAL__NOT_SUPPORTED_YET__fetchOneByRid.name:
          return {
            fetchOneByRid: async (objectType, rid, options) => {
              return await chunkF4L4RFUT_cjs.fetchSingle(clientCtx, objectType, options, createWithRid([rid]));
            }
          };
        case unstable.__EXPERIMENTAL__NOT_SUPPORTED_YET__createMediaReference.name:
          return {
            createMediaReference: async (args) => {
              const {
                data,
                fileName,
                objectType,
                propertyType
              } = args;
              const {
                upload: upload2
              } = await import('./MediaReferenceProperty-QB5RSPON.cjs');
              return await upload2(clientCtx, await clientCtx.ontologyRid, objectType.apiName, propertyType, data, {
                mediaItemPath: fileName,
                preview: true
              });
            }
          };
        case unstable.__EXPERIMENTAL__NOT_SUPPORTED_YET__fetchPageByRid.name:
          return {
            fetchPageByRid: async (objectOrInterfaceType, rids, options = {}) => {
              return await chunkF4L4RFUT_cjs.fetchPage(clientCtx, objectOrInterfaceType, options, createWithRid(rids));
            },
            fetchPageByRidNoType: async (rids, options) => {
              return await chunkF4L4RFUT_cjs.fetchStaticRidPage(clientCtx, rids, options ?? {});
            }
          };
        case unstable.__EXPERIMENTAL__NOT_SUPPORTED_YET__subscribeToNoTypeObjectSet.name:
          return {
            subscribeToNoTypeObjectSet: (rid, listener, opts) => {
              const unsubscribe = chunkCEBSLQQC_cjs.ObjectSetListenerWebsocket.getInstance(clientCtx).subscribeWithoutType({
                type: "reference",
                reference: rid
              }, listener, opts?.includeRid ?? false);
              return {
                unsubscribe
              };
            }
          };
        case unstable.__EXPERIMENTAL__NOT_SUPPORTED_YET__transformAndWait.name:
          return {
            transformAndWait: async (args) => {
              const {
                transformAndWaitInternal
              } = await import('./transformAndWaitInternal-PTYSXCC2.cjs');
              const {
                mediaSetRid,
                mediaItemRid,
                token
              } = args.mediaReference.reference.mediaSetViewItem;
              return transformAndWaitInternal(clientCtx, mediaSetRid, mediaItemRid, makeMediaTransformation(args.transformation), token, args.options);
            }
          };
      }
      throw new Error("not implemented");
    } else {
      throw new Error("not implemented");
    }
  }
  const fetchMetadata = fetchMetadataInternal.bind(void 0, clientCtx);
  const symbolClientContext2 = "__osdkClientContext";
  const client = Object.defineProperties(clientFn, {
    [chunkH6PXPU6F_cjs.symbolClientContext]: {
      value: clientCtx
    },
    [symbolClientContext2]: {
      value: clientCtx
    },
    [chunkF4L4RFUT_cjs.additionalContext]: {
      value: clientCtx
    },
    fetchMetadata: {
      value: fetchMetadata
    }
  });
  return client;
}
var createClient = createClientInternal.bind(void 0, chunkF4L4RFUT_cjs.createObjectSet, void 0, void 0, void 0, void 0);
var createClientWithTransaction = (transactionRid, flushEdits, ...args) => createClientInternal(chunkF4L4RFUT_cjs.createObjectSet, transactionRid, flushEdits, void 0, void 0, ...args);
var createClientWithSubscribe = (subscribeFn, ...args) => createClientInternal(chunkF4L4RFUT_cjs.createObjectSet, void 0, void 0, void 0, subscribeFn, ...args);
var createClientWithScenario = (scenarioRid, ...args) => createClientInternal(chunkF4L4RFUT_cjs.createObjectSet, void 0, void 0, scenarioRid, void 0, ...args);
function createWithRid(rids) {
  const withRid = {
    type: "static",
    "objects": rids
  };
  return withRid;
}

exports.ActionValidationError = ActionValidationError;
exports.ClientRef = ClientRef;
exports.InterfaceDefRef = InterfaceDefRef;
exports.OBSERVABLE_USER_AGENT = OBSERVABLE_USER_AGENT;
exports.ObjectDefRef = ObjectDefRef;
exports.OntologyScenario_exports = OntologyScenario_exports;
exports.USER_AGENT = USER_AGENT;
exports.UnderlyingOsdkObject = UnderlyingOsdkObject;
exports.buildScenarioClient = buildScenarioClient;
exports.createClient = createClient;
exports.createClientFromContext = createClientFromContext;
exports.createClientWithSubscribe = createClientWithSubscribe;
exports.createClientWithTransaction = createClientWithTransaction;
exports.createOsdkObject = createOsdkObject;
