// @osdk/client/build/cjs/chunk-F4L4RFUT.cjs
'use strict';

var chunkROJIXXGF_cjs = require('./chunk-ROJIXXGF.cjs');
var chunkH6PXPU6F_cjs = require('./chunk-H6PXPU6F.cjs');
var shared_client_impl = require('@osdk/shared.client.impl');
var shared_net_fetch = require('@osdk/shared.net.fetch');
var invariant5 = require('tiny-invariant');
var api = require('@osdk/api');
var shared_net_errors = require('@osdk/shared.net.errors');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var invariant5__default = /*#__PURE__*/_interopDefault(invariant5);

var addUserAgentAndRequestContextHeaders = (client, withMetadata) => ({
  ...client,
  fetch: shared_net_fetch.createFetchHeaderMutator(client.fetch, (headers) => {
    headers.set("X-OSDK-Request-Context", JSON.stringify(client.requestContext));
    if (withMetadata.osdkMetadata) {
      headers.set(shared_client_impl.USER_AGENT_HEADER, [headers.get(shared_client_impl.USER_AGENT_HEADER), withMetadata.osdkMetadata.extraUserAgent].filter((x) => x && x?.length > 0).join(" "));
    }
    return headers;
  })
});

// src/util/augmentRequestContext.ts
var augmentRequestContext = (client, augment2) => ({
  ...client,
  requestContext: {
    ...client.requestContext,
    ...augment2(client.requestContext)
  }
});

// ../../node_modules/.pnpm/@osdk+foundry.ontologies@2.63.0/node_modules/@osdk/foundry.ontologies/build/esm/public/OntologyObjectSet.js
var _createTemporary = [1, "/v2/ontologies/{0}/objectSets/createTemporary", 7];
function createTemporary($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _createTemporary, ...args);
}
var _load = [1, "/v2/ontologies/{0}/objectSets/loadObjects", 7];
function load($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _load, ...args);
}
var _loadMultipleObjectTypes = [1, "/v2/ontologies/{0}/objectSets/loadObjectsMultipleObjectTypes", 7];
function loadMultipleObjectTypes($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _loadMultipleObjectTypes, ...args);
}
var _aggregate = [1, "/v2/ontologies/{0}/objectSets/aggregate", 7];
function aggregate($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _aggregate, ...args);
}
var _loadLinks = [1, "/v2/ontologies/{0}/objectSets/loadLinks", 7];
function loadLinks($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _loadLinks, ...args);
}

// src/internal/conversions/extractNamespace.ts
function extractNamespace(fqApiName) {
  const last = fqApiName.lastIndexOf(".");
  if (last === -1) return [void 0, fqApiName];
  return [fqApiName.slice(0, last), fqApiName.slice(last + 1)];
}
async function extractObjectOrInterfaceType(clientCtx, objectSet) {
  switch (objectSet.type) {
    case "searchAround": {
      const def2 = await extractObjectOrInterfaceType(clientCtx, objectSet.objectSet);
      if (def2 === void 0) {
        return void 0;
      }
      const objOrInterfaceDef2 = def2.type === "object" ? await clientCtx.ontologyProvider.getObjectDefinition(def2.apiName) : await clientCtx.ontologyProvider.getInterfaceDefinition(def2.apiName);
      const linkDef2 = objOrInterfaceDef2.links[objectSet.link];
      !linkDef2 ? process.env.NODE_ENV !== "production" ? invariant5__default.default(false, `Missing link definition for '${objectSet.link}'`) : invariant5__default.default(false) : void 0;
      return objOrInterfaceDef2.type === "object" ? {
        apiName: objOrInterfaceDef2.links[objectSet.link].targetType,
        type: "object"
      } : {
        apiName: objOrInterfaceDef2.links[objectSet.link].targetTypeApiName,
        type: objOrInterfaceDef2.links[objectSet.link].targetType
      };
    }
    case "withProperties": {
      return extractObjectOrInterfaceType(clientCtx, objectSet.objectSet);
    }
    case "methodInput":
      return void 0;
    case "base":
      return {
        type: "object",
        apiName: objectSet.objectType
      };
    case "interfaceBase":
      return {
        type: "interface",
        apiName: objectSet.interfaceType
      };
    case "filter":
    case "asBaseObjectTypes":
    case "nearestNeighbors":
      return extractObjectOrInterfaceType(clientCtx, objectSet.objectSet);
    case "asType":
      return {
        type: clientCtx.narrowTypeInterfaceOrObjectMapping[objectSet.entityType],
        apiName: objectSet.entityType
      };
    case "intersect": {
      const objectSets2 = objectSet.objectSets;
      const objectSetTypes2 = await Promise.all(objectSets2.map((os) => extractObjectOrInterfaceType(clientCtx, os)));
      const filteredObjectTypes2 = objectSetTypes2.filter(Boolean);
      const firstInterfaceType = filteredObjectTypes2.find((val) => val?.type === "interface");
      !firstInterfaceType ? process.env.NODE_ENV !== "production" ? invariant5__default.default(false, `Missing interface type in intersect objectset scope'`) : invariant5__default.default(false) : void 0;
      return firstInterfaceType;
    }
    case "subtract":
    case "union":
      const objectSets = objectSet.objectSets;
      const objectSetTypes = await Promise.all(objectSets.map((os) => extractObjectOrInterfaceType(clientCtx, os)));
      const filteredObjectTypes = objectSetTypes.filter(Boolean);
      const firstObjectType = filteredObjectTypes[0];
      !filteredObjectTypes.every((val) => {
        return val?.apiName === firstObjectType?.apiName && val?.type === firstObjectType?.type;
      }) ? process.env.NODE_ENV !== "production" ? invariant5__default.default(false, "Can only have one object type when doing subtract, union") : invariant5__default.default(false) : void 0;
      return filteredObjectTypes[0];
    case "static":
    case "reference":
      return void 0;
    case "interfaceLinkSearchAround":
      const def = await extractObjectOrInterfaceType(clientCtx, objectSet.objectSet);
      if (def === void 0) {
        return void 0;
      }
      const objOrInterfaceDef = def.type === "object" ? await clientCtx.ontologyProvider.getObjectDefinition(def.apiName) : await clientCtx.ontologyProvider.getInterfaceDefinition(def.apiName);
      const linkDef = objOrInterfaceDef.links[objectSet.interfaceLink];
      !linkDef ? process.env.NODE_ENV !== "production" ? invariant5__default.default(false, `Missing link definition for '${objectSet.interfaceLink}'`) : invariant5__default.default(false) : void 0;
      return objOrInterfaceDef.type === "object" ? {
        apiName: objOrInterfaceDef.links[objectSet.interfaceLink].targetType,
        type: "object"
      } : {
        apiName: objOrInterfaceDef.links[objectSet.interfaceLink].targetTypeApiName,
        type: objOrInterfaceDef.links[objectSet.interfaceLink].targetType
      };
    // We don't have to worry about new object sets being added and doing a runtime break and breaking people since the OSDK is always constructing these.
    default:
      process.env.NODE_ENV !== "production" ? invariant5__default.default(false, `Unsupported object set type for deriving object or interface type,`) : invariant5__default.default(false) ;
  }
}
async function extractRdpDefinition(clientCtx, objectSet) {
  if (!hasWithProperties(objectSet)) {
    return {};
  }
  return (await extractRdpDefinitionInternal(clientCtx, objectSet, void 0)).definitions;
}
function hasWithProperties(objectSet) {
  if (objectSet.type === "withProperties") {
    return true;
  }
  if ("objectSet" in objectSet && objectSet.objectSet != null) {
    return hasWithProperties(objectSet.objectSet);
  }
  if ("objectSets" in objectSet && objectSet.objectSets != null) {
    return objectSet.objectSets.some(hasWithProperties);
  }
  return false;
}
async function extractRdpDefinitionInternal(clientCtx, objectSet, methodInputObjectType) {
  switch (objectSet.type) {
    case "searchAround": {
      const {
        definitions: definitions2,
        childObjectType
      } = await extractRdpDefinitionInternal(clientCtx, objectSet.objectSet, methodInputObjectType);
      if (childObjectType === void 0 || childObjectType === "") {
        return {
          definitions: {}
        };
      }
      const objDef = await clientCtx.ontologyProvider.getObjectDefinition(childObjectType);
      const linkDef = objDef.links[objectSet.link];
      !linkDef ? process.env.NODE_ENV !== "production" ? invariant5__default.default(false, `Missing link definition for '${objectSet.link}'`) : invariant5__default.default(false) : void 0;
      return {
        definitions: definitions2,
        childObjectType: objDef.links[objectSet.link].targetType
      };
    }
    case "withProperties": {
      const {
        definitions: definitions2,
        childObjectType
      } = await extractRdpDefinitionInternal(clientCtx, objectSet.objectSet, methodInputObjectType);
      if (childObjectType === void 0 || childObjectType === "") {
        return {
          definitions: {}
        };
      }
      for (const [name, definition] of Object.entries(objectSet.derivedProperties)) {
        if (definition.type !== "selection") {
          definitions2[name] = {
            selectedOrCollectedPropertyType: void 0,
            definition
          };
          continue;
        }
        switch (definition.operation.type) {
          case "collectList":
          case "collectSet":
          case "get":
          case "min":
          case "max":
            const {
              childObjectType: operationLevelObjectType
            } = await extractRdpDefinitionInternal(clientCtx, definition.objectSet, childObjectType);
            if (operationLevelObjectType === void 0 || operationLevelObjectType === "") {
              return {
                definitions: {}
              };
            }
            const objDef = await clientCtx.ontologyProvider.getObjectDefinition(operationLevelObjectType);
            definitions2[name] = {
              selectedOrCollectedPropertyType: objDef.properties[definition.operation.selectedPropertyApiName],
              definition
            };
            break;
          default:
            definitions2[name] = {
              selectedOrCollectedPropertyType: void 0,
              definition
            };
        }
      }
      return {
        definitions: definitions2,
        childObjectType
      };
    }
    case "methodInput":
      return {
        definitions: {},
        childObjectType: methodInputObjectType
      };
    case "base":
      return {
        definitions: {},
        childObjectType: objectSet.objectType
      };
    case "interfaceBase":
      return {
        definitions: {},
        childObjectType: objectSet.interfaceType
      };
    case "filter":
    case "asBaseObjectTypes":
    case "asType":
    case "nearestNeighbors":
      return extractRdpDefinitionInternal(clientCtx, objectSet.objectSet, methodInputObjectType);
    // These will throw in OSS so we should throw here so no request is made
    case "intersect":
    case "subtract":
    case "union":
      const objectSets = objectSet.objectSets;
      const objectSetTypes = await Promise.all(objectSets.map((os) => extractRdpDefinitionInternal(clientCtx, os, methodInputObjectType)));
      const definitions = objectSetTypes.reduce((acc, {
        definitions: definitions2
      }) => ({
        ...acc,
        ...definitions2
      }), {});
      !(Object.keys(definitions).length === 0) ? process.env.NODE_ENV !== "production" ? invariant5__default.default(false, "Object sets combined using intersect, subtract, or union must not contain any derived property definitions") : invariant5__default.default(false) : void 0;
      const firstValidChildObjectType = objectSetTypes.find(({
        childObjectType
      }) => childObjectType != null)?.childObjectType;
      !objectSetTypes.every(({
        childObjectType
      }) => childObjectType === firstValidChildObjectType || childObjectType == null) ? process.env.NODE_ENV !== "production" ? invariant5__default.default(false, "All object sets in an intersect, subtract, or union must have the same child object type") : invariant5__default.default(false) : void 0;
      return {
        definitions: {},
        childObjectType: firstValidChildObjectType
      };
    case "static":
    case "reference":
      return {
        definitions: {}
      };
    // We don't have to worry about new object sets being added and doing a runtime break and breaking people since the OSDK is always constructing these.
    case "interfaceLinkSearchAround":
      process.env.NODE_ENV !== "production" ? invariant5__default.default(false, `Unsupported object set type for Runtime Derived Properties`) : invariant5__default.default(false) ;
    default:
      process.env.NODE_ENV !== "production" ? invariant5__default.default(false, `Unsupported object set type for Runtime Derived Properties`) : invariant5__default.default(false) ;
  }
}

// src/util/objectSetUtils.ts
function resolveBaseObjectSetType(objectType) {
  return objectType.type === "interface" ? {
    type: "interfaceBase",
    interfaceType: objectType["apiName"]
  } : {
    type: "base",
    objectType: objectType["apiName"]
  };
}

// src/object/fetchPage.ts
function modifierToLoadLevelType(modifier) {
  switch (modifier) {
    case "applyMainValue":
      return "extractMainValue";
    case "applyReducers":
      return "applyReducers";
    case "applyReducersAndExtractMainValue":
      return "applyReducersAndExtractMainValue";
    default: {
      const _exhaustiveCheck = modifier;
      throw new Error(`Unknown modifier: ${_exhaustiveCheck}`);
    }
  }
}
function buildSelectV2(select, modifiers, allProperties) {
  const modifiersMap = modifiers ?? {};
  const modifierProps = new Set(Object.keys(modifiersMap));
  const hasModifiers = modifierProps.size > 0;
  const entries = [];
  if (select && select.length > 0) {
    for (const [prop, _] of Object.entries(modifiersMap)) {
      !select.includes(prop) ? process.env.NODE_ENV !== "production" ? invariant5__default.default(false, "Modified properties must be included in $select when manually specifying properties") : invariant5__default.default(false) : void 0;
    }
    for (const prop of select) {
      if (modifierProps.has(prop)) {
        entries.push({
          type: "propertyWithLoadLevel",
          propertyIdentifier: {
            type: "property",
            apiName: prop
          },
          loadLevel: {
            type: modifierToLoadLevelType(modifiersMap[prop])
          }
        });
      } else {
        entries.push({
          type: "property",
          apiName: prop
        });
      }
    }
  } else if (hasModifiers && allProperties && allProperties.length > 0) {
    for (const prop of allProperties) {
      if (modifierProps.has(prop)) {
        entries.push({
          type: "propertyWithLoadLevel",
          propertyIdentifier: {
            type: "property",
            apiName: prop
          },
          loadLevel: {
            type: modifierToLoadLevelType(modifiersMap[prop])
          }
        });
      } else {
        entries.push({
          type: "property",
          apiName: prop
        });
      }
    }
  }
  return entries;
}
function augment(type, ...properties) {
  return {
    [type.apiName]: properties
  };
}
function resolveInterfaceObjectSet(objectSet, interfaceTypeApiName, args) {
  return args?.$includeAllBaseObjectProperties ? {
    type: "intersect",
    objectSets: [objectSet, {
      type: "interfaceBase",
      interfaceType: interfaceTypeApiName,
      includeAllBaseObjectProperties: true
    }]
  } : objectSet;
}
async function fetchStaticRidPage(client, rids, args) {
  const shouldLoadPropertySecurities = args.$loadPropertySecurityMetadata ?? false;
  const requestBody = await applyFetchArgs(args, {
    objectSet: {
      type: "static",
      objects: rids
    },
    select: args?.$select ?? [],
    excludeRid: !args?.$includeRid,
    snapshot: args.$snapshot ?? false,
    loadPropertySecurities: shouldLoadPropertySecurities
  }, client, {
    type: "object",
    apiName: ""
  });
  if (client.flushEdits != null) {
    await client.flushEdits();
  }
  const result = await loadMultipleObjectTypes(addUserAgentAndRequestContextHeaders(client, {
    osdkMetadata: void 0
  }), await client.ontologyRid, requestBody, {
    preview: true,
    transactionId: client.transactionId,
    scenarioRid: client.scenarioRid
  });
  return Promise.resolve({
    data: await client.objectFactory(client, result.data, void 0, {}, shouldLoadPropertySecurities ? result.propertySecurities : void 0, !args.$includeRid, args.$select, false, result.interfaceToObjectTypeMappings, result.interfaceToObjectTypeMappingsV2),
    nextPageToken: result.nextPageToken,
    totalCount: result.totalCount
  });
}
async function fetchInterfacePage(client, interfaceType, args, objectSet) {
  const extractedInterfaceTypeApiName = (await extractObjectOrInterfaceType(client, objectSet))?.apiName ?? interfaceType.apiName;
  const resolvedInterfaceObjectSet = resolveInterfaceObjectSet(objectSet, extractedInterfaceTypeApiName, args);
  const shouldLoadPropertySecurities = args.$loadPropertySecurityMetadata ?? false;
  const modifiers = args.$applyModifiers;
  const hasModifiers = modifiers && Object.keys(modifiers).length > 0;
  const hasSelect = args?.$select && args.$select.length > 0;
  let allProperties;
  if (!hasSelect && hasModifiers) {
    const ifaceDef = await client.ontologyProvider.getInterfaceDefinition(interfaceType.apiName);
    allProperties = ifaceDef ? Object.keys(ifaceDef.properties) : void 0;
  }
  const selectV2 = buildSelectV2(args?.$select ? [...args.$select] : void 0, modifiers, allProperties);
  const requestBody = await buildAndRemapRequestBody(args, {
    objectSet: resolvedInterfaceObjectSet,
    select: [],
    selectV2,
    loadPropertySecurities: shouldLoadPropertySecurities,
    excludeRid: !args?.$includeRid,
    snapshot: args.$snapshot ?? false
  }, client, interfaceType);
  if (client.flushEdits != null) {
    await client.flushEdits();
  }
  const result = await loadMultipleObjectTypes(addUserAgentAndRequestContextHeaders(client, interfaceType), await client.ontologyRid, requestBody, {
    preview: true,
    branch: client.branch,
    transactionId: client.transactionId,
    scenarioRid: client.scenarioRid
  });
  return Promise.resolve({
    data: await client.objectFactory(client, result.data, extractedInterfaceTypeApiName, {}, shouldLoadPropertySecurities ? result.propertySecurities : void 0, !args.$includeRid, args.$select, false, result.interfaceToObjectTypeMappings, result.interfaceToObjectTypeMappingsV2),
    nextPageToken: result.nextPageToken,
    totalCount: result.totalCount
  });
}
async function fetchPageInternal(client, objectType, objectSet, args = {}) {
  if (objectType.type === "interface") {
    return await fetchInterfacePage(client, objectType, args, objectSet);
  } else {
    return await fetchObjectPage(client, objectType, args, objectSet);
  }
}
async function fetchPageWithErrorsInternal(client, objectType, objectSet, args = {}) {
  try {
    const result = await fetchPageInternal(client, objectType, objectSet, args);
    return {
      value: result
    };
  } catch (e) {
    if (e instanceof Error) {
      return {
        error: e
      };
    }
    return {
      error: e
    };
  }
}
async function fetchPage(client, objectType, args, objectSet = resolveBaseObjectSetType(objectType)) {
  return fetchPageInternal(client, objectType, objectSet, args);
}
async function buildAndRemapRequestBody(args, baseBody, client, objectType) {
  const requestBody = await applyFetchArgs(args, baseBody, client, objectType);
  if (requestBody.selectV2 != null && requestBody.selectV2.length > 0) {
    const remapped = remapSelectV2(objectType, requestBody.selectV2);
    return {
      ...requestBody,
      selectV2: remapped
    };
  }
  return requestBody;
}
function remapSelectV2(objectOrInterface, selectV2) {
  if (objectOrInterface == null) {
    return selectV2;
  }
  if (objectOrInterface.type !== "interface") {
    return selectV2;
  }
  const [objApiNamespace] = extractNamespace(objectOrInterface.apiName);
  if (objApiNamespace == null) {
    return selectV2;
  }
  return selectV2.map((entry) => {
    if (entry.type === "property") {
      const [fieldApiNamespace, fieldShortName] = extractNamespace(entry.apiName);
      if (fieldApiNamespace == null) {
        return {
          type: "property",
          apiName: `${objApiNamespace}.${fieldShortName}`
        };
      }
      return entry;
    } else {
      const [fieldApiNamespace, fieldShortName] = extractNamespace(entry.propertyIdentifier.apiName);
      if (fieldApiNamespace == null) {
        return {
          ...entry,
          propertyIdentifier: {
            type: "property",
            apiName: `${objApiNamespace}.${fieldShortName}`
          }
        };
      }
      return entry;
    }
  });
}
function remapPropertyNames(objectOrInterface, propertyNames) {
  if (objectOrInterface == null) {
    return propertyNames;
  }
  if (objectOrInterface.type === "interface") {
    const [objApiNamespace] = extractNamespace(objectOrInterface.apiName);
    return propertyNames.map((name) => {
      const [fieldApiNamespace, fieldShortName] = extractNamespace(name);
      return fieldApiNamespace == null && objApiNamespace != null ? `${objApiNamespace}.${fieldShortName}` : name;
    });
  }
  return propertyNames;
}
async function applyFetchArgs(args, body, _client, objectType) {
  if (args?.$nextPageToken) {
    body.pageToken = args.$nextPageToken;
  }
  if (args?.$pageSize != null) {
    body.pageSize = args.$pageSize;
  }
  if (args?.$loadPropertySecurityMetadata) {
    body.loadPropertySecurities = true;
  }
  const orderBy = args?.$orderBy;
  if (orderBy) {
    if (orderBy === "relevance") {
      body.orderBy = {
        orderType: "relevance",
        fields: []
      };
    } else {
      const orderByEntries = Object.entries(orderBy);
      const fieldNames = orderByEntries.map(([field]) => field);
      const remappedFields = remapPropertyNames(objectType, fieldNames);
      body.orderBy = {
        fields: orderByEntries.map(([, direction], index) => ({
          field: remappedFields[index],
          direction
        }))
      };
    }
  }
  return body;
}
async function fetchObjectPage(client, objectType, args, objectSet) {
  const modifiers = args.$applyModifiers;
  const hasModifiers = modifiers && Object.keys(modifiers).length > 0;
  const hasSelect = args?.$select && args.$select.length > 0;
  let allProperties;
  if (!hasSelect && hasModifiers) {
    const objDef = await client.ontologyProvider.getObjectDefinition(objectType.apiName);
    allProperties = objDef ? Object.keys(objDef.properties) : void 0;
  } else {
    void client.ontologyProvider.getObjectDefinition(objectType.apiName).catch(() => {
    });
  }
  const shouldLoadPropertySecurities = args.$loadPropertySecurityMetadata ?? false;
  const selectV2 = buildSelectV2(args?.$select ? [...args.$select] : void 0, modifiers, allProperties);
  const requestBody = await buildAndRemapRequestBody(args, {
    objectSet,
    select: [],
    selectV2,
    loadPropertySecurities: shouldLoadPropertySecurities,
    excludeRid: !args?.$includeRid,
    snapshot: args.$snapshot ?? false
  }, client, objectType);
  if (client.flushEdits != null) {
    await client.flushEdits();
  }
  const r = await load(addUserAgentAndRequestContextHeaders(client, objectType), await client.ontologyRid, requestBody, {
    branch: client.branch,
    transactionId: client.transactionId,
    scenarioRid: client.scenarioRid
  });
  return Promise.resolve({
    data: await client.objectFactory(client, r.data, void 0, await extractRdpDefinition(client, objectSet), shouldLoadPropertySecurities ? r.propertySecurities : void 0, !args.$includeRid, args.$select, false),
    nextPageToken: r.nextPageToken,
    totalCount: r.totalCount
  });
}

// src/internal/conversions/fullyQualifyPropName.ts
function fullyQualifyPropName(fieldName, objectOrInterface) {
  if (objectOrInterface.type === "interface") {
    const [objApiNamespace] = extractNamespace(objectOrInterface.apiName);
    const [fieldApiNamespace, fieldShortName] = extractNamespace(fieldName);
    return fieldApiNamespace == null && objApiNamespace != null ? `${objApiNamespace}.${fieldShortName}` : fieldName;
  }
  return fieldName;
}

// src/internal/conversions/makeGeoFilterBbox.ts
function makeGeoFilterBbox(bbox, filterType, propertyIdentifier, field) {
  return {
    type: filterType === "$within" ? "withinBoundingBox" : "intersectsBoundingBox",
    /**
     * This is a bit ugly, but did this so that propertyIdentifier only shows up in the return object if its defined,
     * this makes it so we don't need to go update our entire test bed either to include a field which may change in near future.
     * Once we solidify that this is the way forward, I can remove field and clean this up
     */
    ...propertyIdentifier != null && {
      propertyIdentifier
    },
    field,
    value: {
      topLeft: {
        type: "Point",
        coordinates: [bbox[0], bbox[3]]
      },
      bottomRight: {
        type: "Point",
        coordinates: [bbox[2], bbox[1]]
      }
    }
  };
}

// src/internal/conversions/makeGeoFilterPolygon.ts
function makeGeoFilterPolygon(coordinates, filterType, propertyIdentifier, field) {
  return {
    type: filterType,
    ...propertyIdentifier != null && {
      propertyIdentifier
    },
    field,
    value: {
      type: "Polygon",
      coordinates
    }
  };
}

// src/internal/conversions/makeGeoFilterIntersects.ts
function makeGeoFilterIntersects(intersectsBody, propertyIdentifier, field) {
  if (Array.isArray(intersectsBody)) {
    return makeGeoFilterBbox(intersectsBody, "$intersects", propertyIdentifier, field);
  } else if ("$bbox" in intersectsBody && intersectsBody.$bbox != null) {
    return makeGeoFilterBbox(intersectsBody.$bbox, "$intersects", propertyIdentifier, field);
  } else {
    const coordinates = "$polygon" in intersectsBody ? intersectsBody.$polygon : intersectsBody.coordinates;
    return makeGeoFilterPolygon(coordinates, "intersectsPolygon", propertyIdentifier, field);
  }
}
function makeGeoFilterWithin(withinBody, propertyIdentifier, field) {
  if (Array.isArray(withinBody)) {
    return makeGeoFilterBbox(withinBody, "$within", propertyIdentifier, field);
  } else if ("$bbox" in withinBody && withinBody.$bbox != null) {
    return makeGeoFilterBbox(withinBody.$bbox, "$within", propertyIdentifier, field);
  } else if ("$distance" in withinBody && "$of" in withinBody && withinBody.$distance != null && withinBody.$of != null) {
    return {
      type: "withinDistanceOf",
      ...propertyIdentifier != null && {
        propertyIdentifier
      },
      field,
      value: {
        center: Array.isArray(withinBody.$of) ? {
          type: "Point",
          coordinates: withinBody.$of
        } : withinBody.$of,
        distance: {
          value: withinBody.$distance[0],
          unit: api.DistanceUnitMapping[withinBody.$distance[1]]
        }
      }
    };
  } else {
    const coordinates = "$polygon" in withinBody ? withinBody.$polygon : withinBody.coordinates;
    return makeGeoFilterPolygon(coordinates, "withinPolygon", propertyIdentifier, field);
  }
}
function toIntervalQueryRule(rule) {
  if (rule.$match != null) {
    if (rule.$prefixOnLastTerm) {
      return {
        type: "prefixOnLastToken",
        query: rule.$match
      };
    }
    return {
      type: "match",
      query: rule.$match,
      ordered: rule.$ordered,
      maxGaps: rule.$maxGaps
    };
  }
  if (rule.$and != null) {
    return {
      type: "allOf",
      rules: rule.$and.map(toIntervalQueryRule),
      ordered: rule.$ordered,
      maxGaps: rule.$maxGaps
    };
  }
  if (rule.$or != null) {
    return {
      type: "anyOf",
      rules: rule.$or.map(toIntervalQueryRule)
    };
  }
  if (rule.$fuzzy != null) {
    return {
      type: "fuzzy",
      term: rule.$fuzzy,
      fuzziness: rule.$fuzziness
    };
  }
  process.env.NODE_ENV !== "production" ? invariant5__default.default(false, "Unknown interval rule type") : invariant5__default.default(false) ;
}

// src/internal/conversions/modernToLegacyWhereClause.ts
function isAndClause(whereClause) {
  return "$and" in whereClause && whereClause.$and !== void 0;
}
function isOrClause(whereClause) {
  return "$or" in whereClause && whereClause.$or !== void 0;
}
function isNotClause(whereClause) {
  return "$not" in whereClause && whereClause.$not !== void 0;
}
function modernToLegacyWhereClause(whereClause, objectOrInterface, rdpNames) {
  const parts = Object.entries(whereClause).map(([key, value]) => ({
    [key]: value
  }));
  if (parts.length === 1) {
    return modernToLegacyWhereClauseInner(whereClause, objectOrInterface, rdpNames);
  }
  return {
    type: "and",
    value: parts.map((v) => modernToLegacyWhereClauseInner(v, objectOrInterface, rdpNames))
  };
}
function modernToLegacyWhereClauseInner(whereClause, objectOrInterface, rdpNames) {
  const parts = Object.entries(whereClause);
  !(parts.length === 1) ? process.env.NODE_ENV !== "production" ? invariant5__default.default(false, "Invalid where clause provided.") : invariant5__default.default(false) : void 0;
  if (isAndClause(whereClause)) {
    return {
      type: "and",
      value: whereClause.$and.map((clause) => modernToLegacyWhereClause(clause, objectOrInterface, rdpNames))
    };
  } else if (isOrClause(whereClause)) {
    return {
      type: "or",
      value: whereClause.$or.map((clause) => modernToLegacyWhereClause(clause, objectOrInterface, rdpNames))
    };
  } else if (isNotClause(whereClause)) {
    return {
      type: "not",
      value: modernToLegacyWhereClause(whereClause.$not, objectOrInterface, rdpNames)
    };
  }
  return handleWherePair(parts[0], objectOrInterface, void 0, rdpNames);
}
function resolvePropertyIdentifier(fieldName, objectOrInterface, isTitleProperty, isPrimaryKeyProperty, isRdp, structFieldSelector) {
  if (isTitleProperty) {
    return {
      type: "titleProperty"
    };
  } else if (isPrimaryKeyProperty) {
    return {
      type: "primaryKeyProperty"
    };
  } else if (isRdp) {
    return {
      type: "property",
      apiName: fieldName
    };
  } else if (structFieldSelector != null) {
    return {
      type: "structField",
      ...structFieldSelector,
      propertyApiName: fullyQualifyPropName(structFieldSelector.propertyApiName, objectOrInterface)
    };
  }
  return void 0;
}
function handleWherePair([fieldName, filter], objectOrInterface, structFieldSelector, rdpNames) {
  !(filter != null) ? process.env.NODE_ENV !== "production" ? invariant5__default.default(false, `Cannot filter on property "${fieldName}" with an undefined or null value. If the value might be undefined, check it before adding to the where clause.`) : invariant5__default.default(false) : void 0;
  const isTitleProperty = fieldName === "$title";
  const isPrimaryKeyProperty = fieldName === "$primaryKey";
  const isSpecialProperty = isTitleProperty || isPrimaryKeyProperty;
  const isRdp = !structFieldSelector && !isSpecialProperty && rdpNames?.has(fieldName);
  const propertyIdentifier = resolvePropertyIdentifier(fieldName, objectOrInterface, isTitleProperty, isPrimaryKeyProperty, isRdp, structFieldSelector);
  const field = !isRdp && !isSpecialProperty && structFieldSelector == null ? fullyQualifyPropName(fieldName, objectOrInterface) : void 0;
  !(field == null || propertyIdentifier == null && (field != null || isRdp != null)) ? process.env.NODE_ENV !== "production" ? invariant5__default.default(false, "Encountered error constructing where clause: field and propertyIdentifier cannot both be defined") : invariant5__default.default(false) : void 0;
  if (typeof filter === "string" || typeof filter === "number" || typeof filter === "boolean") {
    return {
      type: "eq",
      ...propertyIdentifier != null && {
        propertyIdentifier
      },
      field,
      value: filter
    };
  }
  const keysOfFilter = Object.keys(filter);
  const hasDollarSign = keysOfFilter.some((key) => key.startsWith("$"));
  !(!hasDollarSign || keysOfFilter.length === 1) ? process.env.NODE_ENV !== "production" ? invariant5__default.default(false, "A WhereClause Filter with multiple clauses/fields is not allowed. Instead, use an 'or'/'and' clause to combine multiple filters.") : invariant5__default.default(false) : void 0;
  if (!hasDollarSign) {
    const structFilter = Object.entries(filter);
    !(structFilter.length === 1) ? process.env.NODE_ENV !== "production" ? invariant5__default.default(false, "Cannot filter on more than one struct field in the same clause, need to use an and clause") : invariant5__default.default(false) : void 0;
    const structFieldApiName = keysOfFilter[0];
    return handleWherePair(Object.entries(filter)[0], objectOrInterface, {
      propertyApiName: fieldName,
      structFieldApiName
    }, rdpNames);
  }
  const firstKey = keysOfFilter[0];
  !(filter[firstKey] != null) ? process.env.NODE_ENV !== "production" ? invariant5__default.default(false) : invariant5__default.default(false) : void 0;
  if (firstKey === "$contains" && filter[firstKey] instanceof Object) {
    const containsValue = filter[firstKey];
    const containsKeys = Object.keys(containsValue);
    const isFilterObject = containsKeys.some((key) => key.startsWith("$"));
    if (isFilterObject) {
      return handleWherePair([fieldName, containsValue], objectOrInterface, structFieldSelector, rdpNames);
    } else {
      const structFilter = Object.entries(containsValue);
      !(structFilter.length === 1) ? process.env.NODE_ENV !== "production" ? invariant5__default.default(false, "Cannot filter on more than one struct field in the same clause, need to use an and clause") : invariant5__default.default(false) : void 0;
      const structFieldApiName = structFilter[0][0];
      return handleWherePair(structFilter[0], objectOrInterface, {
        propertyApiName: fieldName,
        structFieldApiName
      });
    }
  }
  if (firstKey === "$ne") {
    return {
      type: "not",
      value: {
        type: "eq",
        ...propertyIdentifier != null && {
          propertyIdentifier
        },
        field,
        value: filter[firstKey]
      }
    };
  }
  if (firstKey === "$within") {
    return makeGeoFilterWithin(filter[firstKey], propertyIdentifier, field);
  }
  if (firstKey === "$intersects") {
    return makeGeoFilterIntersects(filter[firstKey], propertyIdentifier, field);
  }
  if (firstKey === "$containsAllTerms" || firstKey === "$containsAnyTerm") {
    return {
      type: firstKey.substring(1),
      ...propertyIdentifier != null && {
        propertyIdentifier
      },
      field,
      value: typeof filter[firstKey] === "string" ? filter[firstKey] : filter[firstKey].term,
      fuzzy: typeof filter[firstKey] === "string" ? false : filter[firstKey].fuzzySearch ?? false
    };
  }
  if (firstKey === "$matchesRegex") {
    return {
      type: "regex",
      ...propertyIdentifier != null && {
        propertyIdentifier
      },
      field,
      value: filter[firstKey]
    };
  }
  if (firstKey === "$interval") {
    return {
      type: "interval",
      ...propertyIdentifier != null && {
        propertyIdentifier
      },
      field,
      rule: toIntervalQueryRule(filter[firstKey])
    };
  }
  return {
    type: firstKey.substring(1),
    ...propertyIdentifier != null && {
      propertyIdentifier
    },
    field,
    value: filter[firstKey]
  };
}
function derivedPropertyDefinitionFactory(wireDefinition, definitionMap) {
  const definition = {
    abs() {
      return derivedPropertyDefinitionFactory({
        type: "absoluteValue",
        property: wireDefinition
      }, definitionMap);
    },
    negate() {
      return derivedPropertyDefinitionFactory({
        type: "negate",
        property: wireDefinition
      }, definitionMap);
    },
    max(value) {
      return derivedPropertyDefinitionFactory({
        type: "greatest",
        properties: [wireDefinition, getDefinitionFromMap(value, definitionMap)]
      }, definitionMap);
    },
    min(value) {
      return derivedPropertyDefinitionFactory({
        type: "least",
        properties: [wireDefinition, getDefinitionFromMap(value, definitionMap)]
      }, definitionMap);
    },
    add(value) {
      return derivedPropertyDefinitionFactory({
        type: "add",
        properties: [wireDefinition, getDefinitionFromMap(value, definitionMap)]
      }, definitionMap);
    },
    subtract(value) {
      return derivedPropertyDefinitionFactory({
        "type": "subtract",
        "left": wireDefinition,
        "right": getDefinitionFromMap(value, definitionMap)
      }, definitionMap);
    },
    multiply(value) {
      return derivedPropertyDefinitionFactory({
        type: "multiply",
        properties: [wireDefinition, getDefinitionFromMap(value, definitionMap)]
      }, definitionMap);
    },
    divide(value) {
      return derivedPropertyDefinitionFactory({
        "type": "divide",
        "left": wireDefinition,
        "right": getDefinitionFromMap(value, definitionMap)
      }, definitionMap);
    },
    extractPart: (part) => {
      return derivedPropertyDefinitionFactory({
        type: "extract",
        part,
        property: wireDefinition
      }, definitionMap);
    }
  };
  definitionMap.set(definition, wireDefinition);
  return definition;
}
var getDefinitionFromMap = (arg, definitionMap) => {
  if (typeof arg === "object") {
    const definition = definitionMap.get(arg);
    !definition ? process.env.NODE_ENV !== "production" ? invariant5__default.default(false, "Derived Property is not defined") : invariant5__default.default(false) : void 0;
    return definition;
  } else if (typeof arg === "number") {
    process.env.NODE_ENV !== "production" ? invariant5__default.default(false, "Literals for derived properties are not yet supported") : invariant5__default.default(false) ;
  }
  process.env.NODE_ENV !== "production" ? invariant5__default.default(false, "Invalid argument type for a derived property") : invariant5__default.default(false) ;
};

// src/derivedProperties/createWithPropertiesObjectSet.ts
function createWithPropertiesObjectSet(objectType, objectSet, definitionMap, fromBaseObjectSet = false) {
  return {
    pivotTo: (link) => {
      return createWithPropertiesObjectSet(objectType, {
        type: "searchAround",
        objectSet,
        link
      }, definitionMap);
    },
    where: (clause) => {
      const rdpNames = new Set(definitionMap.keys());
      return createWithPropertiesObjectSet(objectType, {
        type: "filter",
        objectSet,
        where: modernToLegacyWhereClause(clause, objectType, rdpNames)
      }, definitionMap);
    },
    aggregate: (aggregation, opt) => {
      const splitAggregation = aggregation.split(":");
      !(splitAggregation.length === 2 || splitAggregation[0] === "$count") ? process.env.NODE_ENV !== "production" ? invariant5__default.default(false, "Invalid aggregation format") : invariant5__default.default(false) : void 0;
      const [aggregationPropertyName, aggregationOperation] = splitAggregation;
      let aggregationOperationDefinition;
      switch (aggregationOperation) {
        case "sum":
        case "avg":
        case "min":
        case "max":
        case "exactDistinct":
        case "approximateDistinct":
          aggregationOperationDefinition = {
            type: aggregationOperation,
            selectedPropertyApiName: aggregationPropertyName
          };
          break;
        case "approximatePercentile":
          aggregationOperationDefinition = {
            type: "approximatePercentile",
            selectedPropertyApiName: aggregationPropertyName,
            approximatePercentile: opt?.percentile ?? 0.5
          };
          break;
        case "collectSet":
        case "collectList":
          aggregationOperationDefinition = {
            type: aggregationOperation,
            selectedPropertyApiName: aggregationPropertyName,
            limit: opt?.limit ?? 100
          };
          break;
        case void 0:
          if (aggregationPropertyName === "$count") {
            aggregationOperationDefinition = {
              type: "count"
            };
            break;
          }
        default:
          process.env.NODE_ENV !== "production" ? invariant5__default.default(false, "Invalid aggregation operation " + aggregationOperation) : invariant5__default.default(false) ;
      }
      const wrappedObjectSet = {
        type: "selection",
        objectSet,
        operation: aggregationOperationDefinition
      };
      const selectorResult = derivedPropertyDefinitionFactory(wrappedObjectSet, definitionMap);
      definitionMap.set(selectorResult, wrappedObjectSet);
      return selectorResult;
    },
    selectProperty: (name) => {
      if (fromBaseObjectSet) {
        const wrappedObjectSet2 = {
          type: "property",
          apiName: name
        };
        const selectorResult2 = derivedPropertyDefinitionFactory(wrappedObjectSet2, definitionMap);
        definitionMap.set(selectorResult2, wrappedObjectSet2);
        return selectorResult2;
      }
      const wrappedObjectSet = {
        type: "selection",
        objectSet,
        operation: {
          type: "get",
          selectedPropertyApiName: name
        }
      };
      const selectorResult = derivedPropertyDefinitionFactory(wrappedObjectSet, definitionMap);
      definitionMap.set(selectorResult, wrappedObjectSet);
      return selectorResult;
    }
  };
}

// src/internal/conversions/modernToLegacyAggregationClause.ts
var directionFieldMap = (dir) => dir === "asc" ? "ASC" : dir === "desc" ? "DESC" : void 0;
function splitAggregationKey(key) {
  const colonPos = key.lastIndexOf(":");
  return {
    property: key.slice(0, colonPos),
    metric: key.slice(colonPos + 1)
  };
}
function modernToLegacyAggregationClause(select) {
  return Object.entries(select).flatMap(([propAndMetric, aggregationType]) => {
    if (propAndMetric === "$count") {
      return {
        type: "count",
        name: "count",
        direction: directionFieldMap(aggregationType)
      };
    }
    const {
      property,
      metric
    } = splitAggregationKey(propAndMetric);
    return [{
      type: metric,
      name: `${property}.${metric}`,
      direction: directionFieldMap(aggregationType),
      field: property
    }];
  });
}

// src/internal/conversions/legacyToModernSingleAggregationResult.ts
function legacyToModernSingleAggregationResult(entry, select) {
  const result = {};
  for (const selectKey of Object.keys(select)) {
    if (selectKey === "$count") {
      continue;
    }
    const {
      property,
      metric
    } = splitAggregationKey(selectKey);
    (result[property] ??= {})[metric] = void 0;
  }
  for (const {
    name,
    value
  } of entry.metrics) {
    if (name === "count") {
      continue;
    }
    const [property, metricType] = name.split(".");
    if (result[property]) {
      result[property][metricType] = value;
    }
  }
  return result;
}
function modernToLegacyGroupByClause(groupByClause) {
  if (!groupByClause) return [];
  return Object.entries(groupByClause).flatMap(([field, type]) => {
    if (type === "exact") {
      return [{
        type,
        field
      }];
    } else if ("$exactWithLimit" in type) {
      {
        return [{
          type: "exact",
          field,
          maxGroupCount: type.$exactWithLimit
        }];
      }
    } else if ("$exact" in type) {
      return [{
        type: "exact",
        field,
        maxGroupCount: type.$exact?.$limit ?? void 0,
        defaultValue: type.$exact.$defaultValue ?? void 0,
        includeNullValues: type.$exact.$includeNullValue === true ? true : void 0
      }];
    } else if ("$fixedWidth" in type) {
      return [{
        type: "fixedWidth",
        field,
        fixedWidth: type.$fixedWidth
      }];
    } else if ("$ranges" in type) {
      return [{
        type: "ranges",
        field,
        ranges: type.$ranges.map((range) => convertRange(range))
      }];
    } else if ("$duration" in type) {
      return [{
        type: "duration",
        field,
        value: type.$duration[0],
        unit: api.DurationMapping[type.$duration[1]]
      }];
    } else return [];
  });
}
function convertRange(range) {
  return {
    startValue: range[0],
    endValue: range[1]
  };
}

// src/object/aggregate.ts
async function aggregate2(clientCtx, objectType, objectSet = resolveBaseObjectSetType(objectType), req) {
  resolveBaseObjectSetType(objectType);
  const body = {
    aggregation: modernToLegacyAggregationClause(req.$select),
    groupBy: []};
  if (req.$groupBy) {
    body.groupBy = modernToLegacyGroupByClause(req.$groupBy);
  }
  if (clientCtx.flushEdits != null) {
    await clientCtx.flushEdits();
  }
  const result = await aggregate(addUserAgentAndRequestContextHeaders(clientCtx, objectType), await clientCtx.ontologyRid, {
    objectSet,
    groupBy: body.groupBy,
    aggregation: body.aggregation
  }, {
    branch: clientCtx.branch,
    transactionId: clientCtx.transactionId,
    scenarioRid: clientCtx.scenarioRid
  });
  if (!result.data || !Array.isArray(result.data)) {
    throw new Error(`Aggregation request failed: ${JSON.stringify(result)}`);
  }
  if (!req.$groupBy) {
    !(result.data.length === 1) ? process.env.NODE_ENV !== "production" ? invariant5__default.default(false, "no group by clause should mean only one data result") : invariant5__default.default(false) : void 0;
    return {
      ...aggregationToCountResult(result.data[0]),
      ...legacyToModernSingleAggregationResult(result.data[0], req.$select)
    };
  }
  const ret = result.data.map((entry) => {
    return {
      $group: entry.group,
      ...aggregationToCountResult(entry),
      ...legacyToModernSingleAggregationResult(entry, req.$select)
    };
  });
  return ret;
}
function aggregationToCountResult(entry) {
  for (const aggregateResult of entry.metrics) {
    if (aggregateResult.name === "count") {
      return {
        $count: aggregateResult.value
      };
    }
  }
}
async function fetchSingle(client, objectType, args, objectSet) {
  const result = await fetchPage(client, objectType, {
    ...args,
    $pageSize: 1
  }, objectSet);
  if (result.data.length !== 1 || result.nextPageToken != null) {
    throw new shared_net_errors.PalantirApiError(`Expected a single result but got ${result.data.length} instead${result.nextPageToken != null ? " with nextPageToken set" : ""}`);
  }
  return result.data[0];
}
async function fetchSingleWithErrors(client, objectType, args, objectSet) {
  try {
    const result = await fetchSingle(client, objectType, args, objectSet);
    return {
      value: result
    };
  } catch (e) {
    if (e instanceof Error) {
      return {
        error: e
      };
    }
    return {
      error: e
    };
  }
}

// src/util/WireObjectSet.ts
var WIRE_OBJECT_SET_TYPES = /* @__PURE__ */ new Set(["base", "filter", "intersect", "reference", "searchAround", "static", "subtract", "union", "interfaceBase"]);
function isWireObjectSet(o) {
  return o != null && typeof o === "object" && WIRE_OBJECT_SET_TYPES.has(o.type);
}

// src/objectSet/fetchLinksPage.ts
var fetchLinksPage = async (client, objectType, objectSet, links) => {
  if (objectType.type === "interface") {
    throw new Error("Interface object sets are not supported yet.");
  }
  void client.ontologyProvider.getObjectDefinition(objectType.apiName).catch(() => {
  });
  const result = await loadLinks(client, await client.ontologyRid, {
    objectSet,
    links
  }, {
    branch: client.branch,
    preview: true
  });
  return remapLinksPage(result);
};
var remapLinksPage = (wireLinksPage) => {
  return {
    ...wireLinksPage,
    data: wireLinksPage.data.flatMap(({
      sourceObject,
      linkedObjects
    }) => linkedObjects.map(({
      targetObject,
      linkType
    }) => ({
      source: remapObjectLocator(sourceObject),
      target: remapObjectLocator(targetObject),
      linkType
    })))
  };
};
var remapObjectLocator = (wireObjectLocator) => ({
  $apiName: wireObjectLocator.__apiName,
  $primaryKey: wireObjectLocator.__primaryKey
});

// src/objectSet/createObjectSet.ts
function isObjectTypeDefinition(def) {
  return def.type === "object";
}
function isObjectSet(o) {
  return o != null && typeof o === "object" && isWireObjectSet(objectSetDefinitions.get(o));
}
function getWireObjectSet(objectSet) {
  return objectSetDefinitions.get(objectSet);
}
var objectSetDefinitions = /* @__PURE__ */ new WeakMap();
function createObjectSet(objectType, clientCtx, objectSet = resolveBaseObjectSetType(objectType)) {
  const base = {
    aggregate: aggregate2.bind(globalThis, augmentRequestContext(clientCtx, (_) => ({
      finalMethodCall: "aggregate"
    })), objectType, objectSet),
    fetchPage: fetchPageInternal.bind(globalThis, augmentRequestContext(clientCtx, (_) => ({
      finalMethodCall: "fetchPage"
    })), objectType, objectSet),
    fetchPageWithErrors: fetchPageWithErrorsInternal.bind(globalThis, augmentRequestContext(clientCtx, (_) => ({
      finalMethodCall: "fetchPageWithErrors"
    })), objectType, objectSet),
    where: (clause) => {
      return clientCtx.objectSetFactory(objectType, clientCtx, {
        type: "filter",
        objectSet,
        where: modernToLegacyWhereClause(clause, objectType)
      });
    },
    pivotTo(type) {
      return createSearchAround(type)();
    },
    union: (...objectSets) => {
      return clientCtx.objectSetFactory(objectType, clientCtx, {
        type: "union",
        objectSets: [objectSet, ...objectSets.map((os) => objectSetDefinitions.get(os))]
      });
    },
    intersect: (...objectSets) => {
      return clientCtx.objectSetFactory(objectType, clientCtx, {
        type: "intersect",
        objectSets: [objectSet, ...objectSets.map((os) => objectSetDefinitions.get(os))]
      });
    },
    subtract: (...objectSets) => {
      return clientCtx.objectSetFactory(objectType, clientCtx, {
        type: "subtract",
        objectSets: [objectSet, ...objectSets.map((os) => objectSetDefinitions.get(os))]
      });
    },
    nearestNeighbors: (query, numNeighbors, property) => {
      const nearestNeighborsQuery = isTextQuery(query) ? {
        "type": "text",
        "value": query
      } : {
        "type": "vector",
        "value": query
      };
      return clientCtx.objectSetFactory(objectType, clientCtx, {
        type: "nearestNeighbors",
        objectSet,
        propertyIdentifier: {
          type: "property",
          apiName: property
        },
        numNeighbors,
        query: nearestNeighborsQuery
      });
    },
    async *asyncIter(args) {
      let $nextPageToken;
      do {
        const result = await fetchPageInternal(augmentRequestContext(clientCtx, (_) => ({
          finalMethodCall: "asyncIter"
        })), objectType, objectSet, {
          ...args,
          $pageSize: 1e4,
          $nextPageToken,
          $snapshot: true
        });
        $nextPageToken = result.nextPageToken;
        for (const obj of result.data) {
          yield obj;
        }
      } while ($nextPageToken != null);
    },
    fetchOne: isObjectTypeDefinition(objectType) ? async (primaryKey, options) => {
      return await fetchSingle(augmentRequestContext(clientCtx, (_) => ({
        finalMethodCall: "fetchOne"
      })), objectType, options, await createWithPk(clientCtx, objectType, objectSet, primaryKey));
    } : void 0,
    fetchOneWithErrors: isObjectTypeDefinition(objectType) ? async (primaryKey, options) => {
      return await fetchSingleWithErrors(augmentRequestContext(clientCtx, (_) => ({
        finalMethodCall: "fetchOneWithErrors"
      })), objectType, options, await createWithPk(clientCtx, objectType, objectSet, primaryKey));
    } : void 0,
    subscribe: (listener, opts) => {
      const pendingSubscribe = clientCtx.subscribeFn(objectType, objectSet, listener, opts?.properties, opts?.includeRid);
      return {
        unsubscribe: async () => (await pendingSubscribe)()
      };
    },
    withProperties: (clause) => {
      const definitionMap = /* @__PURE__ */ new Map();
      const derivedProperties = {};
      for (const key of Object.keys(clause)) {
        const derivedPropertyDefinition = clause[key](createWithPropertiesObjectSet(objectType, {
          type: "methodInput"
        }, definitionMap, true));
        derivedProperties[key] = definitionMap.get(derivedPropertyDefinition);
      }
      return clientCtx.objectSetFactory(objectType, clientCtx, {
        type: "withProperties",
        derivedProperties,
        objectSet
      });
    },
    narrowToType: (objectTypeDef) => {
      const existingMapping = clientCtx.narrowTypeInterfaceOrObjectMapping[objectTypeDef.apiName];
      !(!existingMapping || existingMapping === objectTypeDef.type) ? process.env.NODE_ENV !== "production" ? invariant5__default.default(false, `${objectTypeDef.apiName} was previously used as an ${existingMapping}, but now used as a ${objectTypeDef.type}.`) : invariant5__default.default(false) : void 0;
      clientCtx.narrowTypeInterfaceOrObjectMapping[objectTypeDef.apiName] = objectTypeDef.type;
      return clientCtx.objectSetFactory(objectTypeDef, clientCtx, {
        type: "asType",
        objectSet,
        entityType: objectTypeDef.apiName
      });
    },
    async *experimental_asyncIterLinks(links) {
      let $nextPageToken;
      do {
        const result = await fetchLinksPage(augmentRequestContext(clientCtx, (_) => ({
          finalMethodCall: "asyncIterLinks"
        })), objectType, objectSet, links);
        $nextPageToken = result.nextPageToken;
        for (const obj of result.data) {
          yield obj;
        }
      } while ($nextPageToken != null);
    },
    $objectSetInternals: {
      def: objectType
    }
  };
  function createSearchAround(link) {
    return () => {
      return clientCtx.objectSetFactory(objectType, clientCtx, objectType.type === "object" ? {
        type: "searchAround",
        objectSet,
        link
      } : {
        type: "interfaceLinkSearchAround",
        objectSet,
        interfaceLink: link
      });
    };
  }
  objectSetDefinitions.set(base, objectSet);
  return base;
}
async function createWithPk(clientCtx, objectType, objectSet, primaryKey) {
  const resolved = await extractObjectOrInterfaceType(clientCtx, objectSet);
  const targetApiName = resolved?.type === "object" ? resolved.apiName : objectType.apiName;
  const objDef = await clientCtx.ontologyProvider.getObjectDefinition(targetApiName);
  const withPk = {
    type: "filter",
    objectSet,
    where: {
      type: "eq",
      field: objDef.primaryKeyApiName,
      value: primaryKey
    }
  };
  return withPk;
}
function isTextQuery(query) {
  return typeof query === "string";
}

// ../../node_modules/.pnpm/@osdk+foundry.ontologies@2.63.0/node_modules/@osdk/foundry.ontologies/build/esm/public/Attachment.js
var _upload = [1, "/v2/ontologies/attachments/upload", 7, "*/*"];
function upload($ctx, ...args) {
  const headerParams = {
    ...args[2],
    "Content-Type": args[2]?.["Content-Type"] ?? args[0].type,
    "Content-Length": args[0].size.toString()
  };
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _upload, args[0], args[1], headerParams);
}
var _read = [0, "/v2/ontologies/attachments/{0}/content", , , "*/*"];
function read($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _read, ...args);
}
var _get = [0, "/v2/ontologies/attachments/{0}"];
function get($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _get, ...args);
}

// src/Client.ts
var additionalContext = Symbol("additionalContext");

// src/public-utils/hydrateAttachmentFromRid.ts
function hydrateAttachmentFromRid(client, rid) {
  return hydrateAttachmentFromRidInternal(client[additionalContext], rid);
}
function hydrateAttachmentFromRidInternal(client, rid) {
  return {
    rid,
    async fetchContents() {
      return read(client, rid);
    },
    async fetchMetadata() {
      const r = await get(client, rid);
      return {
        ...r,
        sizeBytes: Number(r.sizeBytes)
      };
    }
  };
}

// src/object/validateMediaItemMetadata.ts
var _knownVariantKeys = {
  document: true,
  imagery: true,
  audio: true,
  video: true,
  dicom: true,
  email: true,
  model3d: true,
  spreadsheet: true,
  untyped: true
};
var KNOWN_VARIANTS = new Set(Object.keys(_knownVariantKeys));
function validateMediaItemMetadata(raw) {
  if (KNOWN_VARIANTS.has(raw.type)) {
    return raw;
  }
  return {
    type: "unknown",
    raw
  };
}

// src/createMediaFromReference.ts
function createMediaFromReference(client, mediaReference) {
  return createMediaFromReferenceInternal(client[additionalContext], mediaReference);
}
function createMediaFromReferenceInternal(client, mediaReference) {
  const {
    mediaSetRid,
    mediaItemRid
  } = mediaReference.reference.mediaSetViewItem;
  const token = mediaReference.reference.mediaSetViewItem.token;
  return {
    async fetchContents() {
      return chunkROJIXXGF_cjs.MediaSet_exports.read(client, mediaSetRid, mediaItemRid, {
        preview: true
      }, token ? {
        ReadToken: token
      } : void 0);
    },
    async fetchMetadata() {
      const info = await chunkROJIXXGF_cjs.MediaSet_exports.info(client, mediaSetRid, mediaItemRid, {
        preview: true
      }, token ? {
        ReadToken: token
      } : void 0);
      !(info.sizeBytes != null) ? process.env.NODE_ENV !== "production" ? invariant5__default.default(false, "Expected sizeBytes in media info") : invariant5__default.default(false) : void 0;
      !(info.mimeType != null) ? process.env.NODE_ENV !== "production" ? invariant5__default.default(false, "Expected mimeType in media info") : invariant5__default.default(false) : void 0;
      return {
        path: info.path,
        sizeBytes: info.sizeBytes,
        mediaType: info.mimeType
      };
    },
    async fetchFullMetadata() {
      const raw = await chunkROJIXXGF_cjs.MediaSet_exports.metadata(client, mediaSetRid, mediaItemRid, {
        preview: true
      }, token ? {
        ReadToken: token
      } : void 0);
      return {
        itemMetadata: validateMediaItemMetadata(raw)
      };
    },
    getMediaReference() {
      return mediaReference;
    }
  };
}

exports.addUserAgentAndRequestContextHeaders = addUserAgentAndRequestContextHeaders;
exports.additionalContext = additionalContext;
exports.augment = augment;
exports.augmentRequestContext = augmentRequestContext;
exports.createMediaFromReference = createMediaFromReference;
exports.createMediaFromReferenceInternal = createMediaFromReferenceInternal;
exports.createObjectSet = createObjectSet;
exports.createTemporary = createTemporary;
exports.createWithPropertiesObjectSet = createWithPropertiesObjectSet;
exports.extractNamespace = extractNamespace;
exports.extractObjectOrInterfaceType = extractObjectOrInterfaceType;
exports.extractRdpDefinition = extractRdpDefinition;
exports.fetchPage = fetchPage;
exports.fetchSingle = fetchSingle;
exports.fetchSingleWithErrors = fetchSingleWithErrors;
exports.fetchStaticRidPage = fetchStaticRidPage;
exports.getWireObjectSet = getWireObjectSet;
exports.hasWithProperties = hasWithProperties;
exports.hydrateAttachmentFromRid = hydrateAttachmentFromRid;
exports.hydrateAttachmentFromRidInternal = hydrateAttachmentFromRidInternal;
exports.isObjectSet = isObjectSet;
exports.isWireObjectSet = isWireObjectSet;
exports.upload = upload;
exports.validateMediaItemMetadata = validateMediaItemMetadata;
