// @osdk/client/build/cjs/chunk-ROJIXXGF.cjs
'use strict';

var chunkH6PXPU6F_cjs = require('./chunk-H6PXPU6F.cjs');
var chunkQ7SFCCGT_cjs = require('./chunk-Q7SFCCGT.cjs');

// ../../node_modules/.pnpm/@osdk+foundry.mediasets@2.63.0/node_modules/@osdk/foundry.mediasets/build/esm/public/MediaSet.js
var MediaSet_exports = {};
chunkQ7SFCCGT_cjs.__export(MediaSet_exports, {
  abort: () => abort,
  calculate: () => calculate,
  clear: () => clear,
  commit: () => commit,
  create: () => create,
  get: () => get,
  getResult: () => getResult,
  getRidByPath: () => getRidByPath,
  getStatus: () => getStatus,
  info: () => info,
  metadata: () => metadata,
  read: () => read,
  readOriginal: () => readOriginal,
  reference: () => reference,
  register: () => register,
  retrieve: () => retrieve,
  transform: () => transform,
  upload: () => upload,
  uploadMedia: () => uploadMedia
});
var _abort = [1, "/v2/mediasets/{0}/transactions/{1}/abort", 2];
function abort($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _abort, ...args);
}
var _calculate = [0, "/v2/mediasets/{0}/items/{1}/transform/imagery/thumbnail/calculate", 6];
function calculate($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _calculate, ...args);
}
var _clear = [3, "/v2/mediasets/{0}/items/clearAtPath", 2];
function clear($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _clear, ...args);
}
var _commit = [1, "/v2/mediasets/{0}/transactions/{1}/commit", 2];
function commit($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _commit, ...args);
}
var _create = [1, "/v2/mediasets/{0}/transactions", 2];
function create($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _create, ...args);
}
var _info = [0, "/v2/mediasets/{0}/items/{1}", 6];
function info($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _info, ...args);
}
var _metadata = [0, "/v2/mediasets/{0}/items/{1}/metadata", 6];
function metadata($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _metadata, ...args);
}
var _reference = [0, "/v2/mediasets/{0}/items/{1}/reference", 6];
function reference($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _reference, ...args);
}
var _getRidByPath = [0, "/v2/mediasets/{0}/items/getRidByPath", 2];
function getRidByPath($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _getRidByPath, ...args);
}
var _get = [0, "/v2/mediasets/{0}", 2];
function get($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _get, ...args);
}
var _getResult = [0, "/v2/mediasets/{0}/items/{1}/transformationJobs/{2}/result", 6, , "*/*"];
function getResult($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _getResult, ...args);
}
var _getStatus = [0, "/v2/mediasets/{0}/items/{1}/transformationJobs/{2}", 6];
function getStatus($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _getStatus, ...args);
}
var _upload = [1, "/v2/mediasets/{0}/items", 3, "*/*"];
function upload($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _upload, ...args);
}
var _read = [0, "/v2/mediasets/{0}/items/{1}/content", 6, , "*/*"];
function read($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _read, ...args);
}
var _readOriginal = [0, "/v2/mediasets/{0}/items/{1}/original", 6, , "*/*"];
function readOriginal($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _readOriginal, ...args);
}
var _register = [1, "/v2/mediasets/{0}/items/register", 3];
function register($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _register, ...args);
}
var _retrieve = [0, "/v2/mediasets/{0}/items/{1}/transform/imagery/thumbnail/retrieve", 6, , "*/*"];
function retrieve($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _retrieve, ...args);
}
var _transform = [1, "/v2/mediasets/{0}/items/{1}/transform", 7];
function transform($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _transform, ...args);
}
var _uploadMedia = [2, "/v2/mediasets/media/upload", 7, "*/*"];
function uploadMedia($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _uploadMedia, ...args);
}

exports.MediaSet_exports = MediaSet_exports;
