// @osdk/client/build/cjs/chunk-JDRCSS7F.cjs
'use strict';

var chunkF4L4RFUT_cjs = require('./chunk-F4L4RFUT.cjs');
var chunkROJIXXGF_cjs = require('./chunk-ROJIXXGF.cjs');
var chunkH6PXPU6F_cjs = require('./chunk-H6PXPU6F.cjs');
var invariant = require('tiny-invariant');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var invariant__default = /*#__PURE__*/_interopDefault(invariant);

// src/object/AttachmentUpload.ts
function isAttachmentUpload(o) {
  return typeof o === "object" && o != null && "name" in o && typeof o.name === "string" && "data" in o && o.data instanceof Blob && !("fileName" in o);
}
function isAttachmentFile(o) {
  return typeof o === "object" && o instanceof Blob && "name" in o;
}
function createAttachmentUpload(data, name) {
  return {
    data,
    name
  };
}

// src/util/streamutils.ts
var START_TOKEN = new Uint8Array([123, 34, 100, 97, 116, 97, 34, 58, 91]);
var OBJECT_OPEN_CHAR_CODE = 123;
var OBJECT_CLOSE_CHAR_CODE = 125;
async function* parseStreamedResponse(asyncIterable) {
  const utf8decoder = new TextDecoder("utf-8");
  let parsedStart = false;
  let prevChunks = [];
  let openBracesCount = 0;
  for await (const chunk of asyncIterable) {
    let i = 0;
    if (!parsedStart) {
      parsedStart = true;
      if (startsWith(chunk, START_TOKEN)) {
        i = START_TOKEN.length;
      }
    }
    for (; i < chunk.length; i++) {
      while (openBracesCount === 0 && chunk[i] !== OBJECT_OPEN_CHAR_CODE && i < chunk.length) {
        i++;
      }
      let j = i;
      for (; j < chunk.length; j++) {
        const c = chunk[j];
        if (c === OBJECT_OPEN_CHAR_CODE) {
          openBracesCount++;
        } else if (c === OBJECT_CLOSE_CHAR_CODE) {
          openBracesCount--;
          if (0 === openBracesCount) {
            yield combineAndParse(utf8decoder, prevChunks, chunk.subarray(i, j + 1));
            prevChunks = [];
            i = j;
            break;
          }
        }
      }
      if (j === chunk.length) {
        prevChunks.push(chunk.subarray(i));
        break;
      }
    }
  }
}
function startsWith(a, b) {
  if (a.length < b.length) {
    return false;
  }
  for (let i = 0; i < b.length; i++) {
    if (a[i] !== b[i]) {
      return false;
    }
  }
  return true;
}
function combineAndParse(utf8decoder, prev, curr) {
  let str = "";
  for (const chunk of prev) {
    str += utf8decoder.decode(chunk, {
      stream: true
    });
  }
  str += utf8decoder.decode(curr);
  return JSON.parse(str);
}
async function* iterateReadableStream(readableStream) {
  let res = await readableStream.read();
  while (!res.done) {
    yield res.value;
    res = await readableStream.read();
  }
}
async function* parseNdjsonStream(asyncIterable) {
  const decoder = new TextDecoder("utf-8");
  let buffer = "";
  for await (const chunk of asyncIterable) {
    buffer += decoder.decode(chunk, {
      stream: true
    });
    let newlineIdx;
    while ((newlineIdx = buffer.indexOf("\n")) !== -1) {
      const line = buffer.slice(0, newlineIdx);
      buffer = buffer.slice(newlineIdx + 1);
      if (line.length > 0) {
        yield JSON.parse(line);
      }
    }
  }
  buffer += decoder.decode();
  const trailing = buffer.trim();
  if (trailing.length > 0) {
    yield JSON.parse(trailing);
  }
}

// src/util/objectSpecifierUtils.ts
function createObjectSpecifierFromPrimaryKey(objectDef, primaryKey) {
  return `${objectDef.apiName}:${primaryKey}`;
}
function createObjectSpecifierFromInterfaceSpecifier(interfaceDef, interfaceSpecifier) {
  return `${interfaceSpecifier.objectTypeApiName}:${interfaceSpecifier.primaryKeyValue}`;
}
function extractPrimaryKeyFromObjectSpecifier(ObjectSpecifier) {
  return ObjectSpecifier.split(":")[1];
}

// ../../node_modules/.pnpm/@osdk+foundry.ontologies@2.63.0/node_modules/@osdk/foundry.ontologies/build/esm/public/Query.js
var _execute = [1, "/v2/ontologies/{0}/queries/{1}/execute", 7];
function execute($ctx, ...args) {
  return chunkH6PXPU6F_cjs.foundryPlatformFetch($ctx, _execute, ...args);
}

// src/object/mediaUpload.ts
function isMedia(o) {
  return typeof o === "object" && o != null && typeof o.fetchMetadata === "function" && typeof o.fetchContents === "function" && typeof o.getMediaReference === "function";
}
function isMediaReference(o) {
  return typeof o === `object` && typeof o.mimeType === "string" && "reference" in o && typeof o.reference === "object" && o.reference.type === "mediaSetViewItem" && "mediaSetViewItem" in o.reference && typeof o.reference.mediaSetViewItem === "object" && typeof o.reference.mediaSetViewItem.mediaSetRid === "string" && typeof o.reference.mediaSetViewItem.mediaSetViewRid === "string" && typeof o.reference.mediaSetViewItem.mediaItemRid === "string";
}
function isMediaUpload(o) {
  return typeof o === "object" && o != null && "fileName" in o && typeof o.fileName === "string" && "data" in o && typeof o.data === "object" && o.data instanceof Blob;
}

// src/util/interfaceUtils.ts
function getDefType(apiNameOrDef) {
  return typeof apiNameOrDef === "string" ? "object" : apiNameOrDef.type;
}
function isInterfaceActionParam(o) {
  return o != null && typeof o === "object" && "$objectType" in o && "$primaryKey" in o;
}
function isInterfaceQueryParam(o) {
  return o != null && typeof o === "object" && "$objectType" in o && "$primaryKey" in o;
}
function isInterfaceSpecifier(o) {
  return o != null && typeof o === "object" && "$objectType" in o && "$primaryKey" in o && o.$objectType !== o.$apiName;
}

// src/util/isObjectSpecifiersObject.ts
function isObjectSpecifiersObject(o) {
  return o && typeof o === "object" && typeof o.$apiName === "string" && o.$primaryKey != null;
}

// src/util/toDataValueQueries.ts
async function toDataValueQueries(value, client, desiredType) {
  if (value == null) {
    return value;
  }
  if (Array.isArray(value) && desiredType.type === "array") {
    const values = Array.from(value);
    if (values.some((dataValue) => isAttachmentUpload(dataValue) || isAttachmentFile(dataValue))) {
      const converted = [];
      for (const value2 of values) {
        converted.push(await toDataValueQueries(value2, client, desiredType));
      }
      return converted;
    }
    const promiseArray = Array.from(value, async (innerValue) => await toDataValueQueries(innerValue, client, desiredType.array));
    return Promise.all(promiseArray);
  }
  switch (desiredType.type) {
    case "attachment": {
      if (isAttachmentUpload(value)) {
        const attachment = await chunkF4L4RFUT_cjs.upload(client, value.data, {
          filename: value.name
        });
        return attachment.rid;
      }
      if (isAttachmentFile(value)) {
        const attachment = await chunkF4L4RFUT_cjs.upload(client, value, {
          filename: value.name
        });
        return attachment.rid;
      }
      return value;
    }
    case "twoDimensionalAggregation": {
      return {
        groups: value
      };
    }
    case "threeDimensionalAggregation": {
      return {
        groups: value
      };
    }
    case "mediaReference": {
      if (isMediaUpload(value)) {
        const mediaRef = await chunkROJIXXGF_cjs.MediaSet_exports.uploadMedia(client, value.data, {
          filename: value.fileName,
          preview: true
        });
        return mediaRef;
      }
      if (isMedia(value)) {
        return value.getMediaReference();
      }
      if (isMediaReference(value)) {
        return value;
      }
      throw new Error("Expected media reference type but got value that is not a MediaReference or MediaUpload");
    }
    case "set": {
      if (value instanceof Set) {
        const promiseArray = Array.from(value, async (innerValue) => await toDataValueQueries(innerValue, client, desiredType.set));
        return Promise.all(promiseArray);
      }
      break;
    }
    case "object": {
      if (isObjectSpecifiersObject(value)) {
        return value.$primaryKey;
      }
      break;
    }
    case "interface": {
      if (isInterfaceSpecifier(value) || isInterfaceQueryParam(value)) {
        return {
          objectTypeApiName: value.$objectType,
          primaryKeyValue: value.$primaryKey
        };
      }
    }
    case "objectSet":
    case "interfaceObjectSet": {
      if (chunkF4L4RFUT_cjs.isWireObjectSet(value)) {
        return value;
      }
      if (chunkF4L4RFUT_cjs.isObjectSet(value)) {
        return chunkF4L4RFUT_cjs.getWireObjectSet(value);
      }
      break;
    }
    case "map": {
      if (typeof value === "object") {
        const entrySet = [];
        for (const [key, mapValue] of Object.entries(value)) {
          entrySet.push({
            key: desiredType.keyType.type === "object" ? extractPrimaryKeyFromObjectSpecifier(key) : await toDataValueQueries(key, client, desiredType.keyType),
            value: await toDataValueQueries(mapValue, client, desiredType.valueType)
          });
        }
        return entrySet;
      }
      break;
    }
    case "struct": {
      if (typeof value === "object") {
        const structMap = {};
        for (const [key, structValue] of Object.entries(value)) {
          structMap[key] = await toDataValueQueries(structValue, client, desiredType.struct[key]);
        }
        return structMap;
      }
    }
    case "boolean":
    case "date":
    case "double":
    case "float":
    case "integer":
    case "long":
    case "string":
    case "timestamp":
      return value;
  }
  return value;
}

// src/queries/applyQuery.ts
async function applyQuery(client, query, params) {
  const qd = client.ontologyProvider.getQueryDefinition(query.apiName, query.isFixedVersion ? query.version : void 0);
  if (client.flushEdits != null) {
    await client.flushEdits();
  }
  const response = await execute(chunkF4L4RFUT_cjs.addUserAgentAndRequestContextHeaders(chunkF4L4RFUT_cjs.augmentRequestContext(client, (_) => ({
    finalMethodCall: "applyQuery"
  })), query), await client.ontologyRid, query.apiName, {
    parameters: params ? await remapQueryParams(params, client, (await qd).parameters) : {}
  }, {
    version: query.isFixedVersion ? query.version : void 0,
    transactionId: client.transactionId,
    scenarioRid: client.scenarioRid,
    branch: client.branch
  });
  const objectOutputDefs = await getRequiredDefinitions((await qd).output, client);
  const remappedResponse = await remapQueryResponse(client, (await qd).output, response.value, objectOutputDefs);
  return remappedResponse;
}
async function remapQueryParams(params, client, paramTypes) {
  const parameterMap = {};
  for (const [key, value] of Object.entries(params)) {
    parameterMap[key] = await toDataValueQueries(value, client, paramTypes[key]);
  }
  return parameterMap;
}
async function remapQueryResponse(client, responseDataType, responseValue, definitions) {
  if (responseValue == null) {
    if (responseDataType.nullable) {
      return void 0;
    } else {
      throw new Error("Got null response when nullable was not allowed");
    }
  }
  switch (responseDataType.type) {
    case "union": {
      throw new Error("Union return types are not yet supported");
    }
    case "array": {
      for (let i = 0; i < responseValue.length; i++) {
        responseValue[i] = await remapQueryResponse(client, responseDataType.array, responseValue[i], definitions);
      }
      return responseValue;
    }
    case "set": {
      for (let i = 0; i < responseValue.length; i++) {
        responseValue[i] = await remapQueryResponse(client, responseDataType.set, responseValue[i], definitions);
      }
      return responseValue;
    }
    case "attachment": {
      return chunkF4L4RFUT_cjs.hydrateAttachmentFromRidInternal(client, responseValue);
    }
    case "mediaReference": {
      return chunkF4L4RFUT_cjs.createMediaFromReferenceInternal(client, responseValue);
    }
    case "object": {
      const def = definitions.get(responseDataType.object);
      if (!def || def.type !== "object") {
        throw new Error(`Missing definition for ${responseDataType.object}`);
      }
      return createQueryObjectResponse(responseValue, def);
    }
    case "interface": {
      const def = definitions.get(responseDataType.interface);
      if (!def || def.type !== "interface") {
        throw new Error(`Missing definition for ${responseDataType.interface}`);
      }
      return createQueryInterfaceResponse(responseValue, def);
    }
    case "objectSet": {
      const def = definitions.get(responseDataType.objectSet);
      if (!def) {
        throw new Error(`Missing definition for ${responseDataType.objectSet}`);
      }
      if (typeof responseValue === "string") {
        return chunkF4L4RFUT_cjs.createObjectSet(def, client, {
          type: "intersect",
          objectSets: [{
            type: "base",
            objectType: responseDataType.objectSet
          }, {
            type: "reference",
            reference: responseValue
          }]
        });
      }
      return chunkF4L4RFUT_cjs.createObjectSet(def, client, responseValue);
    }
    case "struct": {
      for (const [key, subtype] of Object.entries(responseDataType.struct)) {
        if (requiresConversion(subtype) || responseValue[key] == null) {
          responseValue[key] = await remapQueryResponse(client, subtype, responseValue[key], definitions);
        }
      }
      return responseValue;
    }
    case "map": {
      const map = {};
      !Array.isArray(responseValue) ? process.env.NODE_ENV !== "production" ? invariant__default.default(false, "Expected array entry") : invariant__default.default(false) : void 0;
      for (const entry of responseValue) {
        !(entry.key != null) ? process.env.NODE_ENV !== "production" ? invariant__default.default(false, "Expected key") : invariant__default.default(false) : void 0;
        !(responseDataType.valueType.nullable || entry.value != null) ? process.env.NODE_ENV !== "production" ? invariant__default.default(false, "Expected value") : invariant__default.default(false) : void 0;
        const key = responseDataType.keyType.type === "object" ? getObjectSpecifier(entry.key, responseDataType.keyType.object, definitions) : entry.key;
        const value = await remapQueryResponse(client, responseDataType.valueType, entry.value, definitions);
        map[key] = value;
      }
      return map;
    }
    case "twoDimensionalAggregation": {
      const result = [];
      for (const {
        key,
        value
      } of responseValue.groups) {
        result.push({
          key,
          value
        });
      }
      return result;
    }
    case "threeDimensionalAggregation": {
      const result = [];
      for (const {
        key,
        groups
      } of responseValue.groups) {
        const subResult = [];
        for (const {
          key: subKey,
          value
        } of groups) {
          subResult.push({
            key: subKey,
            value
          });
        }
        result.push({
          key,
          groups: subResult
        });
      }
      return result;
    }
  }
  return responseValue;
}
async function getRequiredDefinitions(dataType, client) {
  const result = /* @__PURE__ */ new Map();
  switch (dataType.type) {
    case "objectSet": {
      const objectDef = await client.ontologyProvider.getObjectDefinition(dataType.objectSet);
      result.set(dataType.objectSet, objectDef);
      break;
    }
    case "interfaceObjectSet": {
      const interfaceDef = await client.ontologyProvider.getInterfaceDefinition(dataType.objectSet);
      result.set(dataType.objectSet, interfaceDef);
      break;
    }
    case "object": {
      const objectDef = await client.ontologyProvider.getObjectDefinition(dataType.object);
      result.set(dataType.object, objectDef);
      break;
    }
    case "interface": {
      const interfaceDef = await client.ontologyProvider.getInterfaceDefinition(dataType.interface);
      result.set(dataType.interface, interfaceDef);
      break;
    }
    case "set": {
      return getRequiredDefinitions(dataType.set, client);
    }
    case "array": {
      return getRequiredDefinitions(dataType.array, client);
    }
    case "map": {
      const types = [dataType.keyType, dataType.valueType];
      const allDefs = await Promise.all(types.map((value) => getRequiredDefinitions(value, client)));
      for (const defs of allDefs) {
        for (const [type, objectDef] of defs) {
          result.set(type, objectDef);
        }
      }
      break;
    }
    case "struct": {
      const structValues = Object.values(dataType.struct);
      const allDefs = await Promise.all(structValues.map((value) => getRequiredDefinitions(value, client)));
      for (const defs of allDefs) {
        for (const [type, objectDef] of defs) {
          result.set(type, objectDef);
        }
      }
      break;
    }
  }
  return result;
}
function requiresConversion(dataType) {
  switch (dataType.type) {
    case "boolean":
    case "date":
    case "double":
    case "float":
    case "integer":
    case "long":
    case "string":
    case "timestamp":
      return false;
    case "union":
      return true;
    case "struct":
      return Object.values(dataType.struct).some(requiresConversion);
    case "set":
      return requiresConversion(dataType.set);
    case "attachment":
    case "mediaReference":
    case "objectSet":
    case "twoDimensionalAggregation":
    case "threeDimensionalAggregation":
    case "object":
      return true;
    default:
      return false;
  }
}
function getObjectSpecifier(primaryKey, objectTypeApiName, definitions) {
  const def = definitions.get(objectTypeApiName);
  if (!def || def.type !== "object") {
    throw new Error(`Missing definition for ${objectTypeApiName}`);
  }
  return createObjectSpecifierFromPrimaryKey(def, primaryKey);
}
function createQueryObjectResponse(primaryKey, objectDef) {
  return {
    $apiName: objectDef.apiName,
    $title: void 0,
    $objectType: objectDef.apiName,
    $primaryKey: primaryKey,
    $objectSpecifier: createObjectSpecifierFromPrimaryKey(objectDef, primaryKey)
  };
}
function createQueryInterfaceResponse(interfaceSpecifier, interfaceDef) {
  return {
    $apiName: interfaceDef.apiName,
    $title: void 0,
    $objectType: interfaceSpecifier.objectTypeApiName,
    $primaryKey: interfaceSpecifier.primaryKeyValue,
    $objectSpecifier: createObjectSpecifierFromInterfaceSpecifier(interfaceDef, interfaceSpecifier)
  };
}

exports.applyQuery = applyQuery;
exports.createAttachmentUpload = createAttachmentUpload;
exports.createObjectSpecifierFromPrimaryKey = createObjectSpecifierFromPrimaryKey;
exports.extractPrimaryKeyFromObjectSpecifier = extractPrimaryKeyFromObjectSpecifier;
exports.getDefType = getDefType;
exports.getRequiredDefinitions = getRequiredDefinitions;
exports.isAttachmentFile = isAttachmentFile;
exports.isAttachmentUpload = isAttachmentUpload;
exports.isInterfaceActionParam = isInterfaceActionParam;
exports.isMedia = isMedia;
exports.isMediaReference = isMediaReference;
exports.isMediaUpload = isMediaUpload;
exports.isObjectSpecifiersObject = isObjectSpecifiersObject;
exports.iterateReadableStream = iterateReadableStream;
exports.parseNdjsonStream = parseNdjsonStream;
exports.parseStreamedResponse = parseStreamedResponse;
exports.remapQueryParams = remapQueryParams;
exports.remapQueryResponse = remapQueryResponse;
