// @osdk/client/build/cjs/index.cjs
'use strict';

var chunkXLX6T4OB_cjs = require('./chunk-XLX6T4OB.cjs');
require('./chunk-YJG67XL4.cjs');
require('./chunk-TQUNF2DZ.cjs');
require('./chunk-PIWGMK3H.cjs');
require('./chunk-CEBSLQQC.cjs');
var chunkJDRCSS7F_cjs = require('./chunk-JDRCSS7F.cjs');
var chunkF4L4RFUT_cjs = require('./chunk-F4L4RFUT.cjs');
require('./chunk-ROJIXXGF.cjs');
require('./chunk-H6PXPU6F.cjs');
require('./chunk-Q7SFCCGT.cjs');
var api = require('@osdk/api');
var shared_net_errors = require('@osdk/shared.net.errors');
var shared_client_impl = require('@osdk/shared.client.impl');
var invariant = require('tiny-invariant');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var invariant__default = /*#__PURE__*/_interopDefault(invariant);

function createPlatformClient(baseUrl, tokenProvider, options = void 0, fetchFn = fetch) {
  return shared_client_impl.createSharedClientContext(baseUrl, tokenProvider, chunkXLX6T4OB_cjs.USER_AGENT, fetchFn);
}
var isoRegex = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d+)?(?:Z|[+-]\d{2}:\d{2})?$/;
var extractDate = (dateTime) => {
  !(dateTime.length < 33) ? process.env.NODE_ENV !== "production" ? invariant__default.default(false, "Invalid date format. Provided input is too long.") : invariant__default.default(false) : void 0;
  !isoRegex.test(dateTime) ? process.env.NODE_ENV !== "production" ? invariant__default.default(false, `Invalid date format. Expected ISO 8601 format, but received ${dateTime}`) : invariant__default.default(false) : void 0;
  return extractDateFromIsoString(dateTime);
};
var extractDateInUTC = (date) => {
  return extractDateFromIsoString(date.toISOString());
};
var extractDateInLocalTime = (date) => {
  return extractDateFromIsoString(generateOffsetUtcString(date));
};
var generateOffsetUtcString = (date) => {
  const offsetMs = date.getTimezoneOffset() * 60 * 1e3;
  return new Date(date.getTime() - offsetMs).toISOString();
};
var extractDateFromIsoString = (dateTime) => {
  return dateTime.split("T")[0];
};

Object.defineProperty(exports, "ActionValidationError", {
  enumerable: true,
  get: function () { return chunkXLX6T4OB_cjs.ActionValidationError; }
});
Object.defineProperty(exports, "createClient", {
  enumerable: true,
  get: function () { return chunkXLX6T4OB_cjs.createClient; }
});
Object.defineProperty(exports, "createAttachmentUpload", {
  enumerable: true,
  get: function () { return chunkJDRCSS7F_cjs.createAttachmentUpload; }
});
Object.defineProperty(exports, "createObjectSpecifierFromPrimaryKey", {
  enumerable: true,
  get: function () { return chunkJDRCSS7F_cjs.createObjectSpecifierFromPrimaryKey; }
});
Object.defineProperty(exports, "extractPrimaryKeyFromObjectSpecifier", {
  enumerable: true,
  get: function () { return chunkJDRCSS7F_cjs.extractPrimaryKeyFromObjectSpecifier; }
});
Object.defineProperty(exports, "getWireObjectSet", {
  enumerable: true,
  get: function () { return chunkF4L4RFUT_cjs.getWireObjectSet; }
});
Object.defineProperty(exports, "isObjectSet", {
  enumerable: true,
  get: function () { return chunkF4L4RFUT_cjs.isObjectSet; }
});
Object.defineProperty(exports, "isOk", {
  enumerable: true,
  get: function () { return api.isOk; }
});
Object.defineProperty(exports, "PalantirApiError", {
  enumerable: true,
  get: function () { return shared_net_errors.PalantirApiError; }
});
exports.createPlatformClient = createPlatformClient;
exports.extractDate = extractDate;
exports.extractDateInLocalTime = extractDateInLocalTime;
exports.extractDateInUTC = extractDateInUTC;
