// @osdk/client/build/cjs/public/internal.cjs
'use strict';

var chunkTQUNF2DZ_cjs = require('../chunk-TQUNF2DZ.cjs');
var chunkF4L4RFUT_cjs = require('../chunk-F4L4RFUT.cjs');
require('../chunk-ROJIXXGF.cjs');
require('../chunk-H6PXPU6F.cjs');
require('../chunk-Q7SFCCGT.cjs');

// src/public-utils/createAndFetchTempObjectSetRid.ts
async function createAndFetchTempObjectSetRid(client, objectSet) {
  const response = await chunkF4L4RFUT_cjs.createTemporary(client, await client[chunkF4L4RFUT_cjs.additionalContext].ontologyRid, {
    objectSet: chunkF4L4RFUT_cjs.getWireObjectSet(objectSet)
  });
  return response.objectSetRid;
}

// src/public-utils/hydrateObjectSetFromObjectRids.ts
function hydrateObjectSetFromObjectRids(client, definition, rids) {
  return chunkF4L4RFUT_cjs.createObjectSet(definition, client[chunkF4L4RFUT_cjs.additionalContext], {
    type: "intersect",
    objectSets: [definition.type === "interface" ? {
      type: "interfaceBase",
      interfaceType: definition.apiName
    } : {
      type: "base",
      objectType: definition.apiName
    }, {
      type: "static",
      objects: asMutableArray(rids)
    }]
  });
}
function asMutableArray(array) {
  return array;
}

// src/public-utils/hydrateObjectSetFromRid.ts
function hydrateObjectSetFromRid(client, definition, rid) {
  return chunkF4L4RFUT_cjs.createObjectSet(definition, client[chunkF4L4RFUT_cjs.additionalContext], {
    type: "intersect",
    objectSets: [definition.type === "interface" ? {
      type: "interfaceBase",
      interfaceType: definition.apiName
    } : {
      type: "base",
      objectType: definition.apiName
    }, {
      type: "reference",
      reference: rid
    }]
  });
}

// src/logger/BrowserLogger.ts
function createStyle({
  color
}) {
  return `color: ${color}; border: 1px solid ${color}; padding: 2px; border-radius: 3px;`;
}
var levelStyles = {
  debug: createStyle({
    color: "LightBlue"
  }),
  error: createStyle({
    color: "red"
  }),
  fatal: createStyle({
    color: "red"
  }),
  info: createStyle({
    color: "green"
  }),
  trace: createStyle({
    color: "gray"
  }),
  warn: createStyle({
    color: "orange"
  })
};
var BrowserLogger = class _BrowserLogger extends chunkTQUNF2DZ_cjs.BaseLogger {
  constructor(bindings = {}, options = {}) {
    super(bindings, {
      ...options,
      level: options.level ?? "error"
    }, _BrowserLogger);
  }
  createLogMethod(name, bindings) {
    const msgs = [`%c${name}%c`];
    const styles = [levelStyles[name], ""];
    if (this.options?.msgPrefix) {
      msgs.push(`%c${this.options.msgPrefix}%c`);
      styles.push("font-style: italic; color: gray", "");
    }
    if (typeof bindings === "object" && "methodName" in bindings) {
      msgs.push(`%c.${bindings.methodName}()%c`);
      styles.push("font-style: italic;color: orchid", "");
    }
    return (...args) => {
      console[name === "fatal" ? "error" : name](msgs.join(" "), ...styles, ...args);
    };
  }
};

Object.defineProperty(exports, "MinimalLogger", {
  enumerable: true,
  get: function () { return chunkTQUNF2DZ_cjs.MinimalLogger; }
});
Object.defineProperty(exports, "createMediaFromReference", {
  enumerable: true,
  get: function () { return chunkF4L4RFUT_cjs.createMediaFromReference; }
});
Object.defineProperty(exports, "hydrateAttachmentFromRid", {
  enumerable: true,
  get: function () { return chunkF4L4RFUT_cjs.hydrateAttachmentFromRid; }
});
exports.BrowserLogger = BrowserLogger;
exports.createAndFetchTempObjectSetRid = createAndFetchTempObjectSetRid;
exports.hydrateObjectSetFromObjectRids = hydrateObjectSetFromObjectRids;
exports.hydrateObjectSetFromRid = hydrateObjectSetFromRid;
