// @osdk/client/build/cjs/public/unstable-do-not-use.cjs
'use strict';

var chunkGOW2XJR7_cjs = require('../chunk-GOW2XJR7.cjs');
var chunkXLX6T4OB_cjs = require('../chunk-XLX6T4OB.cjs');
require('../chunk-YJG67XL4.cjs');
require('../chunk-TQUNF2DZ.cjs');
require('../chunk-PIWGMK3H.cjs');
require('../chunk-CEBSLQQC.cjs');
require('../chunk-JDRCSS7F.cjs');
var chunkF4L4RFUT_cjs = require('../chunk-F4L4RFUT.cjs');
require('../chunk-ROJIXXGF.cjs');
require('../chunk-H6PXPU6F.cjs');
require('../chunk-Q7SFCCGT.cjs');
var unstable = require('@osdk/api/unstable');

// src/public-utils/osdkConfig.ts
function getMetaTagContent(name) {
  const element = document.querySelector(`meta[name="${name}"]`);
  const val = element ? element.getAttribute("content") : null;
  if (val == null) {
    throw new Error(`Missing meta tag: ${name}`);
  }
  return val;
}
function getViteEnvVar(name) {
  const val = undefined[name];
  if (val == null) {
    throw new Error(`Missing environment variable: ${name}`);
  }
  return val;
}
function isProduction() {
  return process.env.NODE_ENV === "production";
}
function getConfigValue(metaTagName, viteEnvVarName) {
  return isProduction() ? getMetaTagContent(metaTagName) : getViteEnvVar(viteEnvVarName);
}
function getOntologyRid(ontologyRid) {
  return isProduction() ? getMetaTagContent("osdk-ontologyRid") : ontologyRid;
}
function getOsdkConfig(ontologyRid) {
  return {
    clientId: getConfigValue("osdk-clientId", "VITE_FOUNDRY_CLIENT_ID"),
    redirectUrl: getConfigValue("osdk-redirectUrl", "VITE_FOUNDRY_REDIRECT_URL"),
    foundryUrl: getConfigValue("osdk-foundryUrl", "VITE_FOUNDRY_API_URL"),
    ontologyRid: getOntologyRid(ontologyRid)
  };
}

// src/scenarios/createScenario.ts
async function createScenario(client) {
  const ctx = client[chunkF4L4RFUT_cjs.additionalContext];
  if (ctx.scenarioRid != null) {
    throw new Error("createScenario: the supplied client already has an active scenario. Scenarios cannot be nested.");
  }
  const response = await chunkXLX6T4OB_cjs.OntologyScenario_exports.createScenario(ctx, await ctx.ontologyRid, ctx.branch != null ? {
    base: {
      type: "branch",
      branch: ctx.branch
    }
  } : {}, {
    preview: true
  });
  return chunkXLX6T4OB_cjs.buildScenarioClient(client, response.scenarioRid);
}

// src/scenarios/withScenario.ts
function withScenario(client, scenarioRid) {
  return chunkXLX6T4OB_cjs.buildScenarioClient(client, scenarioRid);
}

// src/shapes/applyShapeTransformations.ts
function applyShapeTransformations(shape, rawObject) {
  if (rawObject === void 0) {
    return {
      data: void 0,
      dropped: false,
      violations: []
    };
  }
  const primaryKey = rawObject.$primaryKey;
  const transformedProps = {};
  const requireProps = [];
  for (const prop of Object.keys(shape.__props)) {
    const config = shape.__props[prop];
    if (!config) {
      continue;
    }
    const originalValue = rawObject[prop];
    const op = config.nullabilityOp;
    switch (op.type) {
      case "dropIfNull": {
        if (originalValue == null) {
          return {
            data: void 0,
            dropped: true,
            violations: [{
              property: prop,
              primaryKey,
              constraint: "dropIfNull"
            }]
          };
        }
        break;
      }
      case "withDefault": {
        transformedProps[prop] = originalValue ?? op.defaultValue;
        break;
      }
      case "withTransform": {
        try {
          transformedProps[prop] = op.transform(originalValue);
        } catch (e) {
          if (process.env.NODE_ENV !== "production") {
            console.error(`Shape transform error for property "${prop}":`, e);
          }
          return {
            data: void 0,
            dropped: false,
            violations: [{
              property: prop,
              primaryKey,
              constraint: "transformError"
            }]
          };
        }
        break;
      }
      case "require": {
        requireProps.push(prop);
        break;
      }
    }
  }
  const clonedObject = Object.keys(transformedProps).length > 0 ? rawObject.$clone(transformedProps) : rawObject;
  const violations = [];
  for (const prop of requireProps) {
    const value = clonedObject[prop];
    if (value == null) {
      violations.push({
        property: prop,
        primaryKey,
        constraint: "require"
      });
    }
  }
  if (violations.length > 0) {
    return {
      data: void 0,
      dropped: false,
      violations
    };
  }
  return {
    data: clonedObject,
    dropped: false,
    violations: []
  };
}
function applyShapeTransformationsToArray(shape, rawObjects) {
  const results = [];
  let droppedCount = 0;
  const allViolations = [];
  for (const rawObject of rawObjects) {
    const result = applyShapeTransformations(shape, rawObject);
    if (result.dropped) {
      droppedCount++;
    } else if (result.data !== void 0) {
      results.push(result.data);
    }
    allViolations.push(...result.violations);
  }
  return {
    data: results,
    droppedCount,
    violations: allViolations
  };
}
function getLinkQueryOptions(linkDef, sourceObject, pageSize) {
  const options = {};
  if (linkDef.where) {
    options.where = resolveSymbolBindings(linkDef.where, sourceObject.$primaryKey);
  }
  if (linkDef.orderBy && linkDef.orderBy.length > 0) {
    options.orderBy = orderByToMap(linkDef.orderBy);
  }
  if (linkDef.limit !== void 0) {
    options.pageSize = linkDef.limit;
  } else if (pageSize !== void 0) {
    options.pageSize = pageSize;
  }
  return options;
}
async function buildObjectSetFromLinkDefByType(client, sourceType, sourcePrimaryKey, linkDef) {
  const metadata = await client[chunkF4L4RFUT_cjs.additionalContext].ontologyProvider.getObjectDefinition(sourceType.apiName);
  const pkFieldName = metadata.primaryKeyApiName;
  let objectSet = client(sourceType).where({
    [pkFieldName]: sourcePrimaryKey
  });
  for (const segment of linkDef.segments) {
    if (segment.type === "pivotTo") {
      objectSet = objectSet.pivotTo(segment.linkName);
    }
  }
  if (linkDef.setOperations && linkDef.setOperations.length > 0) {
    const otherObjectSets = await Promise.all(linkDef.setOperations.map((setOp) => buildObjectSetFromLinkDefByType(client, sourceType, sourcePrimaryKey, setOp.other)));
    for (let i = 0; i < linkDef.setOperations.length; i++) {
      const setOp = linkDef.setOperations[i];
      const otherObjectSet = otherObjectSets[i];
      switch (setOp.type) {
        case "union":
          objectSet = objectSet.union(otherObjectSet);
          break;
        case "intersect":
          objectSet = objectSet.intersect(otherObjectSet);
          break;
        case "subtract":
          objectSet = objectSet.subtract(otherObjectSet);
          break;
        default: {
          setOp.type;
        }
      }
    }
  }
  return objectSet;
}
function orderByToMap(orderBy) {
  const result = {};
  for (const entry of orderBy) {
    result[entry.property] = entry.direction;
  }
  return result;
}
function resolveSymbolBindings(value, sourcePrimaryKey) {
  if (unstable.isSourcePkSymbol(value)) {
    return sourcePrimaryKey;
  }
  if (Array.isArray(value)) {
    return value.map((item) => resolveSymbolBindings(item, sourcePrimaryKey));
  }
  if (value != null && typeof value === "object") {
    const result = {};
    for (const [key, val] of Object.entries(value)) {
      result[key] = resolveSymbolBindings(val, sourcePrimaryKey);
    }
    return result;
  }
  return value;
}

Object.defineProperty(exports, "createObservableClient", {
  enumerable: true,
  get: function () { return chunkGOW2XJR7_cjs.createObservableClient; }
});
Object.defineProperty(exports, "createClientWithSubscribe", {
  enumerable: true,
  get: function () { return chunkXLX6T4OB_cjs.createClientWithSubscribe; }
});
Object.defineProperty(exports, "createClientWithTransaction", {
  enumerable: true,
  get: function () { return chunkXLX6T4OB_cjs.createClientWithTransaction; }
});
Object.defineProperty(exports, "augment", {
  enumerable: true,
  get: function () { return chunkF4L4RFUT_cjs.augment; }
});
Object.defineProperty(exports, "getWireObjectSet", {
  enumerable: true,
  get: function () { return chunkF4L4RFUT_cjs.getWireObjectSet; }
});
Object.defineProperty(exports, "isObjectSet", {
  enumerable: true,
  get: function () { return chunkF4L4RFUT_cjs.isObjectSet; }
});
exports.applyShapeTransformations = applyShapeTransformations;
exports.applyShapeTransformationsToArray = applyShapeTransformationsToArray;
exports.buildObjectSetFromLinkDefByType = buildObjectSetFromLinkDefByType;
exports.createScenario = createScenario;
exports.getLinkQueryOptions = getLinkQueryOptions;
exports.getMetaTagContent = getMetaTagContent;
exports.getOsdkConfig = getOsdkConfig;
exports.withScenario = withScenario;
