// @osdk/client/build/esm/observable/internal/list/ListCacheKey.js
/*
 * Copyright 2026 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// Index constants for accessing otherKeys array elements
export const TYPE_IDX = 0;
export const API_NAME_IDX = 1;
export const WHERE_IDX = 2;
export const ORDER_BY_IDX = 3;
export const RDP_IDX = 4;
export const INTERSECT_IDX = 5;
export const PIVOT_IDX = 6;
export const RIDS_IDX = 7;
export const SELECT_IDX = 8;
export const LOAD_PROPERTY_SECURITY_IDX = 9;
export const INCLUDE_ALL_BASE_PROPERTIES_IDX = 10;
export const RESOLVE_TO_OBJECT_TYPE_IDX = 11;
