// @osdk/client/build/esm/scenarios/ScenarioClient.js
/*
 * Copyright 2026 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { OntologyScenarios } from "@osdk/foundry.ontologies";
import { additionalContext } from "../Client.js";
import { createClientWithScenario } from "../createClient.js";

/**
 * The set of many-to-many link types modified within a scenario for one source object type. Returned as part of
 * {@link ScenarioEditedEntityTypes.linkTypes}.
 */

/**
 * The object types and link types that have been modified within a scenario. Only many-to-many link type edits are
 * surfaced via `linkTypes`; one-to-many edits surface as object edits on the object type that owns the foreign key
 * property.
 */

/**
 * A page of edited object identifiers within a scenario for a given object type. Returned by
 * {@link EXPERIMENTAL_ScenarioClient.getEditedEntities}. To load full property values, pass the primary keys back
 * through the scenario client, e.g.
 * `scenario(MyObject).where({ $primaryKey: { $in: keys } }).fetchPage()`.
 */

/**
 * A page of directed link instances edited within a scenario for a given source object type and link api name.
 * Returned by {@link EXPERIMENTAL_ScenarioClient.getEditedLinks}. Source and target carry only `$primaryKey` + `$apiName`.
 */

/**
 * A {@link Client} attached to an ontology scenario. All read and write operations performed via this client are
 * scoped to that scenario. `EXPERIMENTAL_ScenarioClient` is a superset of {@link Client}.
 *
 * @beta This is an experimental, unstable feature subject to change.
 */

export function isScenarioClient(value) {
  return value != null && (typeof value === "object" || typeof value === "function") && typeof value.getScenarioReference === "function";
}

/**
 * Shared internal builder used by both {@link withScenario} and {@link createScenario}. Throws if the parent client
 * is already inside a scenario. If the parent client has an active transaction, the transaction is ignored (a warning
 * is logged) and the client is scoped to the scenario instead. Constructs a fresh {@link Client} via
 * `createClientWithScenario` and decorates it with {@link EXPERIMENTAL_ScenarioClient}-only methods.
 *
 * @internal
 */
export function buildScenarioClient(parent, scenarioRid) {
  const ctx = parent[additionalContext];
  if (ctx.transactionId != null) {
    // eslint-disable-next-line no-console
    console.warn("withScenario / createScenario: the supplied client has an active transaction. Ignoring the transaction and scoping to the scenario instead.");
  }
  if (ctx.scenarioRid != null) {
    throw new Error("withScenario / createScenario: the supplied client already has an active scenario. Scenarios cannot be nested.");
  }
  const inner = createClientWithScenario(scenarioRid, ctx.baseUrl, ctx.ontologyRid, ctx.tokenProvider, {
    logger: ctx.logger,
    UNSTABLE_DO_NOT_USE_BRANCH: ctx.branch
  }, ctx.fetch);
  const innerCtx = inner[additionalContext];
  async function getEditedEntities(objectType, options) {
    const ontologyRid = await innerCtx.ontologyRid;
    const response = await OntologyScenarios.listScenarioEditedObjects(innerCtx, ontologyRid, scenarioRid, objectType.apiName, {
      pageSize: options?.pageSize,
      pageToken: options?.pageToken,
      preview: true
    });
    const data = response.data.map(entry => {
      const wire = entry;
      return {
        $apiName: wire.__apiName,
        $primaryKey: wire.__primaryKey
      };
    });
    return {
      data,
      nextPageToken: response.nextPageToken
    };
  }
  async function getEditedLinks(sourceObjectType, linkType, options) {
    const ontologyRid = await innerCtx.ontologyRid;
    const response = await OntologyScenarios.listScenarioEditedLinks(innerCtx, ontologyRid, scenarioRid, sourceObjectType.apiName, linkType, {
      pageSize: options?.pageSize,
      pageToken: options?.pageToken,
      preview: true
    });
    const data = [];
    for (const entry of response.data) {
      const sourceWire = entry.sourceObject;
      if (sourceWire?.__apiName == null || sourceWire.__primaryKey == null) {
        continue;
      }
      const source = {
        $apiName: sourceWire.__apiName,
        $primaryKey: sourceWire.__primaryKey
      };
      for (const linked of entry.linkedObjects) {
        const targetWire = linked.targetObject;
        if (targetWire?.__apiName == null || targetWire.__primaryKey == null) {
          continue;
        }
        const target = {
          $apiName: targetWire.__apiName,
          $primaryKey: targetWire.__primaryKey
        };
        data.push({
          source,
          target,
          linkType: linked.linkType ?? linkType
        });
      }
    }
    return {
      data,
      nextPageToken: response.nextPageToken
    };
  }
  return Object.defineProperties(inner, {
    getScenarioReference: {
      value: () => scenarioRid
    },
    getEditedEntityTypes: {
      value: async function () {
        const ontologyRid = await innerCtx.ontologyRid;
        const response = await OntologyScenarios.listScenarioEditedEntityTypes(innerCtx, ontologyRid, scenarioRid, {
          preview: true
        });
        return {
          objectTypes: response.objectTypes,
          linkTypes: response.linkTypes
        };
      }
    },
    getEditedEntities: {
      value: getEditedEntities
    },
    editedEntitiesAsyncIter: {
      value: async function* (objectType, options) {
        const seen = new Set();
        let pageToken;
        do {
          const page = await getEditedEntities(objectType, {
            pageSize: options?.pageSize,
            pageToken
          });
          for (const obj of page.data) {
            const key = obj.$primaryKey;
            if (seen.has(key)) continue;
            seen.add(key);
            yield obj;
          }
          pageToken = page.nextPageToken;
        } while (pageToken != null);
      }
    },
    getEditedLinkTypes: {
      value: async function (sourceObjectType) {
        const ontologyRid = await innerCtx.ontologyRid;
        const response = await OntologyScenarios.listScenarioEditedLinkTypes(innerCtx, ontologyRid, scenarioRid, sourceObjectType.apiName, {
          preview: true
        });
        return response.data;
      }
    },
    getEditedLinks: {
      value: getEditedLinks
    },
    editedLinksAsyncIter: {
      value: async function* (sourceObjectType, linkType, options) {
        let pageToken;
        do {
          const page = await getEditedLinks(sourceObjectType, linkType, {
            pageSize: options?.pageSize,
            pageToken
          });
          for (const link of page.data) {
            yield link;
          }
          pageToken = page.nextPageToken;
        } while (pageToken != null);
      }
    }
  });
}
