// @osdk/client/build/esm/observable/internal/media/MediaHelper.js
/*
 * Copyright 2025 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as OntologiesV2 from "@osdk/foundry.ontologies";
import { additionalContext } from "../../../Client.js";
import { AbstractHelper } from "../AbstractHelper.js";
import { createBlobMemoryManager } from "./BlobMemoryManager.js";
import { getMediaCacheKey } from "./getMediaCacheKey.js";
import { MediaMetadataQuery } from "./MediaMetadataQuery.js";
export class MediaHelper extends AbstractHelper {
  blobManager = createBlobMemoryManager();
  getCacheKey(mediaOrLocation) {
    return getMediaCacheKey(mediaOrLocation);
  }
  getTypedCacheKey(coords) {
    return this.cacheKeys.get("mediaMetadata", coords.objectType, coords.primaryKey, coords.propertyName);
  }
  getQuery(options) {
    const cacheKey = this.getTypedCacheKey(options.coords);
    return this.store.queries.get(cacheKey, () => {
      const subject = this.store.subjects.get(cacheKey);
      return new MediaMetadataQuery(this.store, subject, options.coords.objectType, options.coords.primaryKey, options.coords.propertyName, cacheKey, options);
    });
  }
  observeMediaMetadata(coords, options, observer) {
    const query = this.getQuery({
      ...options,
      coords
    });
    return this._subscribe(query, options, observer);
  }
  async fetchMetadata(coords, options) {
    const ontologyRid = await this.store.client[additionalContext].ontologyRid;
    const response = await OntologiesV2.MediaReferenceProperties.getMediaMetadata(this.store.client[additionalContext], ontologyRid, coords.objectType, String(coords.primaryKey), coords.propertyName, {
      preview: options?.preview ?? true
    });
    return {
      path: String(response.path),
      sizeBytes: Number(response.sizeBytes),
      mediaType: response.mediaType
    };
  }
  async fetchContent(mediaOrLocation, options) {
    const preview = options?.preview ?? true;
    const baseCacheKey = this.getCacheKey(mediaOrLocation);
    const cacheKey = preview ? `${baseCacheKey}:preview` : baseCacheKey;
    const cached = this.blobManager.get(cacheKey);
    if (cached) {
      return cached;
    }
    let response;
    const coords = this.resolveToCoords(mediaOrLocation);
    if (coords) {
      const ontologyRid = await this.store.client[additionalContext].ontologyRid;
      response = await OntologiesV2.MediaReferenceProperties.getMediaContent(this.store.client[additionalContext], ontologyRid, coords.objectType, String(coords.primaryKey), coords.propertyName, {
        preview
      });
    } else if ("fetchContents" in mediaOrLocation) {
      response = await mediaOrLocation.fetchContents();
      const arrayBuffer = await response.arrayBuffer();
      const contentType = response.headers.get("content-type") || "application/octet-stream";
      const blob = new Blob([arrayBuffer], {
        type: contentType
      });
      this.blobManager.add(baseCacheKey, blob);
      return blob;
    } else {
      throw new Error("Cannot fetch media content: no coordinates or fetchContents");
    }
    const arrayBuffer = await response.arrayBuffer();
    const contentType = response.headers.get("content-type") || "application/octet-stream";
    const blob = new Blob([arrayBuffer], {
      type: contentType
    });
    this.blobManager.add(cacheKey, blob);
    return blob;
  }
  resolveToCoords(source) {
    if ("objectType" in source && "primaryKey" in source && "propertyName" in source) {
      return source;
    }
    if ("getMediaSourceLocation" in source) {
      return source.getMediaSourceLocation?.();
    }
    return undefined;
  }
  getCachedContent(mediaOrLocation, options) {
    const preview = options?.preview ?? true;
    const baseCacheKey = this.getCacheKey(mediaOrLocation);
    const cacheKey = preview ? `${baseCacheKey}:preview` : baseCacheKey;
    return this.blobManager.get(cacheKey);
  }
  getCachedMetadata(coords) {
    const typedCacheKey = this.getTypedCacheKey(coords);
    const query = this.store.queries.peek(typedCacheKey);
    if (query) {
      const entry = this.store.getValue(typedCacheKey);
      return entry?.value;
    }
    return undefined;
  }
  createBlobUrl(mediaOrLocation, options) {
    const preview = options?.preview ?? true;
    const baseCacheKey = this.getCacheKey(mediaOrLocation);
    const cacheKey = preview ? `${baseCacheKey}:preview` : baseCacheKey;
    return this.blobManager.createBlobUrl(cacheKey);
  }
  releaseBlobUrl(mediaOrLocation, options) {
    const preview = options?.preview ?? true;
    const baseCacheKey = this.getCacheKey(mediaOrLocation);
    const cacheKey = preview ? `${baseCacheKey}:preview` : baseCacheKey;
    this.blobManager.releaseBlobUrl(cacheKey);
  }
  clearCache(mediaOrLocation) {
    const cacheKey = this.getCacheKey(mediaOrLocation);
    this.blobManager.remove(cacheKey);
    this.blobManager.remove(`${cacheKey}:preview`);
    if ("objectType" in mediaOrLocation) {
      const typedCacheKey = this.getTypedCacheKey(mediaOrLocation);
      this.store.queries.delete(typedCacheKey);
    }
  }
  clearAll() {
    this.blobManager.clear();
    for (const cacheKey of this.store.queries.keys()) {
      if (cacheKey.type === "mediaMetadata") {
        this.store.queries.delete(cacheKey);
      }
    }
  }
  dispose() {
    this.blobManager.dispose();
    for (const cacheKey of this.store.queries.keys()) {
      if (cacheKey.type === "mediaMetadata") {
        const query = this.store.queries.peek(cacheKey);
        if (query) {
          query.dispose?.();
        }
      }
    }
  }
}
