// @osdk/api/build/cjs/public/unstable.cjs
'use strict';

// src/experimental/createMediaReference.ts
var __EXPERIMENTAL__NOT_SUPPORTED_YET__createMediaReference = {
  name: "__EXPERIMENTAL__NOT_SUPPORTED_YET__createMediaReference",
  type: "experiment",
  version: "2.1.0"
};

// src/experimental/executeStreamingFunction.ts
var __EXPERIMENTAL__NOT_SUPPORTED_YET__executeStreamingFunction = {
  name: "__EXPERIMENTAL__NOT_SUPPORTED_YET__executeStreamingFunction",
  type: "experiment",
  version: "2.19.0"
};

// src/experimental/fetchOneByRid.ts
var __EXPERIMENTAL__NOT_SUPPORTED_YET__fetchOneByRid = {
  name: "__EXPERIMENTAL__NOT_SUPPORTED_YET__fetchOneByRid",
  type: "experiment",
  version: "2.1.0"
};

// src/experimental/fetchPageByRid.ts
var __EXPERIMENTAL__NOT_SUPPORTED_YET__fetchPageByRid = {
  name: "__EXPERIMENTAL__NOT_SUPPORTED_YET__fetchPageByRid",
  type: "experiment",
  version: "2.2.0"
};

// src/experimental/getBulkLinks.ts
var __EXPERIMENTAL__NOT_SUPPORTED_YET__getBulkLinks = {
  name: "__EXPERIMENTAL__NOT_SUPPORTED_YET__getBulkLinks",
  type: "experiment",
  version: "2.0.8"
};

// src/experimental/subscribeToNoTypeObjectSet.ts
var __EXPERIMENTAL__NOT_SUPPORTED_YET__subscribeToNoTypeObjectSet = {
  name: "__EXPERIMENTAL__NOT_SUPPORTED_YET__subscribeToNoTypeObjectSet",
  type: "experiment",
  version: "2.19.0"
};

// src/experimental/transformAndWait.ts
var __EXPERIMENTAL__NOT_SUPPORTED_YET__transformAndWait = {
  name: "__EXPERIMENTAL__NOT_SUPPORTED_YET__transformAndWait",
  type: "experiment",
  version: "2.8.0"
};

// src/object/MediaTransformationErrors.ts
var MediaTransformationError = class extends Error {
  constructor(message) {
    super(message);
    this.name = "MediaTransformationError";
  }
};
var MediaTransformationTimeoutError = class extends MediaTransformationError {
  constructor(jobId) {
    super(`Transformation job ${jobId} timed out polling status`);
    this.name = "MediaTransformationTimeoutError";
  }
};
var MediaTransformationFailedError = class extends MediaTransformationError {
  constructor(jobId) {
    super(`Transformation job ${jobId} failed`);
    this.name = "MediaTransformationFailedError";
  }
};

// src/shapes/ShapeDefinition.ts
var SourcePrimaryKeySymbol = Symbol.for("osdk.query.$sourcePk");
var LegacyShapesSymbol = Symbol.for("osdk.shapes.$primaryKey");
function isSourcePkSymbol(value) {
  return typeof value === "symbol" && (value === SourcePrimaryKeySymbol || value === LegacyShapesSymbol);
}
var ShapeNullabilityError = class extends Error {
  constructor(shape, violations) {
    const props = violations.map((v) => v.property).join(", ");
    const shapeName = shape.__debugName ?? shape.__shapeId;
    super(`Shape "${shapeName}" requires these properties to be non-null: ${props}`);
    this.shape = shape;
    this.violations = violations;
    this.name = "ShapeNullabilityError";
  }
};

// src/shapes/computeShapeId.ts
function sortedStringify(obj) {
  return JSON.stringify(obj, (_, v) => v && typeof v === "object" && !Array.isArray(v) ? Object.fromEntries(Object.entries(v).sort(([a], [b]) => a.localeCompare(b))) : v);
}
function computeShapeId(input) {
  const canonical = canonicalizeShapeInput(input);
  return simpleHash(sortedStringify(canonical));
}
function canonicalizeShapeInput(input) {
  const sortedPropKeys = [...Object.keys(input.props)].sort();
  const canonicalProps = {};
  for (const key of sortedPropKeys) {
    const config = input.props[key];
    const op = config.nullabilityOp;
    canonicalProps[key] = op.type === "withDefault" ? {
      type: op.type,
      defaultValue: op.defaultValue
    } : op.type;
  }
  const canonicalLinks = input.derivedLinks.map((link) => ({
    name: link.name,
    objectSetDef: canonicalizeObjectSetDef(link.objectSetDef),
    targetShapeId: link.targetShape.__shapeId
  }));
  canonicalLinks.sort((a, b) => a.name.localeCompare(b.name));
  return {
    baseType: input.baseTypeApiName,
    props: canonicalProps,
    links: canonicalLinks
  };
}
function canonicalizeObjectSetDef(def) {
  const result = {
    segments: def.segments.map((seg) => ({
      type: seg.type,
      linkName: seg.linkName,
      sourceType: seg.sourceType
    }))
  };
  if (def.where) {
    result.where = JSON.parse(sortedStringify(def.where));
  }
  if (def.orderBy && def.orderBy.length > 0) {
    result.orderBy = def.orderBy.map((o) => ({
      property: o.property,
      direction: o.direction
    }));
  }
  if (def.limit !== void 0) {
    result.limit = def.limit;
  }
  if (def.distinct) {
    result.distinct = true;
  }
  if (def.setOperations && def.setOperations.length > 0) {
    result.setOperations = def.setOperations.map((op) => ({
      type: op.type,
      other: canonicalizeObjectSetDef(op.other)
    }));
  }
  return result;
}
function simpleHash(str) {
  let hash = 5381;
  for (let i = 0; i < str.length; i++) {
    const char = str.charCodeAt(i);
    hash = (hash << 5) + hash ^ char;
  }
  const positiveHash = hash >>> 0;
  return positiveHash.toString(16).padStart(8, "0");
}

// src/shapes/ShapeBuilder.ts
var ShapeBuilderImpl = class _ShapeBuilderImpl {
  #state;
  constructor(state) {
    this.#state = state;
  }
  #addProps(props, type) {
    const newProps = {
      ...this.#state.props
    };
    for (const prop of props) {
      newProps[prop] = {
        nullabilityOp: {
          type
        }
      };
    }
    return newProps;
  }
  select(...props) {
    return new _ShapeBuilderImpl({
      ...this.#state,
      props: this.#addProps(props, "select")
    });
  }
  require(...props) {
    return new _ShapeBuilderImpl({
      ...this.#state,
      props: this.#addProps(props, "require")
    });
  }
  dropIfNull(...props) {
    return new _ShapeBuilderImpl({
      ...this.#state,
      props: this.#addProps(props, "dropIfNull")
    });
  }
  withDefault(prop, defaultValue) {
    return new _ShapeBuilderImpl({
      ...this.#state,
      props: {
        ...this.#state.props,
        [prop]: {
          nullabilityOp: {
            type: "withDefault",
            defaultValue
          }
        }
      }
    });
  }
  withTransform(prop, transform) {
    return new _ShapeBuilderImpl({
      ...this.#state,
      props: {
        ...this.#state.props,
        [prop]: {
          nullabilityOp: {
            type: "withTransform",
            transform
          }
        }
      }
    });
  }
  deriveLink(name, builder) {
    const linkBuilder = createShapeLinkBuilder(this.#state.baseTypeApiName);
    const linkResult = builder(linkBuilder);
    return new _ShapeBuilderImpl({
      ...this.#state,
      props: {
        ...this.#state.props
      },
      derivedLinks: [...this.#state.derivedLinks, {
        name,
        objectSetDef: linkResult.objectSetDef,
        targetShape: linkResult.targetShape,
        config: linkResult.config
      }]
    });
  }
  build() {
    const shapeId = computeShapeId({
      baseTypeApiName: this.#state.baseTypeApiName,
      props: this.#state.props,
      derivedLinks: this.#state.derivedLinks
    });
    return {
      __shapeId: shapeId,
      __debugName: this.#state.debugName,
      __baseType: this.#state.baseType,
      __baseTypeApiName: this.#state.baseTypeApiName,
      __props: Object.freeze({
        ...this.#state.props
      }),
      __derivedLinks: Object.freeze([...this.#state.derivedLinks]),
      __selectedPropsType: {},
      __derivedLinksType: {}
    };
  }
};
var ShapeLinkBuilderImpl = class _ShapeLinkBuilderImpl {
  #state;
  $primaryKey = SourcePrimaryKeySymbol;
  constructor(state) {
    this.#state = state;
  }
  pivotTo(link) {
    const newSegments = [...this.#state.segments, {
      type: "pivotTo",
      linkName: link,
      sourceType: this.#state.currentTypeApiName
      // targetType is resolved at link execution time when link metadata is available
    }];
    return new _ShapeLinkBuilderImpl({
      ...this.#state,
      segments: newSegments,
      // currentTypeApiName is a hint for debugging; actual type resolution happens at execution
      currentTypeApiName: `${this.#state.currentTypeApiName}.${link}`
    });
  }
  /** Sets the where filter for this link. Calling multiple times replaces the previous clause (last-write-wins). */
  where(clause) {
    return new _ShapeLinkBuilderImpl({
      ...this.#state,
      where: clause
    });
  }
  #setOp(type, others) {
    return new _ShapeLinkBuilderImpl({
      ...this.#state,
      setOperations: [...this.#state.setOperations, ...others.map((other) => ({
        type,
        other: other.toObjectSetDef()
      }))]
    });
  }
  union(...others) {
    return this.#setOp("union", others);
  }
  intersect(...others) {
    return this.#setOp("intersect", others);
  }
  subtract(...others) {
    return this.#setOp("subtract", others);
  }
  orderBy(property, direction = "asc") {
    return new _ShapeLinkBuilderImpl({
      ...this.#state,
      orderBy: [...this.#state.orderBy, {
        property,
        direction
      }]
    });
  }
  limit(n) {
    if (!Number.isInteger(n) || n <= 0) {
      throw new Error(`ShapeLinkBuilder.limit() requires a positive integer, got: ${n}`);
    }
    return new _ShapeLinkBuilderImpl({
      ...this.#state,
      limit: n
    });
  }
  distinct() {
    return new _ShapeLinkBuilderImpl({
      ...this.#state,
      distinct: true
    });
  }
  as(shape, config = {}) {
    return {
      targetShape: shape,
      objectSetDef: this.toObjectSetDef(),
      config
    };
  }
  toObjectSetDef() {
    return {
      segments: Object.freeze([...this.#state.segments]),
      where: this.#state.where,
      orderBy: this.#state.orderBy.length > 0 ? Object.freeze([...this.#state.orderBy]) : void 0,
      limit: this.#state.limit,
      distinct: this.#state.distinct || void 0,
      setOperations: this.#state.setOperations.length > 0 ? Object.freeze([...this.#state.setOperations]) : void 0
    };
  }
};
function createShapeLinkBuilder(sourceTypeApiName) {
  return new ShapeLinkBuilderImpl({
    sourceTypeApiName,
    currentTypeApiName: sourceTypeApiName,
    segments: [],
    orderBy: [],
    distinct: false,
    setOperations: []
  });
}
function createShapeBuilder(baseType, debugName) {
  const state = {
    baseType,
    baseTypeApiName: baseType.apiName,
    debugName,
    props: {},
    derivedLinks: []
  };
  return new ShapeBuilderImpl(state);
}

// src/shapes/configToShapeDefinition.ts
function configToShapeDefinition(baseType, config) {
  const baseTypeApiName = baseType.apiName;
  const props = {};
  const seen = /* @__PURE__ */ new Set();
  function addProp(prop, config2) {
    if (seen.has(prop)) {
      throw new Error(`Property "${prop}" appears in multiple config arrays`);
    }
    seen.add(prop);
    props[prop] = config2;
  }
  if (config.require) {
    for (const prop of config.require) {
      addProp(prop, {
        nullabilityOp: {
          type: "require"
        }
      });
    }
  }
  if (config.select) {
    for (const prop of config.select) {
      addProp(prop, {
        nullabilityOp: {
          type: "select"
        }
      });
    }
  }
  if (config.dropIfNull) {
    for (const prop of config.dropIfNull) {
      addProp(prop, {
        nullabilityOp: {
          type: "dropIfNull"
        }
      });
    }
  }
  if (config.defaults) {
    for (const [prop, defaultValue] of Object.entries(config.defaults)) {
      addProp(prop, {
        nullabilityOp: {
          type: "withDefault",
          defaultValue
        }
      });
    }
  }
  if (config.transforms) {
    for (const [prop, transform] of Object.entries(config.transforms)) {
      if (typeof transform !== "function") {
        throw new Error(`transforms["${prop}"] must be a function, got: ${typeof transform}`);
      }
      addProp(prop, {
        nullabilityOp: {
          type: "withTransform",
          transform
        }
      });
    }
  }
  const derivedLinks = [];
  if (config.links) {
    for (const [name, linkConfig] of Object.entries(config.links)) {
      const linkBuilder = createShapeLinkBuilder(baseTypeApiName);
      const traversed = linkConfig.via(linkBuilder);
      const objectSetDef = traversed.toObjectSetDef();
      const targetShape = linkConfig.target;
      derivedLinks.push({
        name,
        objectSetDef,
        targetShape,
        config: {
          defer: linkConfig.defer
        }
      });
    }
  }
  const shapeId = computeShapeId({
    baseTypeApiName,
    props,
    derivedLinks
  });
  return {
    __shapeId: shapeId,
    __debugName: void 0,
    __baseType: baseType,
    __baseTypeApiName: baseTypeApiName,
    __props: Object.freeze({
      ...props
    }),
    __derivedLinks: Object.freeze([...derivedLinks]),
    __selectedPropsType: {},
    __derivedLinksType: {}
  };
}

exports.MediaTransformationError = MediaTransformationError;
exports.MediaTransformationFailedError = MediaTransformationFailedError;
exports.MediaTransformationTimeoutError = MediaTransformationTimeoutError;
exports.ShapeNullabilityError = ShapeNullabilityError;
exports.SourcePrimaryKeySymbol = SourcePrimaryKeySymbol;
exports.__EXPERIMENTAL__NOT_SUPPORTED_YET__createMediaReference = __EXPERIMENTAL__NOT_SUPPORTED_YET__createMediaReference;
exports.__EXPERIMENTAL__NOT_SUPPORTED_YET__executeStreamingFunction = __EXPERIMENTAL__NOT_SUPPORTED_YET__executeStreamingFunction;
exports.__EXPERIMENTAL__NOT_SUPPORTED_YET__fetchOneByRid = __EXPERIMENTAL__NOT_SUPPORTED_YET__fetchOneByRid;
exports.__EXPERIMENTAL__NOT_SUPPORTED_YET__fetchPageByRid = __EXPERIMENTAL__NOT_SUPPORTED_YET__fetchPageByRid;
exports.__EXPERIMENTAL__NOT_SUPPORTED_YET__getBulkLinks = __EXPERIMENTAL__NOT_SUPPORTED_YET__getBulkLinks;
exports.__EXPERIMENTAL__NOT_SUPPORTED_YET__subscribeToNoTypeObjectSet = __EXPERIMENTAL__NOT_SUPPORTED_YET__subscribeToNoTypeObjectSet;
exports.__EXPERIMENTAL__NOT_SUPPORTED_YET__transformAndWait = __EXPERIMENTAL__NOT_SUPPORTED_YET__transformAndWait;
exports.configToShapeDefinition = configToShapeDefinition;
exports.createShapeBuilder = createShapeBuilder;
exports.isSourcePkSymbol = isSourcePkSymbol;
