// @osdk/client/build/esm/createMediaReferenceProperty.js
/*
 * Copyright 2024 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { MediaSets } from "@osdk/foundry.mediasets";
import * as MediaReferenceProperties from "@osdk/foundry.ontologies/MediaReferenceProperty";
import { validateMediaItemMetadata } from "./object/validateMediaItemMetadata.js";
export class MediaReferencePropertyImpl {
  #mediaReference;
  #triplet;
  #client;
  constructor(args) {
    const {
      client,
      objectApiName,
      primaryKey,
      propertyName,
      mediaReference
    } = args;
    this.#client = client;
    this.#triplet = [objectApiName, primaryKey, propertyName];
    this.#mediaReference = mediaReference;
  }
  async fetchContents() {
    return MediaReferenceProperties.getMediaContent(this.#client, await this.#client.ontologyRid, ...this.#triplet, {
      preview: true // TODO: Can turn this back off when backend is no longer in beta.
    });
  }
  async fetchMetadata() {
    const r = await MediaReferenceProperties.getMediaMetadata(this.#client, await this.#client.ontologyRid, ...this.#triplet, {
      preview: true // TODO: Can turn this back off when backend is no longer in beta.
    });
    return {
      path: r.path,
      sizeBytes: Number(r.sizeBytes),
      mediaType: r.mediaType
    };
  }

  /**
   * Fetches type-specific metadata via `MediaSets.metadata`. There is no Ontologies-backed
   * full-metadata endpoint, so this routes through the media set view using the embedded media
   * reference.
   */
  async fetchFullMetadata() {
    const {
      mediaSetRid,
      mediaItemRid,
      token
    } = this.#mediaReference.reference.mediaSetViewItem;
    const raw = await MediaSets.metadata(this.#client, mediaSetRid, mediaItemRid, {
      preview: true
    }, token ? {
      ReadToken: token
    } : undefined);
    return {
      itemMetadata: validateMediaItemMetadata(raw)
    };
  }
  getMediaReference() {
    return this.#mediaReference;
  }
}
