// @osdk/client/build/esm/createClient.js
/*
 * Copyright 2024 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { __EXPERIMENTAL__NOT_SUPPORTED_YET__createMediaReference, __EXPERIMENTAL__NOT_SUPPORTED_YET__executeStreamingFunction, __EXPERIMENTAL__NOT_SUPPORTED_YET__fetchOneByRid, __EXPERIMENTAL__NOT_SUPPORTED_YET__fetchPageByRid, __EXPERIMENTAL__NOT_SUPPORTED_YET__getBulkLinks, __EXPERIMENTAL__NOT_SUPPORTED_YET__subscribeToNoTypeObjectSet, __EXPERIMENTAL__NOT_SUPPORTED_YET__transformAndWait } from "@osdk/api/unstable";
import { symbolClientContext as oldSymbolClientContext } from "@osdk/shared.client";
import { applyAction } from "./actions/applyAction.js";
import { additionalContext } from "./Client.js";
import { createMinimalClient } from "./createMinimalClient.js";
import { fetchMetadataInternal } from "./fetchMetadata.js";
import { makeMediaTransformation } from "./internal/conversions/makeMediaTransformation.js";
import { MinimalLogger } from "./logger/MinimalLogger.js";
import { fetchPage, fetchStaticRidPage } from "./object/fetchPage.js";
import { fetchSingle } from "./object/fetchSingle.js";
import { createObjectSet } from "./objectSet/createObjectSet.js";
import { ObjectSetListenerWebsocket } from "./objectSet/ObjectSetListenerWebsocket.js";
import { applyQuery } from "./queries/applyQuery.js";

// We import it this way to keep compatible with CJS. If we referenced the
// value of `symbolClientContext` directly, then we would have to a dynamic import
// in `createClientInternal` which would make it async and a break.
// Since this is just a string in `@osdk/shared.client2` instead of a symbol,
// we can safely perform this trick.

class ActionInvoker {
  constructor(clientCtx, actionDef) {
    // We type the property as a generic function as binding `applyAction`
    // doesn't return a type thats all that useful anyway
    // The implements covers us for the most part here as this exact type doesn't
    // escape this file
    this.applyAction = applyAction.bind(undefined, clientCtx, actionDef);
    this.batchApplyAction = applyAction.bind(undefined, clientCtx, actionDef);
  }
}
class QueryInvoker {
  constructor(clientCtx, queryDef) {
    this.executeFunction = applyQuery.bind(undefined, clientCtx, queryDef);
  }
}

/** @internal */
export function createClientInternal(objectSetFactory, transactionRid, flushEdits, scenarioRid, subscribeFn, baseUrl, ontologyRid, tokenProvider, options = undefined, fetchFn = fetch) {
  if (typeof ontologyRid === "string") {
    if (!ontologyRid.startsWith("ri.")) {
      throw new Error("Invalid ontology RID");
    }
  } else {
    // eslint-disable-next-line @typescript-eslint/no-floating-promises
    ontologyRid.then(ontologyRid => {
      if (!ontologyRid.startsWith("ri.")) {
        // FIXME this promise is not await so this just shows up as an unhandled promise rejection
        throw new Error("Invalid ontology RID");
      }
    });
  }
  const clientCtx = createMinimalClient({
    ontologyRid
  }, baseUrl, tokenProvider, {
    ...options,
    logger: options?.logger ?? new MinimalLogger(),
    transactionId: transactionRid,
    flushEdits,
    scenarioRid,
    branch: options?.UNSTABLE_DO_NOT_USE_BRANCH,
    subscribeFn
  }, fetchFn, objectSetFactory);
  return createClientFromContext(clientCtx);
}

/**
 * @internal
 */
export function createClientFromContext(clientCtx) {
  const fetchMetadata = fetchMetadataInternal.bind(undefined, clientCtx);
  const client = Object.defineProperties(function (o) {
    if (o.type === "object" || o.type === "interface") {
      return clientCtx.objectSetFactory(o, clientCtx);
    } else if (o.type === "action") {
      return new ActionInvoker(clientCtx, o); // then as any for dealing with the conditional return value
    } else if (o.type === "query") {
      return new QueryInvoker(clientCtx, o);
    } else if (o.type === "experiment") {
      switch (o.name) {
        case __EXPERIMENTAL__NOT_SUPPORTED_YET__executeStreamingFunction.name:
          return {
            async *executeStreamingFunction(query, params) {
              const {
                applyStreamingQuery
              } = await import("./queries/applyStreamingQuery.js");
              yield* applyStreamingQuery(clientCtx, query, params);
            }
          };
        case __EXPERIMENTAL__NOT_SUPPORTED_YET__getBulkLinks.name:
          return {
            async *getBulkLinks(objs, linkTypes) {
              const {
                createBulkLinksAsyncIterFactory
              } = await import("./__unstable/createBulkLinksAsyncIterFactory.js");
              yield* createBulkLinksAsyncIterFactory(clientCtx)(objs, linkTypes);
            }
          };
        case __EXPERIMENTAL__NOT_SUPPORTED_YET__fetchOneByRid.name:
          return {
            fetchOneByRid: async (objectType, rid, options) => {
              return await fetchSingle(clientCtx, objectType, options, createWithRid([rid]));
            }
          };
        case __EXPERIMENTAL__NOT_SUPPORTED_YET__createMediaReference.name:
          return {
            createMediaReference: async args => {
              const {
                data,
                fileName,
                objectType,
                propertyType
              } = args;
              const {
                upload
              } = await import("@osdk/foundry.ontologies/MediaReferenceProperty");
              return await upload(clientCtx, await clientCtx.ontologyRid, objectType.apiName, propertyType, data, {
                mediaItemPath: fileName,
                preview: true
              });
            }
          };
        case __EXPERIMENTAL__NOT_SUPPORTED_YET__fetchPageByRid.name:
          return {
            fetchPageByRid: async (objectOrInterfaceType, rids, options = {}) => {
              return await fetchPage(clientCtx, objectOrInterfaceType, options, createWithRid(rids));
            },
            fetchPageByRidNoType: async (rids, options) => {
              return await fetchStaticRidPage(clientCtx, rids, options ?? {});
            }
          };
        case __EXPERIMENTAL__NOT_SUPPORTED_YET__subscribeToNoTypeObjectSet.name:
          return {
            subscribeToNoTypeObjectSet: (rid, listener, opts) => {
              const unsubscribe = ObjectSetListenerWebsocket.getInstance(clientCtx).subscribeWithoutType({
                type: "reference",
                reference: rid
              }, listener, opts?.includeRid ?? false);
              return {
                unsubscribe
              };
            }
          };
        case __EXPERIMENTAL__NOT_SUPPORTED_YET__transformAndWait.name:
          return {
            transformAndWait: async args => {
              const {
                transformAndWaitInternal
              } = await import("./util/transformAndWaitInternal.js");
              const {
                mediaSetRid,
                mediaItemRid,
                token
              } = args.mediaReference.reference.mediaSetViewItem;
              return transformAndWaitInternal(clientCtx, mediaSetRid, mediaItemRid, makeMediaTransformation(args.transformation), token, args.options);
            }
          };
      }
      throw new Error("not implemented");
    } else {
      throw new Error("not implemented");
    }
  }, {
    [oldSymbolClientContext]: {
      value: clientCtx
    },
    ["__osdkClientContext"]: {
      value: clientCtx
    },
    [additionalContext]: {
      value: clientCtx
    },
    fetchMetadata: {
      value: fetchMetadata
    }
  });
  return client;
}

/**
 * Creates a {@link Client} for interacting with a Foundry Ontology. This is the primary entry point for
 * the OSDK and is typically called once per application during setup. The returned client is then used
 * to load object sets, apply actions, and execute queries against the configured ontology.
 * @param baseUrl - The base URL of the Foundry stack (e.g. `"https://example.palantirfoundry.com"`).
 * @param ontologyRid - The ontology RID to scope the client to. May be provided directly or as a `Promise`
 *   that resolves to the RID. Typically the generated `$ontologyRid` export from your generated SDK is passed here.
 * @param tokenProvider - A function returning a `Promise` that resolves to a bearer token used to authenticate
 *   requests. Typically the OAuth client returned by `createPublicOauthClient` or `createConfidentialOauthClient`
 *   from `@osdk/oauth`, which handles caching and refresh; you can also provide a custom function if you
 *   manage tokens yourself.
 * @param options - Optional client configuration: a custom `logger`, an experimental `UNSTABLE_DO_NOT_USE_BRANCH`
 *   for branch-aware requests, and additional `headers` to include on every request.
 * @param fetchFn - An optional `fetch` implementation to use for all requests. Defaults to the global `fetch`.
 * @example
 * ```ts
 * import { createClient, type Client } from "@osdk/client";
 * import { createPublicOauthClient } from "@osdk/oauth";
 * import { $ontologyRid } from "./generatedNoCheck/index.js";
 *
 * const auth = createPublicOauthClient(
 *   "<your-client-id>",
 *   "https://example.palantirfoundry.com",
 *   `${window.location.origin}/auth/callback`,
 * );
 *
 * export const client: Client = createClient(
 *   "https://example.palantirfoundry.com",
 *   $ontologyRid,
 *   auth,
 * );
 * ```
 * @returns a {@link Client} configured to talk to the given Foundry stack and ontology
 */
export const createClient = createClientInternal.bind(undefined, createObjectSet, undefined, undefined, undefined, undefined);
export const createClientWithTransaction = (transactionRid, flushEdits, ...args) => createClientInternal(createObjectSet, transactionRid, flushEdits, undefined, undefined, ...args);

/**
 * Creates a {@link Client} whose `objectSet.subscribe()` calls are routed through a caller-supplied
 * `subscribeFn` instead of the default WebSocket implementation. This is the subscribe analogue of
 * the `fetchFn` parameter on {@link createClient}: it lets an embedder (e.g. an embedded backend)
 * capture the original subscribe request — `objectType`, the wire `objectSet`, the requested
 * `properties`, and `shouldLoadRids` — rather than the resulting WebSocket traffic.
 *
 * @param subscribeFn - The subscribe implementation to use for all `objectSet.subscribe()` calls.
 * @param args - The remaining {@link createClient} arguments (`baseUrl`, `ontologyRid`,
 *   `tokenProvider`, `options`, `fetchFn`).
 */
export const createClientWithSubscribe = (subscribeFn, ...args) => createClientInternal(createObjectSet, undefined, undefined, undefined, subscribeFn, ...args);

/** @internal */
export const createClientWithScenario = (scenarioRid, ...args) => createClientInternal(createObjectSet, undefined, undefined, scenarioRid, undefined, ...args);
function createWithRid(rids) {
  return {
    type: "static",
    "objects": rids
  };
}
