// @osdk/client/build/browser/createMediaFromReference.js
/*
 * Copyright 2024 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { MediaSets } from "@osdk/foundry.mediasets";
import invariant from "tiny-invariant";
import { additionalContext } from "./Client.js";
import { validateMediaItemMetadata } from "./object/validateMediaItemMetadata.js";
export function createMediaFromReference(client, mediaReference) {
  return createMediaFromReferenceInternal(client[additionalContext], mediaReference);
}
export function createMediaFromReferenceInternal(client, mediaReference) {
  const {
    mediaSetRid,
    mediaItemRid
  } = mediaReference.reference.mediaSetViewItem;
  const token = mediaReference.reference.mediaSetViewItem.token;
  return {
    async fetchContents() {
      return MediaSets.read(client, mediaSetRid, mediaItemRid, {
        preview: true
      }, token ? {
        ReadToken: token
      } : undefined);
    },
    async fetchMetadata() {
      const info = await MediaSets.info(client, mediaSetRid, mediaItemRid, {
        preview: true
      }, token ? {
        ReadToken: token
      } : undefined);
      !(info.sizeBytes != null) ? process.env.NODE_ENV !== "production" ? invariant(false, "Expected sizeBytes in media info") : invariant(false) : void 0;
      !(info.mimeType != null) ? process.env.NODE_ENV !== "production" ? invariant(false, "Expected mimeType in media info") : invariant(false) : void 0;
      return {
        path: info.path,
        sizeBytes: info.sizeBytes,
        mediaType: info.mimeType
      };
    },
    async fetchFullMetadata() {
      const raw = await MediaSets.metadata(client, mediaSetRid, mediaItemRid, {
        preview: true
      }, token ? {
        ReadToken: token
      } : undefined);
      return {
        itemMetadata: validateMediaItemMetadata(raw)
      };
    },
    getMediaReference() {
      return mediaReference;
    }
  };
}
