// @osdk/api/build/esm/experimental/MediaTransformation.js
/*
 * Copyright 2026 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
// ─── Top-level transformation ─────────────────────────────────────────────────
/**
 * A media transformation request. The `$`-prefixed key selects the
 * transformation kind, and each variant requires specific encoding and
 * operation fields.
 *
 * @see {@link https://www.palantir.com/docs/foundry/api/v2/media-sets-v2-resources/media-sets/transform-media-item/ | Transform Media Item}
 * @experimental
 */
/**
 * @experimental
 */
export let MediaTransformation;
/**
 * @experimental
 */
// ─── Encoding types ───────────────────────────────────────────────────────────
/**
 * @experimental
 */
/**
 * @experimental
 */
/**
 * @experimental
 */
/**
 * @experimental
 */
/**
 * @experimental
 */
// ─── Shared sub-types ─────────────────────────────────────────────────────────
/**
 * @experimental
 */
/**
 * @experimental
 */
export let OcrOutputFormat;
/**
 * @experimental
 */
export let OcrLanguageOrScript;
/**
 * @experimental
 */
/**
 * @experimental
 */
/**
 * @experimental
 */
/**
 * @experimental
 */
export let TranscribeOutputFormat;
/**
 * @experimental
 */
/**
 * @experimental
 */
export let VlmPreprocessingConfig;
/**
 * @experimental
 */
// ─── Image operations ─────────────────────────────────────────────────────────
/**
 * @experimental
 */
/**
 * @experimental
 */
export let ImageOperation;
/**
 * @experimental
 */
// ─── Video operations ─────────────────────────────────────────────────────────
/**
 * @experimental
 */
/**
 * @experimental
 */
export let VideoOperation;
/**
 * @experimental
 */
// ─── Audio operations ─────────────────────────────────────────────────────────
/**
 * @experimental
 */
/**
 * @experimental
 */
export let AudioOperation;
/**
 * @experimental
 */
// ─── Other operation types ────────────────────────────────────────────────────
/**
 * @experimental
 */
export let VideoToAudioOperation;
/**
 * @experimental
 */
export let AudioToTextOperation;
/**
 * @experimental
 */
export let EmailToTextOperation;
/**
 * @experimental
 */
export let EmailToAttachmentOperation;
/**
 * @experimental
 */
export let SpreadsheetToTextOperation;
/**
 * @experimental
 */
export let VideoToArchiveOperation;
/**
 * @experimental
 */
export let VideoToTextOperation;
/**
 * @experimental
 */
export let ImageToTextOperation;
/**
 * @experimental
 */
export let VideoToImageOperation;
/**
 * @experimental
 */
export let ImageToDocumentOperation;
/**
 * @experimental
 */
export let DicomToImageOperation;
/**
 * @experimental
 */
export let DocumentToDocumentOperation;
/**
 * @experimental
 */
export let DocumentToImageOperation;
/**
 * @experimental
 */
export let ImageToEmbeddingOperation;
/**
 * @experimental
 */
export let DocumentToTextOperation;
