// @osdk/client/build/browser/observable/internal/CacheKeys.js
/*
 * Copyright 2025 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Trie } from "@wry/trie";
import { DEBUG_CACHE_KEYS, DEBUG_REFCOUNTS } from "../DebugFlags.js";
import { RefCounts } from "./RefCounts.js";

/**
 * Cache key management with canonicalization.
 * - Uses Trie structure for efficient storage and lookup
 */
export class CacheKeys {
  #debugRefCounts;
  #debugCacheKeys;
  #cacheKeys = new Trie(false, keys => {
    const cacheKey = {
      type: keys[0],
      otherKeys: keys.slice(1)
    };
    this.#onCreate?.(cacheKey);
    if (process.env.NODE_ENV !== "production" && this.#debugRefCounts) {
      // eslint-disable-next-line no-console
      console.log(`CacheKeys.onCreate(${cacheKey.type}, ${JSON.stringify(cacheKey.otherKeys)})`);
      this.#finalizationRegistry.register(cacheKey, () => {
        // eslint-disable-next-line no-console
        console.log(`CacheKey Finalization(${cacheKey.type}, ${JSON.stringify(cacheKey.otherKeys)})`);
      });
    }
    return cacheKey;
  });
  #refCounts;

  // we are currently only using this for debug logging and should just remove it in the future if that
  // continues to be true
  #finalizationRegistry;
  #onCreate;
  #onDestroy;
  constructor({
    onCreate,
    onDestroy,
    debug
  }) {
    this.#onCreate = onCreate;
    this.#onDestroy = onDestroy;
    this.#debugRefCounts = debug?.refCounts ?? DEBUG_REFCOUNTS;
    this.#debugCacheKeys = debug?.cacheKeys ?? DEBUG_CACHE_KEYS;
    this.#refCounts = new RefCounts(this.#debugRefCounts ? 15_000 : 60_000, k => this.#cleanupCacheKey(k), this.#debugRefCounts);
    setInterval(() => {
      this.#refCounts.gc();
    }, 1000);
    this.#finalizationRegistry = new FinalizationRegistry(cleanupCallback => {
      try {
        cleanupCallback();
      } catch (e) {
        // eslint-disable-next-line no-console
        console.error("Caught an error while running a finalization callback", e);
      }
    });
  }
  get(type, ...args) {
    const cacheKeyArgs = this.#normalizeArgs(type, args);
    if (process.env.NODE_ENV !== "production" && this.#debugCacheKeys) {
      // eslint-disable-next-line no-console
      console.debug(`CacheKeys.get([${type},
        ${cacheKeyArgs.slice(1).map(x => JSON.stringify(x)).join(", ")}]) - already exists? `, this.#cacheKeys.peekArray(cacheKeyArgs) != null);
    }
    const cacheKey = this.#cacheKeys.lookupArray(cacheKeyArgs);

    // This is an idempotent call that ensures the cache key is registered for
    // cleanup. If already registered, this does nothing.
    this.#refCounts.register(cacheKey);
    return cacheKey;
  }

  /**
   * Look up an existing cache key without creating or registering it.
   * Returns undefined if the key does not exist in the trie.
   */
  peek(type, ...args) {
    return this.#cacheKeys.peekArray(this.#normalizeArgs(type, args));
  }
  #normalizeArgs(type, args) {
    // Normalize trailing undefined values to ensure consistent cache key lookup
    // This makes ("object", "Foo", 1, undefined) equivalent to ("object", "Foo", 1)
    const normalizedArgs = [...args];
    while (normalizedArgs.length > 0 && normalizedArgs[normalizedArgs.length - 1] === undefined) {
      normalizedArgs.pop();
    }
    return [type, ...normalizedArgs];
  }
  retain(cacheKey) {
    this.#refCounts.retain(cacheKey);
  }
  release(cacheKey) {
    this.#refCounts.release(cacheKey);
  }
  #remove(cacheKey) {
    this.#cacheKeys.remove(cacheKey.type, ...cacheKey.otherKeys);
  }

  /**
   * Called after a key is no longer retained and the timeout has elapsed
   * @param key
   */
  #cleanupCacheKey = key => {
    this.#onDestroy?.(key);
    this.#remove(key);
  };
}
