// @osdk/client/build/browser/observable/internal/ObservableClientImpl.js
/*
 * Copyright 2025 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Subscription } from "rxjs";
import { additionalContext } from "../../Client.js";
import { getWireObjectSet, isObjectSet } from "../../objectSet/createObjectSet.js";
import { extractObjectOrInterfaceType } from "../../util/extractObjectOrInterfaceType.js";
import { UnsubscribableWrapper } from "./UnsubscribableWrapper.js";

/**
 * Implementation of the public ObservableClient interface.
 * - Delegates all operations to the Store for consistency
 * - Serves as the entry point for reactive data management
 * - Ensures proper method binding and API exposure
 *
 * @internal
 */
export class ObservableClientImpl {
  #unionCache = new WeakMap();
  #intersectCache = new WeakMap();
  #subtractCache = new WeakMap();
  constructor(store) {
    this.__experimentalStore = store;
    this.applyAction = store.applyAction.bind(store);
    this.validateAction = store.validateAction.bind(store);
  }
  observeObject = (apiName, pk, options, subFn) => {
    return this.__experimentalStore.objects.observe({
      ...options,
      apiName,
      pk
    },
    // cast to cross typed to untyped barrier
    subFn);
  };
  observeList = (options, subFn) => {
    return this.__experimentalStore.lists.observe(options,
    // cast to cross typed to untyped barrier
    subFn);
  };
  observeAggregation(options, subFn) {
    if (options.objectSet) {
      return this.__experimentalStore.aggregations.observeAsync(options, subFn);
    }
    return this.__experimentalStore.aggregations.observe(options, subFn);
  }
  observeFunction = (queryDef, params, options, subFn) => {
    const dependsOn = options.dependsOn?.map(dep => typeof dep === "string" ? dep : dep.apiName);

    // Partition dependsOnObjects into instances vs ObjectSets

    const instances = [];
    const objectSetWires = [];
    for (const item of options.dependsOnObjects ?? []) {
      if (isObjectSet(item)) {
        objectSetWires.push(getWireObjectSet(item));
      } else {
        instances.push({
          $apiName: item.$objectType ?? item.$apiName,
          $primaryKey: item.$primaryKey
        });
      }
    }

    // Start async extraction of ObjectSet types
    const objectSetTypesPromise = objectSetWires.length > 0 ? Promise.all(objectSetWires.map(wire => extractObjectOrInterfaceType(this.__experimentalStore.client[additionalContext], wire))).then(types => types.filter(t => t != null).map(t => t.apiName)) : undefined;
    return this.__experimentalStore.functions.observe({
      ...options,
      queryDef,
      params,
      dependsOn,
      dependsOnObjects: instances,
      objectSetTypesPromise
    }, subFn);
  };
  observeLinks = (objects, linkName, options, subFn) => {
    const objectsArray = Array.isArray(objects) ? objects : [objects];
    const observer = subFn;
    return objectsArray.length <= 1 ? observeSingleLink(this.__experimentalStore, objectsArray, linkName, options, observer) : observeMultiLinks(this.__experimentalStore, objectsArray, linkName, options, observer);
  };
  observeObjectSet(baseObjectSet, options, subFn) {
    return this.__experimentalStore.objectSets.observe({
      baseObjectSet,
      ...options
    },
    // cast to cross typed to untyped barrier
    subFn);
  }
  invalidateAll() {
    return this.__experimentalStore.invalidateAll();
  }
  invalidateObjects(objects) {
    return this.__experimentalStore.invalidateObjects(objects);
  }
  invalidateObjectType(type) {
    return this.__experimentalStore.invalidateObjectType(type, undefined);
  }
  invalidateFunction(apiName, params) {
    return this.__experimentalStore.invalidateFunction(apiName, params);
  }
  invalidateFunctionsByObject(apiName, primaryKey) {
    return this.__experimentalStore.invalidateFunctionsByObject(apiName, primaryKey);
  }
  canonicalizeWhereClause(where) {
    return this.__experimentalStore.whereCanonicalizer.canonicalize(where);
  }
  canonicalizeOptions(options) {
    const store = this.__experimentalStore;
    const result = {
      ...options
    };
    result.where = store.whereCanonicalizer.canonicalize(result.where);
    result.withProperties = store.rdpCanonicalizer.canonicalize(result.withProperties);
    result.orderBy = store.orderByCanonicalizer.canonicalize(result.orderBy);
    result.aggregate = store.genericCanonicalizer.canonicalize(result.aggregate);
    result.intersectWith = store.genericCanonicalizer.canonicalize(result.intersectWith);
    result.$select = store.selectCanonicalizer.canonicalize(result.$select);
    result.union = this.#canonObjectSetArray(result.union, store.objectSetArrayCanonicalizer.canonicalizeUnion.bind(store.objectSetArrayCanonicalizer), this.#unionCache);
    result.intersect = this.#canonObjectSetArray(result.intersect, store.objectSetArrayCanonicalizer.canonicalizeIntersect.bind(store.objectSetArrayCanonicalizer), this.#intersectCache);
    result.subtract = this.#canonObjectSetArray(result.subtract, store.objectSetArrayCanonicalizer.canonicalizeSubtract.bind(store.objectSetArrayCanonicalizer), this.#subtractCache);
    return result;
  }
  #canonObjectSetArray(arr, canonicalize, cache) {
    if (!arr || arr.length === 0) {
      return arr;
    }
    const wireStrings = arr.map(os => JSON.stringify(getWireObjectSet(os)));
    const canonKey = canonicalize(wireStrings);
    let cached = cache.get(canonKey);
    if (!cached) {
      cached = arr;
      cache.set(canonKey, cached);
    }
    return cached;
  }
  observeMediaMetadata(coords, options, observer) {
    return this.__experimentalStore.media.observeMediaMetadata(coords, options, observer);
  }
  async getCacheSnapshot() {
    return this.__experimentalStore.getCacheSnapshot();
  }
}
function observeSingleLink(store, objectsArray, linkName, options, observer) {
  if (objectsArray.length === 0) {
    observer.next({
      resolvedList: [],
      linkedObjectsBySourcePrimaryKey: new Map(),
      isOptimistic: false,
      lastUpdated: 0,
      fetchMore: async () => {},
      hasMore: false,
      status: "loaded",
      totalCount: "0"
    });
    return new UnsubscribableWrapper(new Subscription());
  }
  const parentSub = new Subscription();
  for (const obj of objectsArray) {
    const pk = obj.$primaryKey;
    const sourceType = obj.$apiName === obj.$objectType ? "object" : "interface";
    parentSub.add(store.links.observe({
      ...options,
      srcType: {
        type: sourceType,
        apiName: obj.$apiName
      },
      sourceUnderlyingObjectType: obj.$objectType,
      linkName,
      pk
    }, observer));
  }
  return new UnsubscribableWrapper(parentSub);
}
function observeMultiLinks(store, objectsArray, linkName, options, observer) {
  const parentSub = new Subscription();
  const totalExpected = objectsArray.length;
  const perObjectData = new Map();
  let errored = false;
  function mergeAndEmit() {
    if (errored) {
      return;
    }
    const seen = new Map();
    const linkedObjectsBySourcePrimaryKey = new Map();
    const fetchMores = [];
    let latestUpdated = 0;
    let hasMore = false;
    let isOptimistic = false;
    for (const {
      payload,
      pk
    } of perObjectData.values()) {
      linkedObjectsBySourcePrimaryKey.set(pk, payload.resolvedList ?? []);
      for (const obj of payload.resolvedList ?? []) {
        seen.set(`${obj.$objectType}:${obj.$primaryKey}`, obj);
      }
      if (payload.lastUpdated > latestUpdated) {
        latestUpdated = payload.lastUpdated;
      }
      if (payload.isOptimistic) {
        isOptimistic = true;
      }
      if (payload.hasMore) {
        hasMore = true;
        fetchMores.push(payload.fetchMore);
      }
    }
    const payloads = [...perObjectData.values()].map(d => d.payload);
    const loading = perObjectData.size < totalExpected || payloads.some(p => p.status === "init" || p.status === "loading");
    observer.next({
      resolvedList: Array.from(seen.values()),
      linkedObjectsBySourcePrimaryKey,
      isOptimistic,
      lastUpdated: latestUpdated,
      fetchMore: hasMore ? () => Promise.all(fetchMores.map(fn => fn())).then(() => {}) : async () => {},
      hasMore,
      status: loading ? "loading" : payloads.some(p => p.status === "error") ? "error" : "loaded",
      ...(!hasMore ? {
        totalCount: String(seen.size)
      } : {})
    });
  }
  for (const obj of objectsArray) {
    const objKey = `${obj.$objectType ?? obj.$apiName}:${obj.$primaryKey}`;
    const pk = obj.$primaryKey;
    const sourceType = obj.$apiName === obj.$objectType ? "object" : "interface";
    parentSub.add(store.links.observe({
      ...options,
      srcType: {
        type: sourceType,
        apiName: obj.$apiName
      },
      sourceUnderlyingObjectType: obj.$objectType,
      linkName,
      pk
    }, {
      next: payload => {
        if (errored) {
          return;
        }
        perObjectData.set(objKey, {
          payload,
          pk
        });
        mergeAndEmit();
      },
      error: err => {
        if (errored) {
          return;
        }
        errored = true;
        parentSub.unsubscribe();
        observer.error(err);
      },
      // store link queries are long-lived and do not complete
      complete: () => {}
    }));
  }
  return new UnsubscribableWrapper(parentSub);
}
