// @osdk/client/build/browser/observable/internal/objectset/ObjectSetQuery.js
/*
 * Copyright 2026 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { additionalContext } from "../../../Client.js";
import { getWireObjectSet } from "../../../objectSet/createObjectSet.js";
import { extractRdpDefinition } from "../../../util/extractRdpDefinition.js";
import { BaseListQuery } from "../base-list/BaseListQuery.js";
import { DEBUG_ONLY__changesToString } from "../Changes.js";
import { getObjectTypesThatInvalidate } from "../getObjectTypesThatInvalidate.js";
import { API_NAME_IDX as OBJECT_API_NAME_IDX } from "../object/ObjectCacheKey.js";
import { objectSortaMatchesWhereClause as objectMatchesWhereClause } from "../objectMatchesWhereClause.js";
import { OrderBySortingStrategy } from "../sorting/SortingStrategy.js";
export class ObjectSetQuery extends BaseListQuery {
  #baseObjectSetWire;
  #operations;
  #composedObjectSet;
  #objectTypes;
  #requiresServerEvaluation;
  #resultTypeApiName;

  // Object types this query's RDPs traverse; an edit to any of these triggers
  // revalidation. Lazily populated on first fetch when `withProperties` is set.
  #rdpInvalidationSet;
  constructor(store, subject, baseObjectSetWire, operations, cacheKey, opts) {
    super(store, subject, opts, cacheKey, process.env.NODE_ENV !== "production" ? store.client[additionalContext].logger?.child({}, {
      msgPrefix: `ObjectSetQuery<${cacheKey.otherKeys.map(x => JSON.stringify(x)).join(", ")}>`
    }) : undefined);
    this.#baseObjectSetWire = baseObjectSetWire;
    this.#operations = operations;
    this.#composedObjectSet = this.#composeObjectSet(opts);
    const baseWire = JSON.parse(baseObjectSetWire);
    this.#objectTypes = this.#extractObjectTypes(baseWire, opts);
    this.#requiresServerEvaluation = !!(operations.pivotTo || operations.union && operations.union.length > 0 || operations.intersect && operations.intersect.length > 0 || operations.subtract && operations.subtract.length > 0);
    this.#resultTypeApiName = ObjectSetQuery.#extractTypeFromWireObjectSet(baseWire) ?? "";
    if (opts.autoFetchMore === true) {
      this.minResultsToLoad = Number.MAX_SAFE_INTEGER;
    } else if (typeof opts.autoFetchMore === "number") {
      this.minResultsToLoad = Math.max(0, opts.autoFetchMore);
    } else {
      this.minResultsToLoad = opts.pageSize || 0;
    }
  }
  get objectTypes() {
    return this.#objectTypes;
  }
  get rdpConfig() {
    return this.#operations.withProperties;
  }
  get selectFields() {
    return this.#operations.select;
  }
  get rawSelect() {
    return this.#operations.select;
  }
  #composeObjectSet(opts) {
    let result = opts.baseObjectSet;
    if (opts.withProperties) {
      result = result.withProperties(opts.withProperties);
    }
    if (opts.where) {
      result = result.where(opts.where);
    }
    if (opts.union && opts.union.length > 0) {
      result = result.union(...opts.union);
    }
    if (opts.intersect && opts.intersect.length > 0) {
      result = result.intersect(...opts.intersect);
    }
    if (opts.subtract && opts.subtract.length > 0) {
      result = result.subtract(...opts.subtract);
    }
    if (opts.pivotTo) {
      result = result.pivotTo(opts.pivotTo);
    }
    return result;
  }
  #extractObjectTypes(baseWire, opts) {
    const types = new Set();
    const baseTypeName = ObjectSetQuery.#extractTypeFromWireObjectSet(baseWire);
    if (baseTypeName) {
      types.add(baseTypeName);
    }
    ObjectSetQuery.#addTypesFromObjectSets(opts.union, types);
    ObjectSetQuery.#addTypesFromObjectSets(opts.intersect, types);
    ObjectSetQuery.#addTypesFromObjectSets(opts.subtract, types);
    return types;
  }
  static #addTypesFromObjectSets(sets, types) {
    if (!sets) {
      return;
    }
    for (const os of sets) {
      const typeName = ObjectSetQuery.#extractTypeFromWireObjectSet(getWireObjectSet(os));
      if (typeName) {
        types.add(typeName);
      }
    }
  }
  static #extractTypeFromWireObjectSet(wire) {
    if (wire.type === "base") {
      return wire.objectType;
    }
    if (wire.type === "interfaceBase") {
      return wire.interfaceType;
    }
    return undefined;
  }

  /**
   * Register changes to the cache specific to ObjectSetQuery
   */
  registerCacheChanges(batch) {
    batch.changes.registerObjectSet(this.cacheKey);
  }

  /**
   * Implements fetchPageData from BaseListQuery template method
   * Fetches a page of data from the composed ObjectSet
   */
  async fetchPageData(signal) {
    if (this.#operations.orderBy && Object.keys(this.#operations.orderBy).length > 0 && !(this.sortingStrategy instanceof OrderBySortingStrategy)) {
      const wireObjectSet = getWireObjectSet(this.#composedObjectSet);
      const {
        resultType,
        invalidationSet
      } = await getObjectTypesThatInvalidate(this.store.client[additionalContext], wireObjectSet);
      this.sortingStrategy = new OrderBySortingStrategy(resultType.apiName, this.#operations.orderBy, await extractRdpDefinition(this.store.client[additionalContext], wireObjectSet));
      this.#rdpInvalidationSet = invalidationSet;
    }
    if (this.#rdpInvalidationSet == null && this.#operations.withProperties != null) {
      const wireObjectSet = getWireObjectSet(this.#composedObjectSet);
      this.#rdpInvalidationSet = await this.#computeInvalidationTypes(wireObjectSet);
    }

    // Fetch the data with pagination
    const resp = await this.#composedObjectSet.fetchPage({
      $nextPageToken: this.nextPageToken,
      $pageSize: this.getEffectiveFetchPageSize(),
      $includeRid: true,
      ...(this.#operations.select && this.#operations.select.length > 0 ? {
        $select: this.#operations.select
      } : {}),
      // OrderBy is already applied in the composed ObjectSet
      ...(this.#operations.orderBy && Object.keys(this.#operations.orderBy).length > 0 ? {
        $orderBy: this.#operations.orderBy
      } : {}),
      ...(this.options.$loadPropertySecurityMetadata ? {
        $loadPropertySecurityMetadata: true
      } : {})
    });
    if (signal?.aborted) {
      throw new Error("Aborted");
    }
    this.nextPageToken = resp.nextPageToken;
    return resp;
  }
  handleFetchError(error, _status, batch) {
    this.logger?.error("error", error);
    this.store.subjects.get(this.cacheKey).error(error);
    const existingTotalCount = batch.read(this.cacheKey)?.value?.totalCount;
    return this.writeToStore({
      data: [],
      totalCount: existingTotalCount
    }, "error", batch);
  }
  registerStreamUpdates(sub) {
    this.createWebsocketSubscription(this.#composedObjectSet, sub, "observeObjectSet");
  }
  maybeUpdateAndRevalidate = (changes, optimisticId) => {
    if (process.env.NODE_ENV !== "production") {
      this.logger?.child({
        methodName: "maybeUpdateAndRevalidate"
      }).debug(DEBUG_ONLY__changesToString(changes));
      this.logger?.child({
        methodName: "maybeUpdateAndRevalidate"
      }).debug(`Already in changes? ${changes.modified.has(this.cacheKey)}`);
    }
    if (changes.modified.has(this.cacheKey)) {
      return;
    }
    changes.modified.add(this.cacheKey);
    try {
      if (this.#requiresServerEvaluation) {
        return this.#handleServerRevalidation(changes);
      }
      return this.#handleLocalUpdate(changes, optimisticId);
    } finally {
      if (process.env.NODE_ENV !== "production") {
        this.logger?.child({
          methodName: "maybeUpdateAndRevalidate"
        }).debug("in finally");
      }
    }
  };
  #handleServerRevalidation(changes) {
    for (const objectType of this.#objectTypes) {
      const added = changes.addedObjects.get(objectType);
      const modified = changes.modifiedObjects.get(objectType);
      if (added && added.length > 0 || modified && modified.length > 0) {
        return this.revalidate(true);
      }
    }
    for (const deletedKey of changes.deleted) {
      if (deletedKey.type === "object" && this.#objectTypes.has(deletedKey.otherKeys[OBJECT_API_NAME_IDX])) {
        return this.revalidate(true);
      }
    }
    return undefined;
  }
  #getRelevantChanges(changes) {
    const resultApiName = this.#resultTypeApiName;
    const addedObjects = changes.addedObjects.get(resultApiName) ?? [];
    const modifiedObjects = changes.modifiedObjects.get(resultApiName) ?? [];
    let hasRelevantDeletions = false;
    for (const key of changes.deleted) {
      if (key.type === "object" && key.otherKeys[OBJECT_API_NAME_IDX] === resultApiName) {
        hasRelevantDeletions = true;
        break;
      }
    }
    if (addedObjects.length === 0 && modifiedObjects.length === 0 && !hasRelevantDeletions) {
      return undefined;
    }
    return {
      addedObjects,
      modifiedObjects
    };
  }
  #handleLocalUpdate(changes, optimisticId) {
    const whereClause = this.#operations.where;
    const effectiveWhere = whereClause ?? this.store.whereCanonicalizer.canonicalize({
      $and: []
    });
    const relevant = this.#getRelevantChanges(changes);
    if (!relevant) {
      return undefined;
    }
    const addedMatches = this.#classifyByWhereMatch(relevant.addedObjects, effectiveWhere);
    const modifiedMatches = this.#classifyByWhereMatch(relevant.modifiedObjects, effectiveWhere);
    const status = optimisticId || addedMatches.uncertain.size > 0 || modifiedMatches.uncertain.size > 0 ? "loading" : "loaded";
    const {
      retVal: needsRevalidation
    } = this.store.batch({
      optimisticId,
      changes
    }, batch => {
      const existingKeys = new Set(batch.read(this.cacheKey)?.value?.data);
      const {
        newList,
        needsRevalidation
      } = reconcileListChanges(existingKeys, addedMatches.definite, relevant.modifiedObjects, modifiedMatches, changes.deleted, batch.optimisticWrite, obj => this.#getObjectCacheKey(obj));
      const existingTotalCount = batch.read(this.cacheKey)?.value?.totalCount;
      this._updateList(newList, status, batch, {
        type: "clientOrdered"
      }, existingTotalCount);
      return needsRevalidation;
    });
    if (needsRevalidation) {
      return this.revalidate(true);
    }
    return undefined;
  }
  #classifyByWhereMatch(objects, whereClause) {
    const definite = new Set();
    const uncertain = new Set();
    for (const obj of objects) {
      if (objectMatchesWhereClause(obj, whereClause, true)) {
        definite.add(obj);
      } else if (objectMatchesWhereClause(obj, whereClause, false)) {
        uncertain.add(obj);
      }
    }
    return {
      definite,
      uncertain
    };
  }
  async #computeInvalidationTypes(wireObjectSet) {
    try {
      const {
        invalidationSet
      } = await getObjectTypesThatInvalidate(this.store.client[additionalContext], wireObjectSet);
      return invalidationSet;
    } catch (error) {
      this.store.logger?.error("Failed to compute invalidation types for object set query, falling back to empty set", error);
      return new Set();
    }
  }
  #getObjectCacheKey(obj) {
    const pk = obj.$primaryKey;
    return this.cacheKeys.get("object", obj.$objectType, pk, this.rdpConfig ?? undefined);
  }
  invalidateObjectType = async (objectType, changes) => {
    if (this.#objectTypes.has(objectType) || (this.#rdpInvalidationSet?.has(objectType) ?? false)) {
      changes?.modified.add(this.cacheKey);
      return this.revalidate(true);
    }
    return Promise.resolve();
  };
  createPayload(params) {
    return {
      resolvedList: params.resolvedData,
      isOptimistic: params.isOptimistic,
      fetchMore: this.fetchMore,
      hasMore: this.nextPageToken != null,
      status: params.status,
      lastUpdated: params.lastUpdated,
      objectSet: this.#composedObjectSet,
      totalCount: params.totalCount
    };
  }
}
function reconcileListChanges(existingKeys, addedDefiniteMatches, modifiedObjects, modifiedMatches, deleted, isOptimistic, getObjectCacheKey) {
  const objectsToInsert = new Set(addedDefiniteMatches);
  const keysToRemove = new Set(deleted);
  let needsRevalidation = false;
  for (const obj of modifiedObjects) {
    if (modifiedMatches.definite.has(obj)) {
      if (!existingKeys.has(getObjectCacheKey(obj))) {
        objectsToInsert.add(obj);
      }
    } else if (!isOptimistic) {
      keysToRemove.add(getObjectCacheKey(obj));
      if (modifiedMatches.uncertain.has(obj)) {
        needsRevalidation = true;
      }
    }
  }
  const newList = [];
  for (const key of existingKeys) {
    if (!keysToRemove.has(key)) {
      newList.push(key);
    }
  }
  for (const obj of objectsToInsert) {
    newList.push(getObjectCacheKey(obj));
  }
  return {
    newList,
    needsRevalidation
  };
}
