// @osdk/client/build/browser/observable/internal/objectset/ObjectSetHelper.js
/*
 * Copyright 2025 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { getWireObjectSet } from "../../../objectSet/createObjectSet.js";
import { hasWithProperties } from "../../../util/extractRdpDefinition.js";
import { AbstractHelper } from "../AbstractHelper.js";
import { ObjectSetQuery } from "./ObjectSetQuery.js";
export class ObjectSetHelper extends AbstractHelper {
  constructor(store, cacheKeys, whereCanonicalizer, orderByCanonicalizer, rdpCanonicalizer, selectCanonicalizer, objectSetArrayCanonicalizer) {
    super(store, cacheKeys);
    this.whereCanonicalizer = whereCanonicalizer;
    this.orderByCanonicalizer = orderByCanonicalizer;
    this.rdpCanonicalizer = rdpCanonicalizer;
    this.selectCanonicalizer = selectCanonicalizer;
    this.objectSetArrayCanonicalizer = objectSetArrayCanonicalizer;
  }
  observe(options, subFn) {
    const ret = super.observe(options, subFn);
    if (options.streamUpdates) {
      if (options.pivotTo) {
        if (process.env.NODE_ENV !== "production") {
          // eslint-disable-next-line no-console
          console.warn("[@osdk/client] streamUpdates is not supported with pivotTo. " + "The server does not support websocket subscriptions for " + "link-traversal queries. Ignoring streamUpdates.");
        }
      } else if (options.withProperties || hasWithProperties(getWireObjectSet(options.baseObjectSet))) {
        if (process.env.NODE_ENV !== "production") {
          // eslint-disable-next-line no-console
          console.warn("[@osdk/client] streamUpdates is not supported with withProperties. " + "The server does not support websocket subscriptions for " + "object sets that include derived properties. Ignoring streamUpdates.");
        }
      } else {
        ret.query.registerStreamUpdates(ret.subscription);
      }
    }
    return ret;
  }
  getQuery(options) {
    const {
      baseObjectSet
    } = options;
    const baseObjectSetWire = JSON.stringify(getWireObjectSet(baseObjectSet));
    const operations = this.buildCanonicalizedOperations(options);
    const objectSetCacheKey = this.cacheKeys.get("objectSet", baseObjectSetWire, operations);
    return this.store.queries.get(objectSetCacheKey, () => {
      return new ObjectSetQuery(this.store, this.store.subjects.get(objectSetCacheKey), baseObjectSetWire, operations, objectSetCacheKey, options);
    });
  }
  buildCanonicalizedOperations(options) {
    const operations = {};
    if (options.where) {
      operations.where = this.whereCanonicalizer.canonicalize(options.where);
    }
    if (options.withProperties) {
      operations.withProperties = this.rdpCanonicalizer.canonicalize(options.withProperties);
    }
    if (options.union && options.union.length > 0) {
      operations.union = this.objectSetArrayCanonicalizer.canonicalizeUnion(options.union.map(os => JSON.stringify(getWireObjectSet(os))));
    }
    if (options.intersect && options.intersect.length > 0) {
      operations.intersect = this.objectSetArrayCanonicalizer.canonicalizeIntersect(options.intersect.map(os => JSON.stringify(getWireObjectSet(os))));
    }
    if (options.subtract && options.subtract.length > 0) {
      operations.subtract = this.objectSetArrayCanonicalizer.canonicalizeSubtract(options.subtract.map(os => JSON.stringify(getWireObjectSet(os))));
    }
    if (options.pivotTo) {
      operations.pivotTo = options.pivotTo;
    }
    if (options.orderBy) {
      operations.orderBy = this.orderByCanonicalizer.canonicalize(options.orderBy);
    }
    if (options.select && options.select.length > 0) {
      operations.select = this.selectCanonicalizer.canonicalize(options.select);
    }
    if (options.pageSize) {
      operations.pageSize = options.pageSize;
    }
    if (options.$loadPropertySecurityMetadata) {
      operations.loadPropertySecurity = true;
    }
    return operations;
  }
}
